# -*- coding: utf-8 -*-
# Generated by the protocol buffer compiler.  DO NOT EDIT!
# source: ncbi/datasets/v1alpha1/datasets.proto

from google.protobuf.internal import enum_type_wrapper
from google.protobuf import descriptor as _descriptor
from google.protobuf import message as _message
from google.protobuf import reflection as _reflection
from google.protobuf import symbol_database as _symbol_database
# @@protoc_insertion_point(imports)

_sym_db = _symbol_database.Default()


from google.protobuf import timestamp_pb2 as google_dot_protobuf_dot_timestamp__pb2
from ncbi.datasets.v1alpha1 import openapi_options_pb2 as ncbi_dot_datasets_dot_v1alpha1_dot_openapi__options__pb2


DESCRIPTOR = _descriptor.FileDescriptor(
  name='ncbi/datasets/v1alpha1/datasets.proto',
  package='ncbi.datasets.v1alpha1',
  syntax='proto3',
  serialized_options=b'Z\026ncbi/datasets/v1alpha1\370\001\001',
  serialized_pb=b'\n%ncbi/datasets/v1alpha1/datasets.proto\x12\x16ncbi.datasets.v1alpha1\x1a\x1fgoogle/protobuf/timestamp.proto\x1a,ncbi/datasets/v1alpha1/openapi_options.proto\"H\n\x14MethodPayloadRequest\x12\x16\n\x06method\x18\x01 \x01(\tR\x06method\x12\x18\n\x07payload\x18\x02 \x01(\tR\x07payload\"j\n AssemblyDatasetDescriptorsFilter\x12%\n\x0ereference_only\x18\x01 \x01(\x08R\rreferenceOnly\x12\x1f\n\x0brefseq_only\x18\x02 \x01(\x08R\nrefseqOnly\",\n\nAccessions\x12\x1e\n\naccessions\x18\x01 \x03(\tR\naccessions\"\xbf\x04\n\x17\x41ssemblyMetadataRequest\x12\x16\n\x05taxon\x18\x01 \x01(\tH\x00R\x05taxon\x12\x44\n\naccessions\x18\x02 \x01(\x0b\x32\".ncbi.datasets.v1alpha1.AccessionsH\x00R\naccessions\x12_\n\x0b\x62ioprojects\x18\x07 \x01(\x0b\x32;.ncbi.datasets.v1alpha1.AssemblyMetadataRequest.BioprojectsH\x00R\x0b\x62ioprojects\x12\x16\n\x05limit\x18\x03 \x01(\tH\x01R\x05limit\x12R\n\x07\x66ilters\x18\x04 \x01(\x0b\x32\x38.ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsFilterR\x07\x66ilters\x12&\n\x0ftax_exact_match\x18\x05 \x01(\x08R\rtaxExactMatch\x12\x66\n\x10returned_content\x18\x06 \x01(\x0e\x32;.ncbi.datasets.v1alpha1.AssemblyMetadataRequest.ContentTypeR\x0freturnedContent\x1a-\n\x0b\x42ioprojects\x12\x1e\n\naccessions\x18\x01 \x03(\tR\naccessions\")\n\x0b\x43ontentType\x12\x0c\n\x08\x43OMPLETE\x10\x00\x12\x0c\n\x08\x41SSM_ACC\x10\x01\x42\x04\n\x02idB\t\n\x07records\"\x97\x04\n!AssemblyDatasetDescriptorsRequest\x12\x17\n\x06tax_id\x18\x01 \x01(\rH\x00R\x05taxId\x12\x1f\n\x08org_name\x18\x05 \x01(\tB\x02\x18\x01H\x00R\x07orgName\x12\x1b\n\x08tax_name\x18\x07 \x01(\tH\x00R\x07taxName\x12/\n\x12\x61ssembly_accession\x18\x04 \x01(\tH\x00R\x11\x61ssemblyAccession\x12\x1c\n\x06\x63utoff\x18\x02 \x01(\rB\x02\x18\x01H\x01R\x06\x63utoff\x12\x16\n\x05limit\x18\x08 \x01(\tH\x01R\x05limit\x12R\n\x07\x66ilters\x18\x03 \x01(\x0b\x32\x38.ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsFilterR\x07\x66ilters\x12&\n\x0ftax_exact_match\x18\x06 \x01(\x08R\rtaxExactMatch\x12p\n\x10returned_content\x18\t \x01(\x0e\x32\x45.ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsRequest.ContentTypeR\x0freturnedContent\"5\n\x0b\x43ontentType\x12\x0c\n\x08\x43OMPLETE\x10\x00\x12\x0c\n\x08\x41SSM_ACC\x10\x01\x12\n\n\x06\x43OUNTS\x10\x02\x42\x04\n\x02idB\t\n\x07records\"\x8c\x01\n\x1a\x41ssemblyDatasetDescriptors\x12M\n\x08\x64\x61tasets\x18\x01 \x03(\x0b\x32\x31.ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorR\x08\x64\x61tasets\x12\x1f\n\x0btotal_count\x18\x02 \x01(\rR\ntotalCount\"\x98\x01\n\nBioProject\x12\x1c\n\taccession\x18\x01 \x01(\tR\taccession\x12\x14\n\x05title\x18\x02 \x01(\tR\x05title\x12)\n\x10parent_accession\x18\x03 \x01(\tR\x0fparentAccession\x12+\n\x11parent_accessions\x18\x04 \x03(\tR\x10parentAccessions\"Y\n\x11\x42ioProjectLineage\x12\x44\n\x0b\x62ioprojects\x18\x01 \x03(\x0b\x32\".ncbi.datasets.v1alpha1.BioProjectR\x0b\x62ioprojects\"\xe1\x04\n\x19\x41ssemblyDatasetDescriptor\x12-\n\x12\x61ssembly_accession\x18\x01 \x01(\tR\x11\x61ssemblyAccession\x12Z\n\x13\x62ioproject_lineages\x18\x0e \x03(\x0b\x32).ncbi.datasets.v1alpha1.BioProjectLineageR\x12\x62ioprojectLineages\x12!\n\x0c\x64isplay_name\x18\x02 \x01(\tR\x0b\x64isplayName\x12\x32\n\x03org\x18\n \x01(\x0b\x32 .ncbi.datasets.v1alpha1.OrganismR\x03org\x12 \n\x0b\x63hromosomes\x18\x04 \x03(\tR\x0b\x63hromosomes\x12+\n\x11\x61ssembly_category\x18\x05 \x01(\tR\x10\x61ssemblyCategory\x12^\n\x13\x61nnotation_metadata\x18\x0b \x01(\x0b\x32-.ncbi.datasets.v1alpha1.AnnotationForAssemblyR\x12\x61nnotationMetadata\x12%\n\x0e\x61ssembly_level\x18\x07 \x01(\tR\rassemblyLevel\x12\'\n\x0fsubmission_date\x18\x08 \x01(\tR\x0esubmissionDate\x12\x1d\n\ncontig_n50\x18\t \x01(\rR\tcontigN50\x12%\n\x0e\x65stimated_size\x18\x0c \x01(\x04R\restimatedSize\x12\x1d\n\nseq_length\x18\r \x01(\x04R\tseqLength\"\xe7\x03\n\x15\x41nnotationForAssembly\x12\x12\n\x04name\x18\x01 \x01(\tR\x04name\x12\x16\n\x06source\x18\x02 \x01(\tR\x06source\x12%\n\x0erelease_number\x18\x03 \x01(\tR\rreleaseNumber\x12!\n\x0crelease_date\x18\x04 \x01(\tR\x0breleaseDate\x12\x1d\n\nreport_url\x18\x05 \x01(\tR\treportUrl\x12\x46\n\x04\x66ile\x18\x06 \x03(\x0b\x32\x32.ncbi.datasets.v1alpha1.AnnotationForAssembly.FileR\x04\x66ile\x1au\n\x04\x46ile\x12\x46\n\x04type\x18\x01 \x01(\x0e\x32\x32.ncbi.datasets.v1alpha1.AnnotationForAssembly.TypeR\x04type\x12%\n\x0e\x65stimated_size\x18\x02 \x01(\x04R\restimatedSize\"z\n\x04Type\x12\x0b\n\x07\x44\x45\x46\x41ULT\x10\x00\x12\x0e\n\nGENOME_GFF\x10\x01\x12\x0f\n\x0bGENOME_GBFF\x10\x02\x12\x11\n\tGENOME_GB\x10\x02\x1a\x02\x08\x01\x12\r\n\tRNA_FASTA\x10\x03\x12\x0e\n\nPROT_FASTA\x10\x04\x12\x0e\n\nGENOME_GTF\x10\x05\x1a\x02\x10\x01\"\x89\x04\n\x16\x41ssemblyDatasetRequest\x12\x33\n\x13\x61ssembly_accessions\x18\x01 \x03(\tB\x02\x18\x01R\x12\x61ssemblyAccessions\x12\x1e\n\naccessions\x18\n \x03(\tR\naccessions\x12 \n\x0b\x63hromosomes\x18\x02 \x03(\tR\x0b\x63hromosomes\x12\x31\n\x12include_annotation\x18\x03 \x01(\x08\x42\x02\x18\x01R\x11includeAnnotation\x12)\n\x10\x65xclude_sequence\x18\x07 \x01(\x08R\x0f\x65xcludeSequence\x12j\n\x17include_annotation_type\x18\x08 \x03(\x0e\x32\x32.ncbi.datasets.v1alpha1.AnnotationForAssembly.TypeR\x15includeAnnotationType\x12U\n\x08hydrated\x18\t \x01(\x0e\x32\x39.ncbi.datasets.v1alpha1.AssemblyDatasetRequest.ResolutionR\x08hydrated\x12\x1f\n\x0binclude_tsv\x18\x0b \x01(\x08R\nincludeTsv\"6\n\nResolution\x12\x12\n\x0e\x46ULLY_HYDRATED\x10\x00\x12\x14\n\x10\x44\x41TA_REPORT_ONLY\x10\x01\"\x8f\x01\n\x1b\x41ssemblyDatasetAvailability\x12)\n\x10valid_assemblies\x18\x01 \x03(\tR\x0fvalidAssemblies\x12-\n\x12invalid_assemblies\x18\x02 \x03(\tR\x11invalidAssemblies\x12\x16\n\x06reason\x18\x03 \x01(\tR\x06reason\")\n\x0cTaxidRequest\x12\x19\n\x08org_name\x18\x01 \x01(\tR\x07orgName\"#\n\nTaxidReply\x12\x15\n\x06tax_id\x18\x01 \x01(\rR\x05taxId\"\x8b\x0f\n\x08Organism\x12\x15\n\x06tax_id\x18\x01 \x01(\x04R\x05taxId\x12\x19\n\x08sci_name\x18\x02 \x01(\tR\x07sciName\x12\x1f\n\x0b\x63ommon_name\x18\x03 \x01(\tR\ncommonName\x12\x1d\n\nblast_node\x18\x10 \x01(\x08R\tblastNode\x12\x14\n\x05\x62reed\x18\n \x01(\tR\x05\x62reed\x12\x1a\n\x08\x63ultivar\x18\x0b \x01(\tR\x08\x63ultivar\x12\x18\n\x07\x65\x63otype\x18\x0c \x01(\tR\x07\x65\x63otype\x12\x18\n\x07isolate\x18\r \x01(\tR\x07isolate\x12\x10\n\x03sex\x18\x0e \x01(\tR\x03sex\x12\x16\n\x06strain\x18\x0f \x01(\tR\x06strain\x12\x1f\n\x0bsearch_text\x18\x11 \x03(\tR\nsearchText\x12=\n\x04rank\x18\x05 \x01(\x0e\x32).ncbi.datasets.v1alpha1.Organism.RankTypeR\x04rank\x12\"\n\rparent_tax_id\x18\x06 \x01(\x04R\x0bparentTaxId\x12\x31\n\x0e\x61ssembly_count\x18\x07 \x01(\x04\x42\n\x18\x01\xd2\xf0\x19\x04\n\x02\x02\x03R\rassemblyCount\x12Z\n\x0f\x61ssembly_counts\x18\x13 \x01(\x0b\x32\'.ncbi.datasets.v1alpha1.Organism.CountsB\x08\xd2\xf0\x19\x04\n\x02\x02\x03R\x0e\x61ssemblyCounts\x12N\n\x06\x63ounts\x18\x16 \x03(\x0b\x32,.ncbi.datasets.v1alpha1.Organism.CountByTypeB\x08\xd2\xf0\x19\x04\n\x02\x02\x03R\x06\x63ounts\x12\x46\n\x08\x63hildren\x18\x08 \x03(\x0b\x32 .ncbi.datasets.v1alpha1.OrganismB\x08\xd2\xf0\x19\x04\n\x02\x02\x03R\x08\x63hildren\x12\x42\n\x06merged\x18\t \x03(\x0b\x32 .ncbi.datasets.v1alpha1.OrganismB\x08\xd2\xf0\x19\x04\n\x02\x02\x03R\x06merged\x12.\n\x0emerged_tax_ids\x18\x12 \x03(\x04\x42\x08\xd2\xf0\x19\x04\n\x02\x02\x03R\x0cmergedTaxIds\x12!\n\x07min_ord\x18\x14 \x01(\rB\x08\xd2\xf0\x19\x04\n\x02\x02\x03R\x06minOrd\x12!\n\x07max_ord\x18\x15 \x01(\rB\x08\xd2\xf0\x19\x04\n\x02\x02\x03R\x06maxOrd\x12\x1a\n\x03key\x18\x64 \x01(\x04\x42\x08\xd2\xf0\x19\x04\n\x02\x02\x03R\x03key\x12\x1e\n\x05title\x18\x65 \x01(\tB\x08\xd2\xf0\x19\x04\n\x02\x02\x03R\x05title\x12\x1c\n\x04icon\x18\x66 \x01(\x08\x42\x08\xd2\xf0\x19\x04\n\x02\x02\x03R\x04icon\x1a\x36\n\x06\x43ounts\x12\x12\n\x04node\x18\x01 \x01(\rR\x04node\x12\x18\n\x07subtree\x18\x02 \x01(\rR\x07subtree\x1a\x8e\x01\n\x0b\x43ountByType\x12>\n\x04type\x18\x01 \x01(\x0e\x32*.ncbi.datasets.v1alpha1.Organism.CountTypeR\x04type\x12?\n\x06\x63ounts\x18\x02 \x01(\x0b\x32\'.ncbi.datasets.v1alpha1.Organism.CountsR\x06\x63ounts\"\x9d\x05\n\x08RankType\x12\x0b\n\x07NO_RANK\x10\x00\x12\x10\n\x0cSUPERKINGDOM\x10\x01\x12\x0b\n\x07KINGDOM\x10\x02\x12\x0e\n\nSUBKINGDOM\x10\x03\x12\x0f\n\x0bSUPERPHYLUM\x10\x04\x12\r\n\tSUBPHYLUM\x10\x05\x12\n\n\x06PHYLUM\x10\x06\x12\t\n\x05\x43LADE\x10\x1f\x12\x0e\n\nSUPERCLASS\x10\x07\x12\t\n\x05\x43LASS\x10\x08\x12\x0c\n\x08SUBCLASS\x10\t\x12\x0e\n\nINFRACLASS\x10\n\x12\n\n\x06\x43OHORT\x10\x0b\x12\r\n\tSUBCOHORT\x10\x0c\x12\x0e\n\nSUPERORDER\x10\r\x12\t\n\x05ORDER\x10\x0e\x12\x0c\n\x08SUBORDER\x10\x0f\x12\x0e\n\nINFRAORDER\x10\x10\x12\r\n\tPARVORDER\x10\x11\x12\x0f\n\x0bSUPERFAMILY\x10\x12\x12\n\n\x06\x46\x41MILY\x10\x13\x12\r\n\tSUBFAMILY\x10\x14\x12\t\n\x05GENUS\x10\x15\x12\x0c\n\x08SUBGENUS\x10\x16\x12\x11\n\rSPECIES_GROUP\x10\x17\x12\x14\n\x10SPECIES_SUBGROUP\x10\x18\x12\x0b\n\x07SPECIES\x10\x19\x12\x0e\n\nSUBSPECIES\x10\x1a\x12\t\n\x05TRIBE\x10\x1b\x12\x0c\n\x08SUBTRIBE\x10\x1c\x12\t\n\x05\x46ORMA\x10\x1d\x12\x0c\n\x08VARIETAS\x10\x1e\x12\x0b\n\x06STRAIN\x10\xc0\x02\x12\x0c\n\x07SECTION\x10\xca\x02\x12\x0f\n\nSUBSECTION\x10\xd4\x02\x12\x0f\n\nPATHOGROUP\x10\xde\x02\x12\x0f\n\nSUBVARIETY\x10\xe8\x02\x12\r\n\x08GENOTYPE\x10\xf2\x02\x12\r\n\x08SEROTYPE\x10\xfc\x02\x12\x0c\n\x07ISOLATE\x10\x86\x03\x12\n\n\x05MORPH\x10\x90\x03\x12\x0b\n\x06SERIES\x10\x9a\x03\x12\x14\n\x0f\x46ORMA_SPECIALIS\x10\xa4\x03\x12\x0e\n\tSEROGROUP\x10\xae\x03\x12\x0c\n\x07\x42IOTYPE\x10\xb8\x03\"U\n\tCountType\x12\x1a\n\x16\x43OUNT_TYPE_UNSPECIFIED\x10\x00\x12\x17\n\x13\x43OUNT_TYPE_ASSEMBLY\x10\x01\x12\x13\n\x0f\x43OUNT_TYPE_GENE\x10\x02\"t\n\x0eTaxTreeRequest\x12\x1b\n\x06tax_id\x18\x01 \x01(\x04\x42\x02\x18\x01H\x00R\x05taxId\x12!\n\ttax_token\x18\x02 \x01(\tB\x02\x18\x01H\x00R\x08taxToken\x12\x16\n\x05taxon\x18\x03 \x01(\tH\x00R\x05taxonB\n\n\x08taxonomy\"l\n\x14OrganismQueryRequest\x12+\n\x0eorganism_query\x18\x01 \x01(\tB\x02\x18\x01H\x00R\rorganismQuery\x12!\n\x0btaxon_query\x18\x02 \x01(\tH\x00R\ntaxonQueryB\x04\n\x02id\"\xae\x01\n\rSciNameAndIds\x12[\n\x10sci_name_and_ids\x18\x01 \x03(\x0b\x32\x32.ncbi.datasets.v1alpha1.SciNameAndIds.SciNameAndIdR\rsciNameAndIds\x1a@\n\x0cSciNameAndId\x12\x19\n\x08sci_name\x18\x01 \x01(\tR\x07sciName\x12\x15\n\x06tax_id\x18\x02 \x01(\x04R\x05taxId\"\xd8\x01\n\x0fOrthologRequest\x12\x17\n\x07gene_id\x18\x01 \x01(\rR\x06geneId\x12^\n\x10returned_content\x18\x02 \x01(\x0e\x32\x33.ncbi.datasets.v1alpha1.OrthologRequest.ContentTypeR\x0freturnedContent\x12!\n\x0ctaxon_filter\x18\x03 \x03(\tR\x0btaxonFilter\")\n\x0b\x43ontentType\x12\x0c\n\x08\x43OMPLETE\x10\x00\x12\x0c\n\x08IDS_ONLY\x10\x01\"\x89\x01\n\x0bOrthologSet\x12&\n\x0fortholog_set_id\x18\x01 \x01(\rR\rorthologSetId\x12\x16\n\x06method\x18\x02 \x01(\tR\x06method\x12:\n\x05genes\x18\x03 \x01(\x0b\x32$.ncbi.datasets.v1alpha1.GeneMetadataR\x05genes\"\xe5\x07\n\x12GeneDatasetRequest\x12\x19\n\x08gene_ids\x18\x01 \x03(\rR\x07geneIds\x12\x1e\n\naccessions\x18\n \x03(\tR\naccessions\x12\x66\n\x11symbols_for_taxon\x18\t \x01(\x0b\x32:.ncbi.datasets.v1alpha1.GeneDatasetRequest.SymbolsForTaxonR\x0fsymbolsForTaxon\x12\x14\n\x05taxon\x18\x03 \x01(\tR\x05taxon\x12U\n\x17include_annotation_type\x18\x06 \x03(\x0e\x32\x1d.ncbi.datasets.v1alpha1.FastaR\x15includeAnnotationType\x12\x61\n\x10returned_content\x18\x07 \x01(\x0e\x32\x36.ncbi.datasets.v1alpha1.GeneDatasetRequest.ContentTypeR\x0freturnedContent\x12P\n\x0bsort_schema\x18\x08 \x01(\x0b\x32/.ncbi.datasets.v1alpha1.GeneDatasetRequest.SortR\nsortSchema\x12\x14\n\x05limit\x18\x0b \x01(\tR\x05limit\x12!\n\x0c\x66\x61sta_filter\x18\x0c \x03(\tR\x0b\x66\x61staFilter\x1a\x41\n\x0fSymbolsForTaxon\x12\x18\n\x07symbols\x18\x01 \x03(\tR\x07symbols\x12\x14\n\x05taxon\x18\x02 \x01(\tR\x05taxon\x1a$\n\x07GeneIds\x12\x19\n\x08gene_ids\x18\x01 \x03(\rR\x07geneIds\x1a,\n\nAccessions\x12\x1e\n\naccessions\x18\x01 \x03(\tR\naccessions\x1a\x97\x01\n\x04Sort\x12J\n\x05\x66ield\x18\x01 \x01(\x0e\x32\x34.ncbi.datasets.v1alpha1.GeneDatasetRequest.SortFieldR\x05\x66ield\x12\x43\n\tdirection\x18\x02 \x01(\x0e\x32%.ncbi.datasets.v1alpha1.SortDirectionR\tdirection\")\n\x0b\x43ontentType\x12\x0c\n\x08\x43OMPLETE\x10\x00\x12\x0c\n\x08IDS_ONLY\x10\x01\"u\n\tSortField\x12\x1a\n\x16SORT_FIELD_UNSPECIFIED\x10\x00\x12\x16\n\x12SORT_FIELD_GENE_ID\x10\x01\x12\x18\n\x14SORT_FIELD_GENE_TYPE\x10\x02\x12\x1a\n\x16SORT_FIELD_GENE_SYMBOL\x10\x03\"\xbb\x01\n\x05Range\x12\x14\n\x05\x62\x65gin\x18\x01 \x01(\x04R\x05\x62\x65gin\x12\x10\n\x03\x65nd\x18\x02 \x01(\x04R\x03\x65nd\x12\x45\n\x0borientation\x18\x03 \x01(\x0e\x32#.ncbi.datasets.v1alpha1.OrientationR\x0borientation\x12\x14\n\x05order\x18\x04 \x01(\rR\x05order\x12-\n\x12ribosomal_slippage\x18\x05 \x01(\x05R\x11ribosomalSlippage\"o\n\x0bSeqRangeSet\x12+\n\x11\x61\x63\x63\x65ssion_version\x18\x01 \x01(\tR\x10\x61\x63\x63\x65ssionVersion\x12\x33\n\x05range\x18\x02 \x03(\x0b\x32\x1d.ncbi.datasets.v1alpha1.RangeR\x05range\"\xeb\x01\n\x0fGenomicLocation\x12:\n\x19genomic_accession_version\x18\x01 \x01(\tR\x17genomicAccessionVersion\x12#\n\rsequence_name\x18\x02 \x01(\tR\x0csequenceName\x12\x42\n\rgenomic_range\x18\x03 \x01(\x0b\x32\x1d.ncbi.datasets.v1alpha1.RangeR\x0cgenomicRange\x12\x33\n\x05\x65xons\x18\x04 \x03(\x0b\x32\x1d.ncbi.datasets.v1alpha1.RangeR\x05\x65xons\"h\n\rMaturePeptide\x12+\n\x11\x61\x63\x63\x65ssion_version\x18\x01 \x01(\tR\x10\x61\x63\x63\x65ssionVersion\x12\x12\n\x04name\x18\x02 \x01(\tR\x04name\x12\x16\n\x06length\x18\x03 \x01(\rR\x06length\"\xfe\x01\n\x07Protein\x12+\n\x11\x61\x63\x63\x65ssion_version\x18\x01 \x01(\tR\x10\x61\x63\x63\x65ssionVersion\x12\x12\n\x04name\x18\x02 \x01(\tR\x04name\x12\x16\n\x06length\x18\x03 \x01(\rR\x06length\x12!\n\x0cisoform_name\x18\x05 \x01(\tR\x0bisoformName\x12\'\n\x0f\x65nsembl_protein\x18\x06 \x01(\tR\x0e\x65nsemblProtein\x12N\n\x0fmature_peptides\x18\x04 \x03(\x0b\x32%.ncbi.datasets.v1alpha1.MaturePeptideR\x0ematurePeptides\"\xa3\x05\n\nTranscript\x12+\n\x11\x61\x63\x63\x65ssion_version\x18\x01 \x01(\tR\x10\x61\x63\x63\x65ssionVersion\x12\x12\n\x04name\x18\x02 \x01(\tR\x04name\x12\x16\n\x06length\x18\x03 \x01(\rR\x06length\x12L\n\rgenomic_range\x18\x04 \x01(\x0b\x32#.ncbi.datasets.v1alpha1.SeqRangeSetB\x02\x18\x01R\x0cgenomicRange\x12=\n\x05\x65xons\x18\x05 \x01(\x0b\x32#.ncbi.datasets.v1alpha1.SeqRangeSetB\x02\x18\x01R\x05\x65xons\x12\x35\n\x03\x63\x64s\x18\x06 \x01(\x0b\x32#.ncbi.datasets.v1alpha1.SeqRangeSetR\x03\x63\x64s\x12T\n\x11genomic_locations\x18\n \x03(\x0b\x32\'.ncbi.datasets.v1alpha1.GenomicLocationR\x10genomicLocations\x12-\n\x12\x65nsembl_transcript\x18\x08 \x01(\tR\x11\x65nsemblTranscript\x12\x39\n\x07protein\x18\x07 \x01(\x0b\x32\x1f.ncbi.datasets.v1alpha1.ProteinR\x07protein\x12\x45\n\x04type\x18\t \x01(\x0e\x32\x31.ncbi.datasets.v1alpha1.Transcript.TranscriptTypeR\x04type\"q\n\x0eTranscriptType\x12\x0b\n\x07UNKNOWN\x10\x00\x12\x12\n\x0ePROTEIN_CODING\x10\x01\x12\x0e\n\nNON_CODING\x10\x02\x12\x18\n\x14PROTEIN_CODING_MODEL\x10\x03\x12\x14\n\x10NON_CODING_MODEL\x10\x04\"\x85\x02\n\rGenomicRegion\x12\x42\n\ngene_range\x18\x01 \x01(\x0b\x32#.ncbi.datasets.v1alpha1.SeqRangeSetR\tgeneRange\x12K\n\x04type\x18\x02 \x01(\x0e\x32\x37.ncbi.datasets.v1alpha1.GenomicRegion.GenomicRegionTypeR\x04type\"c\n\x11GenomicRegionType\x12\x0b\n\x07UNKNOWN\x10\x00\x12\x0f\n\x0bREFSEQ_GENE\x10\x01\x12\x0e\n\nPSEUDOGENE\x10\x02\x12\x15\n\x11\x42IOLOGICAL_REGION\x10\x03\x12\t\n\x05OTHER\x10\x04\"G\n\x13\x41nnotatedAssemblies\x12\x1c\n\taccession\x18\x01 \x01(\tR\taccession\x12\x12\n\x04name\x18\x02 \x01(\tR\x04name\"\xaf\x01\n\nAnnotation\x12!\n\x0crelease_name\x18\x01 \x01(\tR\x0breleaseName\x12!\n\x0crelease_date\x18\x02 \x01(\tR\x0breleaseDate\x12[\n\x13\x61ssemblies_in_scope\x18\x03 \x03(\x0b\x32+.ncbi.datasets.v1alpha1.AnnotatedAssembliesR\x11\x61ssembliesInScope\"\xa2\x0b\n\x0eGeneDescriptor\x12\x17\n\x07gene_id\x18\x01 \x01(\x04R\x06geneId\x12\x16\n\x06symbol\x18\x02 \x01(\tR\x06symbol\x12 \n\x0b\x64\x65scription\x18\x03 \x01(\tR\x0b\x64\x65scription\x12\x15\n\x06tax_id\x18\x04 \x01(\x04R\x05taxId\x12\x18\n\x07taxname\x18\x05 \x01(\tR\x07taxname\x12\x43\n\x04type\x18\x06 \x01(\x0e\x32/.ncbi.datasets.v1alpha1.GeneDescriptor.GeneTypeR\x04type\x12I\n\x08rna_type\x18\x0b \x01(\x0e\x32..ncbi.datasets.v1alpha1.GeneDescriptor.RnaTypeR\x07rnaType\x12\x45\n\x0borientation\x18\x07 \x01(\x0e\x32#.ncbi.datasets.v1alpha1.OrientationR\x0borientation\x12J\n\x0egenomic_ranges\x18\x08 \x03(\x0b\x32#.ncbi.datasets.v1alpha1.SeqRangeSetR\rgenomicRanges\x12V\n\x13reference_standards\x18\x15 \x03(\x0b\x32%.ncbi.datasets.v1alpha1.GenomicRegionR\x12referenceStandards\x12N\n\x0fgenomic_regions\x18\x16 \x03(\x0b\x32%.ncbi.datasets.v1alpha1.GenomicRegionR\x0egenomicRegions\x12\x44\n\x0btranscripts\x18\t \x03(\x0b\x32\".ncbi.datasets.v1alpha1.TranscriptR\x0btranscripts\x12;\n\x08proteins\x18\n \x03(\x0b\x32\x1f.ncbi.datasets.v1alpha1.ProteinR\x08proteins\x12\x1f\n\x0b\x63ommon_name\x18\x0c \x01(\tR\ncommonName\x12\"\n\nchromosome\x18\r \x01(\tB\x02\x18\x01R\nchromosome\x12 \n\x0b\x63hromosomes\x18\x0e \x03(\tR\x0b\x63hromosomes\x12\x64\n\x16nomenclature_authority\x18\x0f \x01(\x0b\x32-.ncbi.datasets.v1alpha1.NomenclatureAuthorityR\x15nomenclatureAuthority\x12\x32\n\x15swiss_prot_accessions\x18\x10 \x03(\tR\x13swissProtAccessions\x12(\n\x10\x65nsembl_gene_ids\x18\x11 \x03(\tR\x0e\x65nsemblGeneIds\x12\x19\n\x08omim_ids\x18\x12 \x03(\tR\x07omimIds\x12\x1a\n\x08synonyms\x18\x13 \x03(\tR\x08synonyms\x12(\n\x10replaced_gene_id\x18\x14 \x01(\x04R\x0ereplacedGeneId\x12\x44\n\x0b\x61nnotations\x18\x17 \x03(\x0b\x32\".ncbi.datasets.v1alpha1.AnnotationR\x0b\x61nnotations\"\xb8\x01\n\x08GeneType\x12\x0b\n\x07UNKNOWN\x10\x00\x12\x08\n\x04tRNA\x10\x01\x12\x08\n\x04rRNA\x10\x02\x12\t\n\x05snRNA\x10\x03\x12\t\n\x05scRNA\x10\x04\x12\n\n\x06snoRNA\x10\x05\x12\x12\n\x0ePROTEIN_CODING\x10\x06\x12\n\n\x06PSEUDO\x10\x07\x12\x0e\n\nTRANSPOSON\x10\x08\x12\x0b\n\x07miscRNA\x10\t\x12\t\n\x05ncRNA\x10\n\x12\x15\n\x11\x42IOLOGICAL_REGION\x10\x0b\x12\n\n\x05OTHER\x10\xff\x01\"1\n\x07RnaType\x12\x0f\n\x0brna_UNKNOWN\x10\x00\x12\n\n\x06premsg\x10\x01\x12\t\n\x05tmRna\x10\x02\"U\n\x15NomenclatureAuthority\x12\x1c\n\tauthority\x18\x01 \x01(\tR\tauthority\x12\x1e\n\nidentifier\x18\x02 \x01(\tR\nidentifier\"\xbc\x05\n\x05\x45rror\x12\x61\n\x13\x61ssembly_error_code\x18\x06 \x01(\x0e\x32/.ncbi.datasets.v1alpha1.Error.AssemblyErrorCodeH\x00R\x11\x61ssemblyErrorCode\x12U\n\x0fgene_error_code\x18\x07 \x01(\x0e\x32+.ncbi.datasets.v1alpha1.Error.GeneErrorCodeH\x00R\rgeneErrorCode\x12X\n\x10virus_error_code\x18\x08 \x01(\x0e\x32,.ncbi.datasets.v1alpha1.Error.VirusErrorCodeH\x00R\x0evirusErrorCode\x12\x16\n\x06reason\x18\x05 \x01(\tR\x06reason\x12\x18\n\x07message\x18\x02 \x01(\tR\x07message\x12/\n\x11valid_identifiers\x18\x03 \x03(\tB\x02\x18\x01R\x10validIdentifiers\x12/\n\x13invalid_identifiers\x18\x04 \x03(\tR\x12invalidIdentifiers\"\x86\x01\n\x11\x41ssemblyErrorCode\x12\x1f\n\x1bUNKNOWN_ASSEMBLY_ERROR_CODE\x10\x00\x12\x1a\n\x16INVALID_BIOPROJECT_IDS\x10\x01\x12!\n\x1dNO_ASSEMBLIES_FOR_BIOPROJECTS\x10\x02\x12\x11\n\rINVALID_TAXON\x10\x03\"J\n\rGeneErrorCode\x12\x1b\n\x17UNKNOWN_GENE_ERROR_CODE\x10\x00\x12\x1c\n\x18INCOMPLETE_LOOKUP_SYMBOL\x10\x05\".\n\x0eVirusErrorCode\x12\x1c\n\x18UNKNOWN_VIRUS_ERROR_CODE\x10\x00\x42\x06\n\x04\x63ode\"\xe3\x04\n\x07Warning\x12]\n\x11gene_warning_code\x18\x07 \x01(\x0e\x32/.ncbi.datasets.v1alpha1.Warning.GeneWarningCodeH\x00R\x0fgeneWarningCode\x12\x16\n\x06reason\x18\x02 \x01(\tR\x06reason\x12\x18\n\x07message\x18\x03 \x01(\tR\x07message\x12K\n\x0breplaced_id\x18\x04 \x01(\x0b\x32*.ncbi.datasets.v1alpha1.Warning.ReplacedIdR\nreplacedId\x12\x37\n\x17unrecognized_identifier\x18\x05 \x01(\tR\x16unrecognizedIdentifier\x1a\x46\n\nReplacedId\x12\x1c\n\trequested\x18\x01 \x01(\tR\trequested\x12\x1a\n\x08returned\x18\x02 \x01(\tR\x08returned\"\xf0\x01\n\x0fGeneWarningCode\x12\x1d\n\x19UNKNOWN_GENE_WARNING_CODE\x10\x00\x12\x1e\n\x1a\x41\x43\x43\x45SSION_VERSION_MISMATCH\x10\x01\x12\x14\n\x10REPLACED_GENE_ID\x10\x02\x12\x18\n\x14\x44ISCONTINUED_GENE_ID\x10\x03\x12\x18\n\x14UNRECOGNIZED_GENE_ID\x10\x04\x12\x1c\n\x18UNRECOGNIZED_GENE_SYMBOL\x10\x05\x12\x1a\n\x16UNRECOGNIZED_ACCESSION\x10\x06\x12\x1a\n\x16UNRECOGNIZED_TAX_TOKEN\x10\x07\x42\x06\n\x04\x63ode\"\x88\x01\n\x07Message\x12\x35\n\x05\x65rror\x18\x01 \x01(\x0b\x32\x1d.ncbi.datasets.v1alpha1.ErrorH\x00R\x05\x65rror\x12;\n\x07warning\x18\x02 \x01(\x0b\x32\x1f.ncbi.datasets.v1alpha1.WarningH\x00R\x07warningB\t\n\x07message\"\x92\x02\n\tGeneMatch\x12?\n\x08messages\x18\x01 \x03(\x0b\x32\x1f.ncbi.datasets.v1alpha1.MessageB\x02\x18\x01R\x08messages\x12;\n\x08warnings\x18\x03 \x03(\x0b\x32\x1f.ncbi.datasets.v1alpha1.WarningR\x08warnings\x12\x14\n\x05query\x18\x04 \x03(\tR\x05query\x12:\n\x04gene\x18\x02 \x01(\x0b\x32&.ncbi.datasets.v1alpha1.GeneDescriptorR\x04gene\x12\x35\n\x06\x65rrors\x18\x05 \x03(\x0b\x32\x1d.ncbi.datasets.v1alpha1.ErrorR\x06\x65rrors\"\xa5\x01\n\x0cGeneMetadata\x12;\n\x08messages\x18\x01 \x03(\x0b\x32\x1f.ncbi.datasets.v1alpha1.MessageR\x08messages\x12\x37\n\x05genes\x18\x02 \x03(\x0b\x32!.ncbi.datasets.v1alpha1.GeneMatchR\x05genes\x12\x1f\n\x0btotal_count\x18\x03 \x01(\rR\ntotalCount\"\x9b\x01\n\rAssemblyMatch\x12;\n\x08messages\x18\x01 \x03(\x0b\x32\x1f.ncbi.datasets.v1alpha1.MessageR\x08messages\x12M\n\x08\x61ssembly\x18\x02 \x01(\x0b\x32\x31.ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorR\x08\x61ssembly\"\xb7\x01\n\x10\x41ssemblyMetadata\x12;\n\x08messages\x18\x01 \x03(\x0b\x32\x1f.ncbi.datasets.v1alpha1.MessageR\x08messages\x12\x45\n\nassemblies\x18\x02 \x03(\x0b\x32%.ncbi.datasets.v1alpha1.AssemblyMatchR\nassemblies\x12\x1f\n\x0btotal_count\x18\x03 \x01(\rR\ntotalCount\"\x86\x01\n\x0fGeneDescriptors\x12<\n\x05genes\x18\x01 \x03(\x0b\x32&.ncbi.datasets.v1alpha1.GeneDescriptorR\x05genes\x12\x35\n\x06\x65rrors\x18\x02 \x03(\x0b\x32\x1d.ncbi.datasets.v1alpha1.ErrorR\x06\x65rrors\"y\n\x12\x41nnotationForVirus\"c\n\x04Type\x12\x0b\n\x07\x44\x45\x46\x41ULT\x10\x00\x12\r\n\tCDS_FASTA\x10\x01\x12\x0e\n\nPROT_FASTA\x10\x02\x12\x0f\n\x0bGENOME_GBFF\x10\x03\x12\x0f\n\x0bGENOME_GPFF\x10\x04\x12\r\n\tPDB_FILES\x10\x05\"\x95\x05\n\x13VirusDatasetRequest\x12\x1b\n\x06tax_id\x18\x01 \x01(\rB\x02\x18\x01H\x00R\x05taxId\x12\x1f\n\x08tax_name\x18\x02 \x01(\tB\x02\x18\x01H\x00R\x07taxName\x12\x1e\n\taccession\x18\x03 \x01(\tH\x00R\taccession\x12\x16\n\x05taxon\x18\x0b \x01(\tH\x00R\x05taxon\x12\x1f\n\x0brefseq_only\x18\x04 \x01(\x08R\nrefseqOnly\x12%\n\x0e\x61nnotated_only\x18\x05 \x01(\x08R\rannotatedOnly\x12\x41\n\x0ereleased_since\x18\x06 \x01(\x0b\x32\x1a.google.protobuf.TimestampR\rreleasedSince\x12\x12\n\x04host\x18\x07 \x01(\tR\x04host\x12!\n\x0cgeo_location\x18\x08 \x01(\tR\x0bgeoLocation\x12#\n\rcomplete_only\x18\t \x01(\x08R\x0c\x63ompleteOnly\x12J\n\x0ctable_fields\x18\n \x03(\x0e\x32\'.ncbi.datasets.v1alpha1.VirusTableFieldR\x0btableFields\x12)\n\x10\x65xclude_sequence\x18\x0c \x01(\x08R\x0f\x65xcludeSequence\x12g\n\x17include_annotation_type\x18\r \x03(\x0e\x32/.ncbi.datasets.v1alpha1.AnnotationForVirus.TypeR\x15includeAnnotationType\x12;\n\x06\x66ormat\x18\x0e \x01(\x0e\x32#.ncbi.datasets.v1alpha1.TableFormatR\x06\x66ormatB\x04\n\x02id\"\xbc\x04\n\x1aSars2ProteinDatasetRequest\x12\x1a\n\x08proteins\x18\x01 \x03(\tR\x08proteins\x12\x1f\n\x0brefseq_only\x18\x04 \x01(\x08R\nrefseqOnly\x12%\n\x0e\x61nnotated_only\x18\x05 \x01(\x08R\rannotatedOnly\x12\x41\n\x0ereleased_since\x18\x06 \x01(\x0b\x32\x1a.google.protobuf.TimestampR\rreleasedSince\x12\x12\n\x04host\x18\x07 \x01(\tR\x04host\x12!\n\x0cgeo_location\x18\x08 \x01(\tR\x0bgeoLocation\x12#\n\rcomplete_only\x18\t \x01(\x08R\x0c\x63ompleteOnly\x12J\n\x0ctable_fields\x18\n \x03(\x0e\x32\'.ncbi.datasets.v1alpha1.VirusTableFieldR\x0btableFields\x12)\n\x10\x65xclude_sequence\x18\x0c \x01(\x08R\x0f\x65xcludeSequence\x12g\n\x17include_annotation_type\x18\r \x03(\x0e\x32/.ncbi.datasets.v1alpha1.AnnotationForVirus.TypeR\x15includeAnnotationType\x12;\n\x06\x66ormat\x18\x0e \x01(\x0e\x32#.ncbi.datasets.v1alpha1.TableFormatR\x06\x66ormat\"\xa2\x06\n\x0f\x44ownloadSummary\x12!\n\x0crecord_count\x18\x01 \x01(\rR\x0brecordCount\x12)\n\x0e\x61ssembly_count\x18\x02 \x01(\rB\x02\x18\x01R\rassemblyCount\x12J\n\x13resource_updated_on\x18\x04 \x01(\x0b\x32\x1a.google.protobuf.TimestampR\x11resourceUpdatedOn\x12L\n\x08hydrated\x18\x05 \x01(\x0b\x32\x30.ncbi.datasets.v1alpha1.DownloadSummary.HydratedR\x08hydrated\x12R\n\ndehydrated\x18\x06 \x01(\x0b\x32\x32.ncbi.datasets.v1alpha1.DownloadSummary.DehydratedR\ndehydrated\x12\x39\n\x06\x65rrors\x18\x07 \x03(\x0b\x32\x1d.ncbi.datasets.v1alpha1.ErrorB\x02\x18\x01R\x06\x65rrors\x12;\n\x08messages\x18\x08 \x03(\x0b\x32\x1f.ncbi.datasets.v1alpha1.MessageR\x08messages\x1a\x8c\x01\n\x08Hydrated\x12\x33\n\x16\x65stimated_file_size_mb\x18\x01 \x01(\rR\x13\x65stimatedFileSizeMb\x12\x10\n\x03url\x18\x02 \x01(\tR\x03url\x12\x39\n\x19\x63li_download_command_line\x18\x03 \x01(\tR\x16\x63liDownloadCommandLine\x1a\xcb\x01\n\nDehydrated\x12\x33\n\x16\x65stimated_file_size_mb\x18\x01 \x01(\rR\x13\x65stimatedFileSizeMb\x12\x10\n\x03url\x18\x02 \x01(\tR\x03url\x12\x39\n\x19\x63li_download_command_line\x18\x03 \x01(\tR\x16\x63liDownloadCommandLine\x12;\n\x1a\x63li_rehydrate_command_line\x18\x04 \x01(\tR\x17\x63liRehydrateCommandLine\"#\n\rTabularOutput\x12\x12\n\x04\x64\x61ta\x18\x01 \x01(\tR\x04\x64\x61ta\"\xbf\x03\n\x18MicroBiggeDatasetRequest\x12,\n\x11opaque_solr_query\x18\x01 \x01(\tH\x00R\x0fopaqueSolrQuery\x12O\n\x05\x66iles\x18\x02 \x03(\x0e\x32\x39.ncbi.datasets.v1alpha1.MicroBiggeDatasetRequest.FileTypeR\x05\x66iles\x12u\n\x14\x65lement_flank_config\x18\x03 \x01(\x0b\x32\x43.ncbi.datasets.v1alpha1.MicroBiggeDatasetRequest.ElementFlankConfigR\x12\x65lementFlankConfig\x1a,\n\x12\x45lementFlankConfig\x12\x16\n\x06length\x18\x01 \x01(\rR\x06length\"v\n\x08\x46ileType\x12\x19\n\x15\x46ILE_TYPE_UNSPECIFIED\x10\x00\x12\x11\n\relement_fasta\x10\x01\x12\x17\n\x13\x65lement_flank_fasta\x10\x02\x12\x10\n\x0c\x63ontig_fasta\x10\x03\x12\x11\n\rprotein_fasta\x10\x04\x42\x07\n\x05query\"\x7f\n\x0cVersionReply\x12\x18\n\x07version\x18\x01 \x01(\tR\x07version\x12\x1b\n\tmajor_ver\x18\x02 \x01(\rR\x08majorVer\x12\x1b\n\tminor_ver\x18\x03 \x01(\rR\x08minorVer\x12\x1b\n\tpatch_ver\x18\x04 \x01(\rR\x08patchVer\"\xba\x02\n\x15ProkaryoteGeneRequest\x12\x1e\n\naccessions\x18\x01 \x03(\tR\naccessions\x12U\n\x17include_annotation_type\x18\x02 \x03(\x0e\x32\x1d.ncbi.datasets.v1alpha1.FastaR\x15includeAnnotationType\x12i\n\x11gene_flank_config\x18\x03 \x01(\x0b\x32=.ncbi.datasets.v1alpha1.ProkaryoteGeneRequest.GeneFlankConfigR\x0fgeneFlankConfig\x12\x14\n\x05taxon\x18\x04 \x01(\tR\x05taxon\x1a)\n\x0fGeneFlankConfig\x12\x16\n\x06length\x18\x01 \x01(\rR\x06length*l\n\rSortDirection\x12\x1e\n\x1aSORT_DIRECTION_UNSPECIFIED\x10\x00\x12\x1c\n\x18SORT_DIRECTION_ASCENDING\x10\x01\x12\x1d\n\x19SORT_DIRECTION_DESCENDING\x10\x02*f\n\x05\x46\x61sta\x12\x15\n\x11\x46\x41STA_UNSPECIFIED\x10\x00\x12\x0e\n\nFASTA_GENE\x10\x01\x12\r\n\tFASTA_RNA\x10\x02\x12\x11\n\rFASTA_PROTEIN\x10\x03\x12\x14\n\x10\x46\x41STA_GENE_FLANK\x10\x04*,\n\x0bOrientation\x12\x08\n\x04none\x10\x00\x12\x08\n\x04plus\x10\x01\x12\t\n\x05minus\x10\x02*\xa1\x03\n\x0fVirusTableField\x12\x0f\n\x0bunspecified\x10\x00\x12\x18\n\x14nucleotide_accession\x10\x01\x12\x12\n\x0especies_tax_id\x10\x02\x12\x10\n\x0cspecies_name\x10\x03\x12\t\n\x05genus\x10\x04\x12\n\n\x06\x66\x61mily\x10\x05\x12\x15\n\x11nucleotide_length\x10\x07\x12\x10\n\x0cisolate_name\x10\x08\x12\x11\n\rsequence_type\x10\t\x12\x14\n\x10nuc_completeness\x10\n\x12\x10\n\x0cgeo_location\x10\x0b\x12\x0c\n\x08us_state\x10\x0c\x12\r\n\thost_name\x10\r\x12\x0f\n\x0bhost_tax_id\x10\x0e\x12\x13\n\x0f\x63ollection_date\x10\x0f\x12\x0e\n\nbioproject\x10\x10\x12\r\n\tbiosample\x10\x11\x12\x14\n\x10polyprotein_name\x10\x13\x12\x10\n\x0cprotein_name\x10\x14\x12\x15\n\x11protein_accession\x10\x15\x12\x13\n\x0fprotein_synonym\x10\x16\x12\x0c\n\x08\x63\x64s_span\x10\x17**\n\x0bTableFormat\x12\x07\n\x03tsv\x10\x00\x12\x07\n\x03\x63sv\x10\x01\x12\t\n\x05jsonl\x10\x02\x42\x1bZ\x16ncbi/datasets/v1alpha1\xf8\x01\x01\x62\x06proto3'
  ,
  dependencies=[google_dot_protobuf_dot_timestamp__pb2.DESCRIPTOR,ncbi_dot_datasets_dot_v1alpha1_dot_openapi__options__pb2.DESCRIPTOR,])

_SORTDIRECTION = _descriptor.EnumDescriptor(
  name='SortDirection',
  full_name='ncbi.datasets.v1alpha1.SortDirection',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='SORT_DIRECTION_UNSPECIFIED', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SORT_DIRECTION_ASCENDING', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SORT_DIRECTION_DESCENDING', index=2, number=2,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=16538,
  serialized_end=16646,
)
_sym_db.RegisterEnumDescriptor(_SORTDIRECTION)

SortDirection = enum_type_wrapper.EnumTypeWrapper(_SORTDIRECTION)
_FASTA = _descriptor.EnumDescriptor(
  name='Fasta',
  full_name='ncbi.datasets.v1alpha1.Fasta',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='FASTA_UNSPECIFIED', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='FASTA_GENE', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='FASTA_RNA', index=2, number=2,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='FASTA_PROTEIN', index=3, number=3,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='FASTA_GENE_FLANK', index=4, number=4,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=16648,
  serialized_end=16750,
)
_sym_db.RegisterEnumDescriptor(_FASTA)

Fasta = enum_type_wrapper.EnumTypeWrapper(_FASTA)
_ORIENTATION = _descriptor.EnumDescriptor(
  name='Orientation',
  full_name='ncbi.datasets.v1alpha1.Orientation',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='none', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='plus', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='minus', index=2, number=2,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=16752,
  serialized_end=16796,
)
_sym_db.RegisterEnumDescriptor(_ORIENTATION)

Orientation = enum_type_wrapper.EnumTypeWrapper(_ORIENTATION)
_VIRUSTABLEFIELD = _descriptor.EnumDescriptor(
  name='VirusTableField',
  full_name='ncbi.datasets.v1alpha1.VirusTableField',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='unspecified', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='nucleotide_accession', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='species_tax_id', index=2, number=2,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='species_name', index=3, number=3,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='genus', index=4, number=4,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='family', index=5, number=5,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='nucleotide_length', index=6, number=7,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='isolate_name', index=7, number=8,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='sequence_type', index=8, number=9,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='nuc_completeness', index=9, number=10,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='geo_location', index=10, number=11,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='us_state', index=11, number=12,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='host_name', index=12, number=13,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='host_tax_id', index=13, number=14,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='collection_date', index=14, number=15,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='bioproject', index=15, number=16,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='biosample', index=16, number=17,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='polyprotein_name', index=17, number=19,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='protein_name', index=18, number=20,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='protein_accession', index=19, number=21,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='protein_synonym', index=20, number=22,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='cds_span', index=21, number=23,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=16799,
  serialized_end=17216,
)
_sym_db.RegisterEnumDescriptor(_VIRUSTABLEFIELD)

VirusTableField = enum_type_wrapper.EnumTypeWrapper(_VIRUSTABLEFIELD)
_TABLEFORMAT = _descriptor.EnumDescriptor(
  name='TableFormat',
  full_name='ncbi.datasets.v1alpha1.TableFormat',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='tsv', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='csv', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='jsonl', index=2, number=2,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=17218,
  serialized_end=17260,
)
_sym_db.RegisterEnumDescriptor(_TABLEFORMAT)

TableFormat = enum_type_wrapper.EnumTypeWrapper(_TABLEFORMAT)
SORT_DIRECTION_UNSPECIFIED = 0
SORT_DIRECTION_ASCENDING = 1
SORT_DIRECTION_DESCENDING = 2
FASTA_UNSPECIFIED = 0
FASTA_GENE = 1
FASTA_RNA = 2
FASTA_PROTEIN = 3
FASTA_GENE_FLANK = 4
none = 0
plus = 1
minus = 2
unspecified = 0
nucleotide_accession = 1
species_tax_id = 2
species_name = 3
genus = 4
family = 5
nucleotide_length = 7
isolate_name = 8
sequence_type = 9
nuc_completeness = 10
geo_location = 11
us_state = 12
host_name = 13
host_tax_id = 14
collection_date = 15
bioproject = 16
biosample = 17
polyprotein_name = 19
protein_name = 20
protein_accession = 21
protein_synonym = 22
cds_span = 23
tsv = 0
csv = 1
jsonl = 2


_ASSEMBLYMETADATAREQUEST_CONTENTTYPE = _descriptor.EnumDescriptor(
  name='ContentType',
  full_name='ncbi.datasets.v1alpha1.AssemblyMetadataRequest.ContentType',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='COMPLETE', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='ASSM_ACC', index=1, number=1,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=890,
  serialized_end=931,
)
_sym_db.RegisterEnumDescriptor(_ASSEMBLYMETADATAREQUEST_CONTENTTYPE)

_ASSEMBLYDATASETDESCRIPTORSREQUEST_CONTENTTYPE = _descriptor.EnumDescriptor(
  name='ContentType',
  full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsRequest.ContentType',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='COMPLETE', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='ASSM_ACC', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='COUNTS', index=2, number=2,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=1416,
  serialized_end=1469,
)
_sym_db.RegisterEnumDescriptor(_ASSEMBLYDATASETDESCRIPTORSREQUEST_CONTENTTYPE)

_ANNOTATIONFORASSEMBLY_TYPE = _descriptor.EnumDescriptor(
  name='Type',
  full_name='ncbi.datasets.v1alpha1.AnnotationForAssembly.Type',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='DEFAULT', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='GENOME_GFF', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='GENOME_GBFF', index=2, number=2,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='GENOME_GB', index=3, number=2,
      serialized_options=b'\010\001',
      type=None),
    _descriptor.EnumValueDescriptor(
      name='RNA_FASTA', index=4, number=3,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='PROT_FASTA', index=5, number=4,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='GENOME_GTF', index=6, number=5,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=b'\020\001',
  serialized_start=2855,
  serialized_end=2977,
)
_sym_db.RegisterEnumDescriptor(_ANNOTATIONFORASSEMBLY_TYPE)

_ASSEMBLYDATASETREQUEST_RESOLUTION = _descriptor.EnumDescriptor(
  name='Resolution',
  full_name='ncbi.datasets.v1alpha1.AssemblyDatasetRequest.Resolution',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='FULLY_HYDRATED', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='DATA_REPORT_ONLY', index=1, number=1,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=3447,
  serialized_end=3501,
)
_sym_db.RegisterEnumDescriptor(_ASSEMBLYDATASETREQUEST_RESOLUTION)

_ORGANISM_RANKTYPE = _descriptor.EnumDescriptor(
  name='RankType',
  full_name='ncbi.datasets.v1alpha1.Organism.RankType',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='NO_RANK', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SUPERKINGDOM', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='KINGDOM', index=2, number=2,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SUBKINGDOM', index=3, number=3,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SUPERPHYLUM', index=4, number=4,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SUBPHYLUM', index=5, number=5,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='PHYLUM', index=6, number=6,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='CLADE', index=7, number=31,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SUPERCLASS', index=8, number=7,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='CLASS', index=9, number=8,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SUBCLASS', index=10, number=9,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='INFRACLASS', index=11, number=10,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='COHORT', index=12, number=11,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SUBCOHORT', index=13, number=12,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SUPERORDER', index=14, number=13,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='ORDER', index=15, number=14,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SUBORDER', index=16, number=15,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='INFRAORDER', index=17, number=16,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='PARVORDER', index=18, number=17,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SUPERFAMILY', index=19, number=18,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='FAMILY', index=20, number=19,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SUBFAMILY', index=21, number=20,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='GENUS', index=22, number=21,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SUBGENUS', index=23, number=22,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SPECIES_GROUP', index=24, number=23,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SPECIES_SUBGROUP', index=25, number=24,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SPECIES', index=26, number=25,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SUBSPECIES', index=27, number=26,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='TRIBE', index=28, number=27,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SUBTRIBE', index=29, number=28,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='FORMA', index=30, number=29,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='VARIETAS', index=31, number=30,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='STRAIN', index=32, number=320,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SECTION', index=33, number=330,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SUBSECTION', index=34, number=340,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='PATHOGROUP', index=35, number=350,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SUBVARIETY', index=36, number=360,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='GENOTYPE', index=37, number=370,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SEROTYPE', index=38, number=380,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='ISOLATE', index=39, number=390,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='MORPH', index=40, number=400,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SERIES', index=41, number=410,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='FORMA_SPECIALIS', index=42, number=420,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SEROGROUP', index=43, number=430,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='BIOTYPE', index=44, number=440,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=4905,
  serialized_end=5574,
)
_sym_db.RegisterEnumDescriptor(_ORGANISM_RANKTYPE)

_ORGANISM_COUNTTYPE = _descriptor.EnumDescriptor(
  name='CountType',
  full_name='ncbi.datasets.v1alpha1.Organism.CountType',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='COUNT_TYPE_UNSPECIFIED', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='COUNT_TYPE_ASSEMBLY', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='COUNT_TYPE_GENE', index=2, number=2,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=5576,
  serialized_end=5661,
)
_sym_db.RegisterEnumDescriptor(_ORGANISM_COUNTTYPE)

_ORTHOLOGREQUEST_CONTENTTYPE = _descriptor.EnumDescriptor(
  name='ContentType',
  full_name='ncbi.datasets.v1alpha1.OrthologRequest.ContentType',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='COMPLETE', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='IDS_ONLY', index=1, number=1,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=6244,
  serialized_end=6285,
)
_sym_db.RegisterEnumDescriptor(_ORTHOLOGREQUEST_CONTENTTYPE)

_GENEDATASETREQUEST_CONTENTTYPE = _descriptor.EnumDescriptor(
  name='ContentType',
  full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.ContentType',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='COMPLETE', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='IDS_ONLY', index=1, number=1,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=6244,
  serialized_end=6285,
)
_sym_db.RegisterEnumDescriptor(_GENEDATASETREQUEST_CONTENTTYPE)

_GENEDATASETREQUEST_SORTFIELD = _descriptor.EnumDescriptor(
  name='SortField',
  full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.SortField',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='SORT_FIELD_UNSPECIFIED', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SORT_FIELD_GENE_ID', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SORT_FIELD_GENE_TYPE', index=2, number=2,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='SORT_FIELD_GENE_SYMBOL', index=3, number=3,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=7308,
  serialized_end=7425,
)
_sym_db.RegisterEnumDescriptor(_GENEDATASETREQUEST_SORTFIELD)

_TRANSCRIPT_TRANSCRIPTTYPE = _descriptor.EnumDescriptor(
  name='TranscriptType',
  full_name='ncbi.datasets.v1alpha1.Transcript.TranscriptType',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='UNKNOWN', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='PROTEIN_CODING', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='NON_CODING', index=2, number=2,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='PROTEIN_CODING_MODEL', index=3, number=3,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='NON_CODING_MODEL', index=4, number=4,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=8894,
  serialized_end=9007,
)
_sym_db.RegisterEnumDescriptor(_TRANSCRIPT_TRANSCRIPTTYPE)

_GENOMICREGION_GENOMICREGIONTYPE = _descriptor.EnumDescriptor(
  name='GenomicRegionType',
  full_name='ncbi.datasets.v1alpha1.GenomicRegion.GenomicRegionType',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='UNKNOWN', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='REFSEQ_GENE', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='PSEUDOGENE', index=2, number=2,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='BIOLOGICAL_REGION', index=3, number=3,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='OTHER', index=4, number=4,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=9172,
  serialized_end=9271,
)
_sym_db.RegisterEnumDescriptor(_GENOMICREGION_GENOMICREGIONTYPE)

_GENEDESCRIPTOR_GENETYPE = _descriptor.EnumDescriptor(
  name='GeneType',
  full_name='ncbi.datasets.v1alpha1.GeneDescriptor.GeneType',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='UNKNOWN', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='tRNA', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='rRNA', index=2, number=2,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='snRNA', index=3, number=3,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='scRNA', index=4, number=4,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='snoRNA', index=5, number=5,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='PROTEIN_CODING', index=6, number=6,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='PSEUDO', index=7, number=7,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='TRANSPOSON', index=8, number=8,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='miscRNA', index=9, number=9,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='ncRNA', index=10, number=10,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='BIOLOGICAL_REGION', index=11, number=11,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='OTHER', index=12, number=255,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=10732,
  serialized_end=10916,
)
_sym_db.RegisterEnumDescriptor(_GENEDESCRIPTOR_GENETYPE)

_GENEDESCRIPTOR_RNATYPE = _descriptor.EnumDescriptor(
  name='RnaType',
  full_name='ncbi.datasets.v1alpha1.GeneDescriptor.RnaType',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='rna_UNKNOWN', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='premsg', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='tmRna', index=2, number=2,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=10918,
  serialized_end=10967,
)
_sym_db.RegisterEnumDescriptor(_GENEDESCRIPTOR_RNATYPE)

_ERROR_ASSEMBLYERRORCODE = _descriptor.EnumDescriptor(
  name='AssemblyErrorCode',
  full_name='ncbi.datasets.v1alpha1.Error.AssemblyErrorCode',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='UNKNOWN_ASSEMBLY_ERROR_CODE', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='INVALID_BIOPROJECT_IDS', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='NO_ASSEMBLIES_FOR_BIOPROJECTS', index=2, number=2,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='INVALID_TAXON', index=3, number=3,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=11491,
  serialized_end=11625,
)
_sym_db.RegisterEnumDescriptor(_ERROR_ASSEMBLYERRORCODE)

_ERROR_GENEERRORCODE = _descriptor.EnumDescriptor(
  name='GeneErrorCode',
  full_name='ncbi.datasets.v1alpha1.Error.GeneErrorCode',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='UNKNOWN_GENE_ERROR_CODE', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='INCOMPLETE_LOOKUP_SYMBOL', index=1, number=5,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=11627,
  serialized_end=11701,
)
_sym_db.RegisterEnumDescriptor(_ERROR_GENEERRORCODE)

_ERROR_VIRUSERRORCODE = _descriptor.EnumDescriptor(
  name='VirusErrorCode',
  full_name='ncbi.datasets.v1alpha1.Error.VirusErrorCode',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='UNKNOWN_VIRUS_ERROR_CODE', index=0, number=0,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=11703,
  serialized_end=11749,
)
_sym_db.RegisterEnumDescriptor(_ERROR_VIRUSERRORCODE)

_WARNING_GENEWARNINGCODE = _descriptor.EnumDescriptor(
  name='GeneWarningCode',
  full_name='ncbi.datasets.v1alpha1.Warning.GeneWarningCode',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='UNKNOWN_GENE_WARNING_CODE', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='ACCESSION_VERSION_MISMATCH', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='REPLACED_GENE_ID', index=2, number=2,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='DISCONTINUED_GENE_ID', index=3, number=3,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='UNRECOGNIZED_GENE_ID', index=4, number=4,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='UNRECOGNIZED_GENE_SYMBOL', index=5, number=5,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='UNRECOGNIZED_ACCESSION', index=6, number=6,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='UNRECOGNIZED_TAX_TOKEN', index=7, number=7,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=12123,
  serialized_end=12363,
)
_sym_db.RegisterEnumDescriptor(_WARNING_GENEWARNINGCODE)

_ANNOTATIONFORVIRUS_TYPE = _descriptor.EnumDescriptor(
  name='Type',
  full_name='ncbi.datasets.v1alpha1.AnnotationForVirus.Type',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='DEFAULT', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='CDS_FASTA', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='PROT_FASTA', index=2, number=2,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='GENOME_GBFF', index=3, number=3,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='GENOME_GPFF', index=4, number=4,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='PDB_FILES', index=5, number=5,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=13460,
  serialized_end=13559,
)
_sym_db.RegisterEnumDescriptor(_ANNOTATIONFORVIRUS_TYPE)

_MICROBIGGEDATASETREQUEST_FILETYPE = _descriptor.EnumDescriptor(
  name='FileType',
  full_name='ncbi.datasets.v1alpha1.MicroBiggeDatasetRequest.FileType',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='FILE_TYPE_UNSPECIFIED', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='element_fasta', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='element_flank_fasta', index=2, number=2,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='contig_fasta', index=3, number=3,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='protein_fasta', index=4, number=4,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=15963,
  serialized_end=16081,
)
_sym_db.RegisterEnumDescriptor(_MICROBIGGEDATASETREQUEST_FILETYPE)


_METHODPAYLOADREQUEST = _descriptor.Descriptor(
  name='MethodPayloadRequest',
  full_name='ncbi.datasets.v1alpha1.MethodPayloadRequest',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='method', full_name='ncbi.datasets.v1alpha1.MethodPayloadRequest.method', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='method', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='payload', full_name='ncbi.datasets.v1alpha1.MethodPayloadRequest.payload', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='payload', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=144,
  serialized_end=216,
)


_ASSEMBLYDATASETDESCRIPTORSFILTER = _descriptor.Descriptor(
  name='AssemblyDatasetDescriptorsFilter',
  full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsFilter',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='reference_only', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsFilter.reference_only', index=0,
      number=1, type=8, cpp_type=7, label=1,
      has_default_value=False, default_value=False,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='referenceOnly', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='refseq_only', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsFilter.refseq_only', index=1,
      number=2, type=8, cpp_type=7, label=1,
      has_default_value=False, default_value=False,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='refseqOnly', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=218,
  serialized_end=324,
)


_ACCESSIONS = _descriptor.Descriptor(
  name='Accessions',
  full_name='ncbi.datasets.v1alpha1.Accessions',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='accessions', full_name='ncbi.datasets.v1alpha1.Accessions.accessions', index=0,
      number=1, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='accessions', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=326,
  serialized_end=370,
)


_ASSEMBLYMETADATAREQUEST_BIOPROJECTS = _descriptor.Descriptor(
  name='Bioprojects',
  full_name='ncbi.datasets.v1alpha1.AssemblyMetadataRequest.Bioprojects',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='accessions', full_name='ncbi.datasets.v1alpha1.AssemblyMetadataRequest.Bioprojects.accessions', index=0,
      number=1, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='accessions', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=843,
  serialized_end=888,
)

_ASSEMBLYMETADATAREQUEST = _descriptor.Descriptor(
  name='AssemblyMetadataRequest',
  full_name='ncbi.datasets.v1alpha1.AssemblyMetadataRequest',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='taxon', full_name='ncbi.datasets.v1alpha1.AssemblyMetadataRequest.taxon', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='taxon', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='accessions', full_name='ncbi.datasets.v1alpha1.AssemblyMetadataRequest.accessions', index=1,
      number=2, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='accessions', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='bioprojects', full_name='ncbi.datasets.v1alpha1.AssemblyMetadataRequest.bioprojects', index=2,
      number=7, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='bioprojects', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='limit', full_name='ncbi.datasets.v1alpha1.AssemblyMetadataRequest.limit', index=3,
      number=3, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='limit', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='filters', full_name='ncbi.datasets.v1alpha1.AssemblyMetadataRequest.filters', index=4,
      number=4, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='filters', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='tax_exact_match', full_name='ncbi.datasets.v1alpha1.AssemblyMetadataRequest.tax_exact_match', index=5,
      number=5, type=8, cpp_type=7, label=1,
      has_default_value=False, default_value=False,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='taxExactMatch', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='returned_content', full_name='ncbi.datasets.v1alpha1.AssemblyMetadataRequest.returned_content', index=6,
      number=6, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='returnedContent', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[_ASSEMBLYMETADATAREQUEST_BIOPROJECTS, ],
  enum_types=[
    _ASSEMBLYMETADATAREQUEST_CONTENTTYPE,
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
    _descriptor.OneofDescriptor(
      name='id', full_name='ncbi.datasets.v1alpha1.AssemblyMetadataRequest.id',
      index=0, containing_type=None, fields=[]),
    _descriptor.OneofDescriptor(
      name='records', full_name='ncbi.datasets.v1alpha1.AssemblyMetadataRequest.records',
      index=1, containing_type=None, fields=[]),
  ],
  serialized_start=373,
  serialized_end=948,
)


_ASSEMBLYDATASETDESCRIPTORSREQUEST = _descriptor.Descriptor(
  name='AssemblyDatasetDescriptorsRequest',
  full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsRequest',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='tax_id', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsRequest.tax_id', index=0,
      number=1, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='taxId', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='org_name', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsRequest.org_name', index=1,
      number=5, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='orgName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='tax_name', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsRequest.tax_name', index=2,
      number=7, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='taxName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='assembly_accession', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsRequest.assembly_accession', index=3,
      number=4, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='assemblyAccession', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='cutoff', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsRequest.cutoff', index=4,
      number=2, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='cutoff', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='limit', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsRequest.limit', index=5,
      number=8, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='limit', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='filters', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsRequest.filters', index=6,
      number=3, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='filters', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='tax_exact_match', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsRequest.tax_exact_match', index=7,
      number=6, type=8, cpp_type=7, label=1,
      has_default_value=False, default_value=False,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='taxExactMatch', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='returned_content', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsRequest.returned_content', index=8,
      number=9, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='returnedContent', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
    _ASSEMBLYDATASETDESCRIPTORSREQUEST_CONTENTTYPE,
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
    _descriptor.OneofDescriptor(
      name='id', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsRequest.id',
      index=0, containing_type=None, fields=[]),
    _descriptor.OneofDescriptor(
      name='records', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsRequest.records',
      index=1, containing_type=None, fields=[]),
  ],
  serialized_start=951,
  serialized_end=1486,
)


_ASSEMBLYDATASETDESCRIPTORS = _descriptor.Descriptor(
  name='AssemblyDatasetDescriptors',
  full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptors',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='datasets', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptors.datasets', index=0,
      number=1, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='datasets', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='total_count', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptors.total_count', index=1,
      number=2, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='totalCount', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=1489,
  serialized_end=1629,
)


_BIOPROJECT = _descriptor.Descriptor(
  name='BioProject',
  full_name='ncbi.datasets.v1alpha1.BioProject',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='accession', full_name='ncbi.datasets.v1alpha1.BioProject.accession', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='accession', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='title', full_name='ncbi.datasets.v1alpha1.BioProject.title', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='title', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='parent_accession', full_name='ncbi.datasets.v1alpha1.BioProject.parent_accession', index=2,
      number=3, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='parentAccession', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='parent_accessions', full_name='ncbi.datasets.v1alpha1.BioProject.parent_accessions', index=3,
      number=4, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='parentAccessions', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=1632,
  serialized_end=1784,
)


_BIOPROJECTLINEAGE = _descriptor.Descriptor(
  name='BioProjectLineage',
  full_name='ncbi.datasets.v1alpha1.BioProjectLineage',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='bioprojects', full_name='ncbi.datasets.v1alpha1.BioProjectLineage.bioprojects', index=0,
      number=1, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='bioprojects', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=1786,
  serialized_end=1875,
)


_ASSEMBLYDATASETDESCRIPTOR = _descriptor.Descriptor(
  name='AssemblyDatasetDescriptor',
  full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptor',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='assembly_accession', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptor.assembly_accession', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='assemblyAccession', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='bioproject_lineages', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptor.bioproject_lineages', index=1,
      number=14, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='bioprojectLineages', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='display_name', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptor.display_name', index=2,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='displayName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='org', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptor.org', index=3,
      number=10, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='org', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='chromosomes', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptor.chromosomes', index=4,
      number=4, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='chromosomes', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='assembly_category', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptor.assembly_category', index=5,
      number=5, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='assemblyCategory', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='annotation_metadata', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptor.annotation_metadata', index=6,
      number=11, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='annotationMetadata', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='assembly_level', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptor.assembly_level', index=7,
      number=7, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='assemblyLevel', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='submission_date', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptor.submission_date', index=8,
      number=8, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='submissionDate', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='contig_n50', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptor.contig_n50', index=9,
      number=9, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='contigN50', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='estimated_size', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptor.estimated_size', index=10,
      number=12, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='estimatedSize', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='seq_length', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetDescriptor.seq_length', index=11,
      number=13, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='seqLength', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=1878,
  serialized_end=2487,
)


_ANNOTATIONFORASSEMBLY_FILE = _descriptor.Descriptor(
  name='File',
  full_name='ncbi.datasets.v1alpha1.AnnotationForAssembly.File',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='type', full_name='ncbi.datasets.v1alpha1.AnnotationForAssembly.File.type', index=0,
      number=1, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='type', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='estimated_size', full_name='ncbi.datasets.v1alpha1.AnnotationForAssembly.File.estimated_size', index=1,
      number=2, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='estimatedSize', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=2736,
  serialized_end=2853,
)

_ANNOTATIONFORASSEMBLY = _descriptor.Descriptor(
  name='AnnotationForAssembly',
  full_name='ncbi.datasets.v1alpha1.AnnotationForAssembly',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='name', full_name='ncbi.datasets.v1alpha1.AnnotationForAssembly.name', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='name', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='source', full_name='ncbi.datasets.v1alpha1.AnnotationForAssembly.source', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='source', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='release_number', full_name='ncbi.datasets.v1alpha1.AnnotationForAssembly.release_number', index=2,
      number=3, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='releaseNumber', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='release_date', full_name='ncbi.datasets.v1alpha1.AnnotationForAssembly.release_date', index=3,
      number=4, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='releaseDate', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='report_url', full_name='ncbi.datasets.v1alpha1.AnnotationForAssembly.report_url', index=4,
      number=5, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='reportUrl', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='file', full_name='ncbi.datasets.v1alpha1.AnnotationForAssembly.file', index=5,
      number=6, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='file', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[_ANNOTATIONFORASSEMBLY_FILE, ],
  enum_types=[
    _ANNOTATIONFORASSEMBLY_TYPE,
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=2490,
  serialized_end=2977,
)


_ASSEMBLYDATASETREQUEST = _descriptor.Descriptor(
  name='AssemblyDatasetRequest',
  full_name='ncbi.datasets.v1alpha1.AssemblyDatasetRequest',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='assembly_accessions', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetRequest.assembly_accessions', index=0,
      number=1, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='assemblyAccessions', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='accessions', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetRequest.accessions', index=1,
      number=10, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='accessions', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='chromosomes', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetRequest.chromosomes', index=2,
      number=2, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='chromosomes', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='include_annotation', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetRequest.include_annotation', index=3,
      number=3, type=8, cpp_type=7, label=1,
      has_default_value=False, default_value=False,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='includeAnnotation', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='exclude_sequence', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetRequest.exclude_sequence', index=4,
      number=7, type=8, cpp_type=7, label=1,
      has_default_value=False, default_value=False,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='excludeSequence', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='include_annotation_type', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetRequest.include_annotation_type', index=5,
      number=8, type=14, cpp_type=8, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='includeAnnotationType', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='hydrated', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetRequest.hydrated', index=6,
      number=9, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='hydrated', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='include_tsv', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetRequest.include_tsv', index=7,
      number=11, type=8, cpp_type=7, label=1,
      has_default_value=False, default_value=False,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='includeTsv', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
    _ASSEMBLYDATASETREQUEST_RESOLUTION,
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=2980,
  serialized_end=3501,
)


_ASSEMBLYDATASETAVAILABILITY = _descriptor.Descriptor(
  name='AssemblyDatasetAvailability',
  full_name='ncbi.datasets.v1alpha1.AssemblyDatasetAvailability',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='valid_assemblies', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetAvailability.valid_assemblies', index=0,
      number=1, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='validAssemblies', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='invalid_assemblies', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetAvailability.invalid_assemblies', index=1,
      number=2, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='invalidAssemblies', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='reason', full_name='ncbi.datasets.v1alpha1.AssemblyDatasetAvailability.reason', index=2,
      number=3, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='reason', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=3504,
  serialized_end=3647,
)


_TAXIDREQUEST = _descriptor.Descriptor(
  name='TaxidRequest',
  full_name='ncbi.datasets.v1alpha1.TaxidRequest',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='org_name', full_name='ncbi.datasets.v1alpha1.TaxidRequest.org_name', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='orgName', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=3649,
  serialized_end=3690,
)


_TAXIDREPLY = _descriptor.Descriptor(
  name='TaxidReply',
  full_name='ncbi.datasets.v1alpha1.TaxidReply',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='tax_id', full_name='ncbi.datasets.v1alpha1.TaxidReply.tax_id', index=0,
      number=1, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='taxId', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=3692,
  serialized_end=3727,
)


_ORGANISM_COUNTS = _descriptor.Descriptor(
  name='Counts',
  full_name='ncbi.datasets.v1alpha1.Organism.Counts',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='node', full_name='ncbi.datasets.v1alpha1.Organism.Counts.node', index=0,
      number=1, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='node', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='subtree', full_name='ncbi.datasets.v1alpha1.Organism.Counts.subtree', index=1,
      number=2, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='subtree', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=4703,
  serialized_end=4757,
)

_ORGANISM_COUNTBYTYPE = _descriptor.Descriptor(
  name='CountByType',
  full_name='ncbi.datasets.v1alpha1.Organism.CountByType',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='type', full_name='ncbi.datasets.v1alpha1.Organism.CountByType.type', index=0,
      number=1, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='type', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='counts', full_name='ncbi.datasets.v1alpha1.Organism.CountByType.counts', index=1,
      number=2, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='counts', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=4760,
  serialized_end=4902,
)

_ORGANISM = _descriptor.Descriptor(
  name='Organism',
  full_name='ncbi.datasets.v1alpha1.Organism',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='tax_id', full_name='ncbi.datasets.v1alpha1.Organism.tax_id', index=0,
      number=1, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='taxId', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='sci_name', full_name='ncbi.datasets.v1alpha1.Organism.sci_name', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='sciName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='common_name', full_name='ncbi.datasets.v1alpha1.Organism.common_name', index=2,
      number=3, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='commonName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='blast_node', full_name='ncbi.datasets.v1alpha1.Organism.blast_node', index=3,
      number=16, type=8, cpp_type=7, label=1,
      has_default_value=False, default_value=False,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='blastNode', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='breed', full_name='ncbi.datasets.v1alpha1.Organism.breed', index=4,
      number=10, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='breed', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='cultivar', full_name='ncbi.datasets.v1alpha1.Organism.cultivar', index=5,
      number=11, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='cultivar', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='ecotype', full_name='ncbi.datasets.v1alpha1.Organism.ecotype', index=6,
      number=12, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='ecotype', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='isolate', full_name='ncbi.datasets.v1alpha1.Organism.isolate', index=7,
      number=13, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='isolate', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='sex', full_name='ncbi.datasets.v1alpha1.Organism.sex', index=8,
      number=14, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='sex', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='strain', full_name='ncbi.datasets.v1alpha1.Organism.strain', index=9,
      number=15, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='strain', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='search_text', full_name='ncbi.datasets.v1alpha1.Organism.search_text', index=10,
      number=17, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='searchText', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='rank', full_name='ncbi.datasets.v1alpha1.Organism.rank', index=11,
      number=5, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='rank', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='parent_tax_id', full_name='ncbi.datasets.v1alpha1.Organism.parent_tax_id', index=12,
      number=6, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='parentTaxId', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='assembly_count', full_name='ncbi.datasets.v1alpha1.Organism.assembly_count', index=13,
      number=7, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001\322\360\031\004\n\002\002\003', json_name='assemblyCount', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='assembly_counts', full_name='ncbi.datasets.v1alpha1.Organism.assembly_counts', index=14,
      number=19, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\322\360\031\004\n\002\002\003', json_name='assemblyCounts', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='counts', full_name='ncbi.datasets.v1alpha1.Organism.counts', index=15,
      number=22, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\322\360\031\004\n\002\002\003', json_name='counts', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='children', full_name='ncbi.datasets.v1alpha1.Organism.children', index=16,
      number=8, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\322\360\031\004\n\002\002\003', json_name='children', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='merged', full_name='ncbi.datasets.v1alpha1.Organism.merged', index=17,
      number=9, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\322\360\031\004\n\002\002\003', json_name='merged', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='merged_tax_ids', full_name='ncbi.datasets.v1alpha1.Organism.merged_tax_ids', index=18,
      number=18, type=4, cpp_type=4, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\322\360\031\004\n\002\002\003', json_name='mergedTaxIds', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='min_ord', full_name='ncbi.datasets.v1alpha1.Organism.min_ord', index=19,
      number=20, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\322\360\031\004\n\002\002\003', json_name='minOrd', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='max_ord', full_name='ncbi.datasets.v1alpha1.Organism.max_ord', index=20,
      number=21, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\322\360\031\004\n\002\002\003', json_name='maxOrd', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='key', full_name='ncbi.datasets.v1alpha1.Organism.key', index=21,
      number=100, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\322\360\031\004\n\002\002\003', json_name='key', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='title', full_name='ncbi.datasets.v1alpha1.Organism.title', index=22,
      number=101, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\322\360\031\004\n\002\002\003', json_name='title', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='icon', full_name='ncbi.datasets.v1alpha1.Organism.icon', index=23,
      number=102, type=8, cpp_type=7, label=1,
      has_default_value=False, default_value=False,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\322\360\031\004\n\002\002\003', json_name='icon', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[_ORGANISM_COUNTS, _ORGANISM_COUNTBYTYPE, ],
  enum_types=[
    _ORGANISM_RANKTYPE,
    _ORGANISM_COUNTTYPE,
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=3730,
  serialized_end=5661,
)


_TAXTREEREQUEST = _descriptor.Descriptor(
  name='TaxTreeRequest',
  full_name='ncbi.datasets.v1alpha1.TaxTreeRequest',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='tax_id', full_name='ncbi.datasets.v1alpha1.TaxTreeRequest.tax_id', index=0,
      number=1, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='taxId', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='tax_token', full_name='ncbi.datasets.v1alpha1.TaxTreeRequest.tax_token', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='taxToken', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='taxon', full_name='ncbi.datasets.v1alpha1.TaxTreeRequest.taxon', index=2,
      number=3, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='taxon', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
    _descriptor.OneofDescriptor(
      name='taxonomy', full_name='ncbi.datasets.v1alpha1.TaxTreeRequest.taxonomy',
      index=0, containing_type=None, fields=[]),
  ],
  serialized_start=5663,
  serialized_end=5779,
)


_ORGANISMQUERYREQUEST = _descriptor.Descriptor(
  name='OrganismQueryRequest',
  full_name='ncbi.datasets.v1alpha1.OrganismQueryRequest',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='organism_query', full_name='ncbi.datasets.v1alpha1.OrganismQueryRequest.organism_query', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='organismQuery', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='taxon_query', full_name='ncbi.datasets.v1alpha1.OrganismQueryRequest.taxon_query', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='taxonQuery', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
    _descriptor.OneofDescriptor(
      name='id', full_name='ncbi.datasets.v1alpha1.OrganismQueryRequest.id',
      index=0, containing_type=None, fields=[]),
  ],
  serialized_start=5781,
  serialized_end=5889,
)


_SCINAMEANDIDS_SCINAMEANDID = _descriptor.Descriptor(
  name='SciNameAndId',
  full_name='ncbi.datasets.v1alpha1.SciNameAndIds.SciNameAndId',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='sci_name', full_name='ncbi.datasets.v1alpha1.SciNameAndIds.SciNameAndId.sci_name', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='sciName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='tax_id', full_name='ncbi.datasets.v1alpha1.SciNameAndIds.SciNameAndId.tax_id', index=1,
      number=2, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='taxId', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=6002,
  serialized_end=6066,
)

_SCINAMEANDIDS = _descriptor.Descriptor(
  name='SciNameAndIds',
  full_name='ncbi.datasets.v1alpha1.SciNameAndIds',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='sci_name_and_ids', full_name='ncbi.datasets.v1alpha1.SciNameAndIds.sci_name_and_ids', index=0,
      number=1, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='sciNameAndIds', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[_SCINAMEANDIDS_SCINAMEANDID, ],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=5892,
  serialized_end=6066,
)


_ORTHOLOGREQUEST = _descriptor.Descriptor(
  name='OrthologRequest',
  full_name='ncbi.datasets.v1alpha1.OrthologRequest',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='gene_id', full_name='ncbi.datasets.v1alpha1.OrthologRequest.gene_id', index=0,
      number=1, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='geneId', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='returned_content', full_name='ncbi.datasets.v1alpha1.OrthologRequest.returned_content', index=1,
      number=2, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='returnedContent', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='taxon_filter', full_name='ncbi.datasets.v1alpha1.OrthologRequest.taxon_filter', index=2,
      number=3, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='taxonFilter', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
    _ORTHOLOGREQUEST_CONTENTTYPE,
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=6069,
  serialized_end=6285,
)


_ORTHOLOGSET = _descriptor.Descriptor(
  name='OrthologSet',
  full_name='ncbi.datasets.v1alpha1.OrthologSet',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='ortholog_set_id', full_name='ncbi.datasets.v1alpha1.OrthologSet.ortholog_set_id', index=0,
      number=1, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='orthologSetId', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='method', full_name='ncbi.datasets.v1alpha1.OrthologSet.method', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='method', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='genes', full_name='ncbi.datasets.v1alpha1.OrthologSet.genes', index=2,
      number=3, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='genes', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=6288,
  serialized_end=6425,
)


_GENEDATASETREQUEST_SYMBOLSFORTAXON = _descriptor.Descriptor(
  name='SymbolsForTaxon',
  full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.SymbolsForTaxon',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='symbols', full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.SymbolsForTaxon.symbols', index=0,
      number=1, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='symbols', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='taxon', full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.SymbolsForTaxon.taxon', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='taxon', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=6960,
  serialized_end=7025,
)

_GENEDATASETREQUEST_GENEIDS = _descriptor.Descriptor(
  name='GeneIds',
  full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.GeneIds',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='gene_ids', full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.GeneIds.gene_ids', index=0,
      number=1, type=13, cpp_type=3, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='geneIds', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=7027,
  serialized_end=7063,
)

_GENEDATASETREQUEST_ACCESSIONS = _descriptor.Descriptor(
  name='Accessions',
  full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.Accessions',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='accessions', full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.Accessions.accessions', index=0,
      number=1, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='accessions', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=326,
  serialized_end=370,
)

_GENEDATASETREQUEST_SORT = _descriptor.Descriptor(
  name='Sort',
  full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.Sort',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='field', full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.Sort.field', index=0,
      number=1, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='field', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='direction', full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.Sort.direction', index=1,
      number=2, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='direction', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=7112,
  serialized_end=7263,
)

_GENEDATASETREQUEST = _descriptor.Descriptor(
  name='GeneDatasetRequest',
  full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='gene_ids', full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.gene_ids', index=0,
      number=1, type=13, cpp_type=3, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='geneIds', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='accessions', full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.accessions', index=1,
      number=10, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='accessions', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='symbols_for_taxon', full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.symbols_for_taxon', index=2,
      number=9, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='symbolsForTaxon', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='taxon', full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.taxon', index=3,
      number=3, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='taxon', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='include_annotation_type', full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.include_annotation_type', index=4,
      number=6, type=14, cpp_type=8, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='includeAnnotationType', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='returned_content', full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.returned_content', index=5,
      number=7, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='returnedContent', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='sort_schema', full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.sort_schema', index=6,
      number=8, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='sortSchema', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='limit', full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.limit', index=7,
      number=11, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='limit', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='fasta_filter', full_name='ncbi.datasets.v1alpha1.GeneDatasetRequest.fasta_filter', index=8,
      number=12, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='fastaFilter', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[_GENEDATASETREQUEST_SYMBOLSFORTAXON, _GENEDATASETREQUEST_GENEIDS, _GENEDATASETREQUEST_ACCESSIONS, _GENEDATASETREQUEST_SORT, ],
  enum_types=[
    _GENEDATASETREQUEST_CONTENTTYPE,
    _GENEDATASETREQUEST_SORTFIELD,
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=6428,
  serialized_end=7425,
)


_RANGE = _descriptor.Descriptor(
  name='Range',
  full_name='ncbi.datasets.v1alpha1.Range',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='begin', full_name='ncbi.datasets.v1alpha1.Range.begin', index=0,
      number=1, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='begin', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='end', full_name='ncbi.datasets.v1alpha1.Range.end', index=1,
      number=2, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='end', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='orientation', full_name='ncbi.datasets.v1alpha1.Range.orientation', index=2,
      number=3, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='orientation', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='order', full_name='ncbi.datasets.v1alpha1.Range.order', index=3,
      number=4, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='order', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='ribosomal_slippage', full_name='ncbi.datasets.v1alpha1.Range.ribosomal_slippage', index=4,
      number=5, type=5, cpp_type=1, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='ribosomalSlippage', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=7428,
  serialized_end=7615,
)


_SEQRANGESET = _descriptor.Descriptor(
  name='SeqRangeSet',
  full_name='ncbi.datasets.v1alpha1.SeqRangeSet',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='accession_version', full_name='ncbi.datasets.v1alpha1.SeqRangeSet.accession_version', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='accessionVersion', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='range', full_name='ncbi.datasets.v1alpha1.SeqRangeSet.range', index=1,
      number=2, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='range', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=7617,
  serialized_end=7728,
)


_GENOMICLOCATION = _descriptor.Descriptor(
  name='GenomicLocation',
  full_name='ncbi.datasets.v1alpha1.GenomicLocation',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='genomic_accession_version', full_name='ncbi.datasets.v1alpha1.GenomicLocation.genomic_accession_version', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='genomicAccessionVersion', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='sequence_name', full_name='ncbi.datasets.v1alpha1.GenomicLocation.sequence_name', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='sequenceName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='genomic_range', full_name='ncbi.datasets.v1alpha1.GenomicLocation.genomic_range', index=2,
      number=3, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='genomicRange', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='exons', full_name='ncbi.datasets.v1alpha1.GenomicLocation.exons', index=3,
      number=4, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='exons', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=7731,
  serialized_end=7966,
)


_MATUREPEPTIDE = _descriptor.Descriptor(
  name='MaturePeptide',
  full_name='ncbi.datasets.v1alpha1.MaturePeptide',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='accession_version', full_name='ncbi.datasets.v1alpha1.MaturePeptide.accession_version', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='accessionVersion', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='name', full_name='ncbi.datasets.v1alpha1.MaturePeptide.name', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='name', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='length', full_name='ncbi.datasets.v1alpha1.MaturePeptide.length', index=2,
      number=3, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='length', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=7968,
  serialized_end=8072,
)


_PROTEIN = _descriptor.Descriptor(
  name='Protein',
  full_name='ncbi.datasets.v1alpha1.Protein',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='accession_version', full_name='ncbi.datasets.v1alpha1.Protein.accession_version', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='accessionVersion', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='name', full_name='ncbi.datasets.v1alpha1.Protein.name', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='name', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='length', full_name='ncbi.datasets.v1alpha1.Protein.length', index=2,
      number=3, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='length', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='isoform_name', full_name='ncbi.datasets.v1alpha1.Protein.isoform_name', index=3,
      number=5, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='isoformName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='ensembl_protein', full_name='ncbi.datasets.v1alpha1.Protein.ensembl_protein', index=4,
      number=6, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='ensemblProtein', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='mature_peptides', full_name='ncbi.datasets.v1alpha1.Protein.mature_peptides', index=5,
      number=4, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='maturePeptides', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=8075,
  serialized_end=8329,
)


_TRANSCRIPT = _descriptor.Descriptor(
  name='Transcript',
  full_name='ncbi.datasets.v1alpha1.Transcript',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='accession_version', full_name='ncbi.datasets.v1alpha1.Transcript.accession_version', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='accessionVersion', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='name', full_name='ncbi.datasets.v1alpha1.Transcript.name', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='name', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='length', full_name='ncbi.datasets.v1alpha1.Transcript.length', index=2,
      number=3, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='length', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='genomic_range', full_name='ncbi.datasets.v1alpha1.Transcript.genomic_range', index=3,
      number=4, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='genomicRange', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='exons', full_name='ncbi.datasets.v1alpha1.Transcript.exons', index=4,
      number=5, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='exons', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='cds', full_name='ncbi.datasets.v1alpha1.Transcript.cds', index=5,
      number=6, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='cds', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='genomic_locations', full_name='ncbi.datasets.v1alpha1.Transcript.genomic_locations', index=6,
      number=10, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='genomicLocations', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='ensembl_transcript', full_name='ncbi.datasets.v1alpha1.Transcript.ensembl_transcript', index=7,
      number=8, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='ensemblTranscript', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='protein', full_name='ncbi.datasets.v1alpha1.Transcript.protein', index=8,
      number=7, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='protein', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='type', full_name='ncbi.datasets.v1alpha1.Transcript.type', index=9,
      number=9, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='type', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
    _TRANSCRIPT_TRANSCRIPTTYPE,
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=8332,
  serialized_end=9007,
)


_GENOMICREGION = _descriptor.Descriptor(
  name='GenomicRegion',
  full_name='ncbi.datasets.v1alpha1.GenomicRegion',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='gene_range', full_name='ncbi.datasets.v1alpha1.GenomicRegion.gene_range', index=0,
      number=1, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='geneRange', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='type', full_name='ncbi.datasets.v1alpha1.GenomicRegion.type', index=1,
      number=2, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='type', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
    _GENOMICREGION_GENOMICREGIONTYPE,
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=9010,
  serialized_end=9271,
)


_ANNOTATEDASSEMBLIES = _descriptor.Descriptor(
  name='AnnotatedAssemblies',
  full_name='ncbi.datasets.v1alpha1.AnnotatedAssemblies',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='accession', full_name='ncbi.datasets.v1alpha1.AnnotatedAssemblies.accession', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='accession', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='name', full_name='ncbi.datasets.v1alpha1.AnnotatedAssemblies.name', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='name', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=9273,
  serialized_end=9344,
)


_ANNOTATION = _descriptor.Descriptor(
  name='Annotation',
  full_name='ncbi.datasets.v1alpha1.Annotation',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='release_name', full_name='ncbi.datasets.v1alpha1.Annotation.release_name', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='releaseName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='release_date', full_name='ncbi.datasets.v1alpha1.Annotation.release_date', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='releaseDate', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='assemblies_in_scope', full_name='ncbi.datasets.v1alpha1.Annotation.assemblies_in_scope', index=2,
      number=3, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='assembliesInScope', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=9347,
  serialized_end=9522,
)


_GENEDESCRIPTOR = _descriptor.Descriptor(
  name='GeneDescriptor',
  full_name='ncbi.datasets.v1alpha1.GeneDescriptor',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='gene_id', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.gene_id', index=0,
      number=1, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='geneId', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='symbol', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.symbol', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='symbol', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='description', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.description', index=2,
      number=3, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='description', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='tax_id', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.tax_id', index=3,
      number=4, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='taxId', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='taxname', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.taxname', index=4,
      number=5, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='taxname', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='type', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.type', index=5,
      number=6, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='type', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='rna_type', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.rna_type', index=6,
      number=11, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='rnaType', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='orientation', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.orientation', index=7,
      number=7, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='orientation', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='genomic_ranges', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.genomic_ranges', index=8,
      number=8, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='genomicRanges', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='reference_standards', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.reference_standards', index=9,
      number=21, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='referenceStandards', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='genomic_regions', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.genomic_regions', index=10,
      number=22, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='genomicRegions', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='transcripts', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.transcripts', index=11,
      number=9, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='transcripts', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='proteins', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.proteins', index=12,
      number=10, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='proteins', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='common_name', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.common_name', index=13,
      number=12, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='commonName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='chromosome', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.chromosome', index=14,
      number=13, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='chromosome', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='chromosomes', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.chromosomes', index=15,
      number=14, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='chromosomes', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='nomenclature_authority', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.nomenclature_authority', index=16,
      number=15, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='nomenclatureAuthority', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='swiss_prot_accessions', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.swiss_prot_accessions', index=17,
      number=16, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='swissProtAccessions', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='ensembl_gene_ids', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.ensembl_gene_ids', index=18,
      number=17, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='ensemblGeneIds', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='omim_ids', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.omim_ids', index=19,
      number=18, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='omimIds', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='synonyms', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.synonyms', index=20,
      number=19, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='synonyms', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='replaced_gene_id', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.replaced_gene_id', index=21,
      number=20, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='replacedGeneId', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='annotations', full_name='ncbi.datasets.v1alpha1.GeneDescriptor.annotations', index=22,
      number=23, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='annotations', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
    _GENEDESCRIPTOR_GENETYPE,
    _GENEDESCRIPTOR_RNATYPE,
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=9525,
  serialized_end=10967,
)


_NOMENCLATUREAUTHORITY = _descriptor.Descriptor(
  name='NomenclatureAuthority',
  full_name='ncbi.datasets.v1alpha1.NomenclatureAuthority',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='authority', full_name='ncbi.datasets.v1alpha1.NomenclatureAuthority.authority', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='authority', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='identifier', full_name='ncbi.datasets.v1alpha1.NomenclatureAuthority.identifier', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='identifier', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=10969,
  serialized_end=11054,
)


_ERROR = _descriptor.Descriptor(
  name='Error',
  full_name='ncbi.datasets.v1alpha1.Error',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='assembly_error_code', full_name='ncbi.datasets.v1alpha1.Error.assembly_error_code', index=0,
      number=6, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='assemblyErrorCode', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='gene_error_code', full_name='ncbi.datasets.v1alpha1.Error.gene_error_code', index=1,
      number=7, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='geneErrorCode', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='virus_error_code', full_name='ncbi.datasets.v1alpha1.Error.virus_error_code', index=2,
      number=8, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='virusErrorCode', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='reason', full_name='ncbi.datasets.v1alpha1.Error.reason', index=3,
      number=5, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='reason', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='message', full_name='ncbi.datasets.v1alpha1.Error.message', index=4,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='message', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='valid_identifiers', full_name='ncbi.datasets.v1alpha1.Error.valid_identifiers', index=5,
      number=3, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='validIdentifiers', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='invalid_identifiers', full_name='ncbi.datasets.v1alpha1.Error.invalid_identifiers', index=6,
      number=4, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='invalidIdentifiers', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
    _ERROR_ASSEMBLYERRORCODE,
    _ERROR_GENEERRORCODE,
    _ERROR_VIRUSERRORCODE,
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
    _descriptor.OneofDescriptor(
      name='code', full_name='ncbi.datasets.v1alpha1.Error.code',
      index=0, containing_type=None, fields=[]),
  ],
  serialized_start=11057,
  serialized_end=11757,
)


_WARNING_REPLACEDID = _descriptor.Descriptor(
  name='ReplacedId',
  full_name='ncbi.datasets.v1alpha1.Warning.ReplacedId',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='requested', full_name='ncbi.datasets.v1alpha1.Warning.ReplacedId.requested', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='requested', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='returned', full_name='ncbi.datasets.v1alpha1.Warning.ReplacedId.returned', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='returned', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=12050,
  serialized_end=12120,
)

_WARNING = _descriptor.Descriptor(
  name='Warning',
  full_name='ncbi.datasets.v1alpha1.Warning',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='gene_warning_code', full_name='ncbi.datasets.v1alpha1.Warning.gene_warning_code', index=0,
      number=7, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='geneWarningCode', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='reason', full_name='ncbi.datasets.v1alpha1.Warning.reason', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='reason', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='message', full_name='ncbi.datasets.v1alpha1.Warning.message', index=2,
      number=3, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='message', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='replaced_id', full_name='ncbi.datasets.v1alpha1.Warning.replaced_id', index=3,
      number=4, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='replacedId', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='unrecognized_identifier', full_name='ncbi.datasets.v1alpha1.Warning.unrecognized_identifier', index=4,
      number=5, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='unrecognizedIdentifier', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[_WARNING_REPLACEDID, ],
  enum_types=[
    _WARNING_GENEWARNINGCODE,
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
    _descriptor.OneofDescriptor(
      name='code', full_name='ncbi.datasets.v1alpha1.Warning.code',
      index=0, containing_type=None, fields=[]),
  ],
  serialized_start=11760,
  serialized_end=12371,
)


_MESSAGE = _descriptor.Descriptor(
  name='Message',
  full_name='ncbi.datasets.v1alpha1.Message',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='error', full_name='ncbi.datasets.v1alpha1.Message.error', index=0,
      number=1, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='error', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='warning', full_name='ncbi.datasets.v1alpha1.Message.warning', index=1,
      number=2, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='warning', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
    _descriptor.OneofDescriptor(
      name='message', full_name='ncbi.datasets.v1alpha1.Message.message',
      index=0, containing_type=None, fields=[]),
  ],
  serialized_start=12374,
  serialized_end=12510,
)


_GENEMATCH = _descriptor.Descriptor(
  name='GeneMatch',
  full_name='ncbi.datasets.v1alpha1.GeneMatch',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='messages', full_name='ncbi.datasets.v1alpha1.GeneMatch.messages', index=0,
      number=1, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='messages', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='warnings', full_name='ncbi.datasets.v1alpha1.GeneMatch.warnings', index=1,
      number=3, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='warnings', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='query', full_name='ncbi.datasets.v1alpha1.GeneMatch.query', index=2,
      number=4, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='query', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='gene', full_name='ncbi.datasets.v1alpha1.GeneMatch.gene', index=3,
      number=2, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='gene', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='errors', full_name='ncbi.datasets.v1alpha1.GeneMatch.errors', index=4,
      number=5, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='errors', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=12513,
  serialized_end=12787,
)


_GENEMETADATA = _descriptor.Descriptor(
  name='GeneMetadata',
  full_name='ncbi.datasets.v1alpha1.GeneMetadata',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='messages', full_name='ncbi.datasets.v1alpha1.GeneMetadata.messages', index=0,
      number=1, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='messages', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='genes', full_name='ncbi.datasets.v1alpha1.GeneMetadata.genes', index=1,
      number=2, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='genes', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='total_count', full_name='ncbi.datasets.v1alpha1.GeneMetadata.total_count', index=2,
      number=3, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='totalCount', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=12790,
  serialized_end=12955,
)


_ASSEMBLYMATCH = _descriptor.Descriptor(
  name='AssemblyMatch',
  full_name='ncbi.datasets.v1alpha1.AssemblyMatch',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='messages', full_name='ncbi.datasets.v1alpha1.AssemblyMatch.messages', index=0,
      number=1, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='messages', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='assembly', full_name='ncbi.datasets.v1alpha1.AssemblyMatch.assembly', index=1,
      number=2, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='assembly', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=12958,
  serialized_end=13113,
)


_ASSEMBLYMETADATA = _descriptor.Descriptor(
  name='AssemblyMetadata',
  full_name='ncbi.datasets.v1alpha1.AssemblyMetadata',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='messages', full_name='ncbi.datasets.v1alpha1.AssemblyMetadata.messages', index=0,
      number=1, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='messages', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='assemblies', full_name='ncbi.datasets.v1alpha1.AssemblyMetadata.assemblies', index=1,
      number=2, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='assemblies', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='total_count', full_name='ncbi.datasets.v1alpha1.AssemblyMetadata.total_count', index=2,
      number=3, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='totalCount', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=13116,
  serialized_end=13299,
)


_GENEDESCRIPTORS = _descriptor.Descriptor(
  name='GeneDescriptors',
  full_name='ncbi.datasets.v1alpha1.GeneDescriptors',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='genes', full_name='ncbi.datasets.v1alpha1.GeneDescriptors.genes', index=0,
      number=1, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='genes', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='errors', full_name='ncbi.datasets.v1alpha1.GeneDescriptors.errors', index=1,
      number=2, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='errors', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=13302,
  serialized_end=13436,
)


_ANNOTATIONFORVIRUS = _descriptor.Descriptor(
  name='AnnotationForVirus',
  full_name='ncbi.datasets.v1alpha1.AnnotationForVirus',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
    _ANNOTATIONFORVIRUS_TYPE,
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=13438,
  serialized_end=13559,
)


_VIRUSDATASETREQUEST = _descriptor.Descriptor(
  name='VirusDatasetRequest',
  full_name='ncbi.datasets.v1alpha1.VirusDatasetRequest',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='tax_id', full_name='ncbi.datasets.v1alpha1.VirusDatasetRequest.tax_id', index=0,
      number=1, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='taxId', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='tax_name', full_name='ncbi.datasets.v1alpha1.VirusDatasetRequest.tax_name', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='taxName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='accession', full_name='ncbi.datasets.v1alpha1.VirusDatasetRequest.accession', index=2,
      number=3, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='accession', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='taxon', full_name='ncbi.datasets.v1alpha1.VirusDatasetRequest.taxon', index=3,
      number=11, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='taxon', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='refseq_only', full_name='ncbi.datasets.v1alpha1.VirusDatasetRequest.refseq_only', index=4,
      number=4, type=8, cpp_type=7, label=1,
      has_default_value=False, default_value=False,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='refseqOnly', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='annotated_only', full_name='ncbi.datasets.v1alpha1.VirusDatasetRequest.annotated_only', index=5,
      number=5, type=8, cpp_type=7, label=1,
      has_default_value=False, default_value=False,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='annotatedOnly', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='released_since', full_name='ncbi.datasets.v1alpha1.VirusDatasetRequest.released_since', index=6,
      number=6, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='releasedSince', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='host', full_name='ncbi.datasets.v1alpha1.VirusDatasetRequest.host', index=7,
      number=7, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='host', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='geo_location', full_name='ncbi.datasets.v1alpha1.VirusDatasetRequest.geo_location', index=8,
      number=8, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='geoLocation', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='complete_only', full_name='ncbi.datasets.v1alpha1.VirusDatasetRequest.complete_only', index=9,
      number=9, type=8, cpp_type=7, label=1,
      has_default_value=False, default_value=False,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='completeOnly', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='table_fields', full_name='ncbi.datasets.v1alpha1.VirusDatasetRequest.table_fields', index=10,
      number=10, type=14, cpp_type=8, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='tableFields', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='exclude_sequence', full_name='ncbi.datasets.v1alpha1.VirusDatasetRequest.exclude_sequence', index=11,
      number=12, type=8, cpp_type=7, label=1,
      has_default_value=False, default_value=False,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='excludeSequence', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='include_annotation_type', full_name='ncbi.datasets.v1alpha1.VirusDatasetRequest.include_annotation_type', index=12,
      number=13, type=14, cpp_type=8, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='includeAnnotationType', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='format', full_name='ncbi.datasets.v1alpha1.VirusDatasetRequest.format', index=13,
      number=14, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='format', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
    _descriptor.OneofDescriptor(
      name='id', full_name='ncbi.datasets.v1alpha1.VirusDatasetRequest.id',
      index=0, containing_type=None, fields=[]),
  ],
  serialized_start=13562,
  serialized_end=14223,
)


_SARS2PROTEINDATASETREQUEST = _descriptor.Descriptor(
  name='Sars2ProteinDatasetRequest',
  full_name='ncbi.datasets.v1alpha1.Sars2ProteinDatasetRequest',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='proteins', full_name='ncbi.datasets.v1alpha1.Sars2ProteinDatasetRequest.proteins', index=0,
      number=1, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='proteins', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='refseq_only', full_name='ncbi.datasets.v1alpha1.Sars2ProteinDatasetRequest.refseq_only', index=1,
      number=4, type=8, cpp_type=7, label=1,
      has_default_value=False, default_value=False,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='refseqOnly', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='annotated_only', full_name='ncbi.datasets.v1alpha1.Sars2ProteinDatasetRequest.annotated_only', index=2,
      number=5, type=8, cpp_type=7, label=1,
      has_default_value=False, default_value=False,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='annotatedOnly', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='released_since', full_name='ncbi.datasets.v1alpha1.Sars2ProteinDatasetRequest.released_since', index=3,
      number=6, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='releasedSince', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='host', full_name='ncbi.datasets.v1alpha1.Sars2ProteinDatasetRequest.host', index=4,
      number=7, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='host', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='geo_location', full_name='ncbi.datasets.v1alpha1.Sars2ProteinDatasetRequest.geo_location', index=5,
      number=8, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='geoLocation', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='complete_only', full_name='ncbi.datasets.v1alpha1.Sars2ProteinDatasetRequest.complete_only', index=6,
      number=9, type=8, cpp_type=7, label=1,
      has_default_value=False, default_value=False,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='completeOnly', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='table_fields', full_name='ncbi.datasets.v1alpha1.Sars2ProteinDatasetRequest.table_fields', index=7,
      number=10, type=14, cpp_type=8, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='tableFields', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='exclude_sequence', full_name='ncbi.datasets.v1alpha1.Sars2ProteinDatasetRequest.exclude_sequence', index=8,
      number=12, type=8, cpp_type=7, label=1,
      has_default_value=False, default_value=False,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='excludeSequence', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='include_annotation_type', full_name='ncbi.datasets.v1alpha1.Sars2ProteinDatasetRequest.include_annotation_type', index=9,
      number=13, type=14, cpp_type=8, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='includeAnnotationType', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='format', full_name='ncbi.datasets.v1alpha1.Sars2ProteinDatasetRequest.format', index=10,
      number=14, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='format', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=14226,
  serialized_end=14798,
)


_DOWNLOADSUMMARY_HYDRATED = _descriptor.Descriptor(
  name='Hydrated',
  full_name='ncbi.datasets.v1alpha1.DownloadSummary.Hydrated',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='estimated_file_size_mb', full_name='ncbi.datasets.v1alpha1.DownloadSummary.Hydrated.estimated_file_size_mb', index=0,
      number=1, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='estimatedFileSizeMb', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='url', full_name='ncbi.datasets.v1alpha1.DownloadSummary.Hydrated.url', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='url', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='cli_download_command_line', full_name='ncbi.datasets.v1alpha1.DownloadSummary.Hydrated.cli_download_command_line', index=2,
      number=3, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='cliDownloadCommandLine', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=15257,
  serialized_end=15397,
)

_DOWNLOADSUMMARY_DEHYDRATED = _descriptor.Descriptor(
  name='Dehydrated',
  full_name='ncbi.datasets.v1alpha1.DownloadSummary.Dehydrated',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='estimated_file_size_mb', full_name='ncbi.datasets.v1alpha1.DownloadSummary.Dehydrated.estimated_file_size_mb', index=0,
      number=1, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='estimatedFileSizeMb', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='url', full_name='ncbi.datasets.v1alpha1.DownloadSummary.Dehydrated.url', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='url', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='cli_download_command_line', full_name='ncbi.datasets.v1alpha1.DownloadSummary.Dehydrated.cli_download_command_line', index=2,
      number=3, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='cliDownloadCommandLine', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='cli_rehydrate_command_line', full_name='ncbi.datasets.v1alpha1.DownloadSummary.Dehydrated.cli_rehydrate_command_line', index=3,
      number=4, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='cliRehydrateCommandLine', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=15400,
  serialized_end=15603,
)

_DOWNLOADSUMMARY = _descriptor.Descriptor(
  name='DownloadSummary',
  full_name='ncbi.datasets.v1alpha1.DownloadSummary',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='record_count', full_name='ncbi.datasets.v1alpha1.DownloadSummary.record_count', index=0,
      number=1, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='recordCount', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='assembly_count', full_name='ncbi.datasets.v1alpha1.DownloadSummary.assembly_count', index=1,
      number=2, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='assemblyCount', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='resource_updated_on', full_name='ncbi.datasets.v1alpha1.DownloadSummary.resource_updated_on', index=2,
      number=4, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='resourceUpdatedOn', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='hydrated', full_name='ncbi.datasets.v1alpha1.DownloadSummary.hydrated', index=3,
      number=5, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='hydrated', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='dehydrated', full_name='ncbi.datasets.v1alpha1.DownloadSummary.dehydrated', index=4,
      number=6, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='dehydrated', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='errors', full_name='ncbi.datasets.v1alpha1.DownloadSummary.errors', index=5,
      number=7, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='errors', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='messages', full_name='ncbi.datasets.v1alpha1.DownloadSummary.messages', index=6,
      number=8, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='messages', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[_DOWNLOADSUMMARY_HYDRATED, _DOWNLOADSUMMARY_DEHYDRATED, ],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=14801,
  serialized_end=15603,
)


_TABULAROUTPUT = _descriptor.Descriptor(
  name='TabularOutput',
  full_name='ncbi.datasets.v1alpha1.TabularOutput',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='data', full_name='ncbi.datasets.v1alpha1.TabularOutput.data', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='data', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=15605,
  serialized_end=15640,
)


_MICROBIGGEDATASETREQUEST_ELEMENTFLANKCONFIG = _descriptor.Descriptor(
  name='ElementFlankConfig',
  full_name='ncbi.datasets.v1alpha1.MicroBiggeDatasetRequest.ElementFlankConfig',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='length', full_name='ncbi.datasets.v1alpha1.MicroBiggeDatasetRequest.ElementFlankConfig.length', index=0,
      number=1, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='length', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=15917,
  serialized_end=15961,
)

_MICROBIGGEDATASETREQUEST = _descriptor.Descriptor(
  name='MicroBiggeDatasetRequest',
  full_name='ncbi.datasets.v1alpha1.MicroBiggeDatasetRequest',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='opaque_solr_query', full_name='ncbi.datasets.v1alpha1.MicroBiggeDatasetRequest.opaque_solr_query', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='opaqueSolrQuery', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='files', full_name='ncbi.datasets.v1alpha1.MicroBiggeDatasetRequest.files', index=1,
      number=2, type=14, cpp_type=8, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='files', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='element_flank_config', full_name='ncbi.datasets.v1alpha1.MicroBiggeDatasetRequest.element_flank_config', index=2,
      number=3, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='elementFlankConfig', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[_MICROBIGGEDATASETREQUEST_ELEMENTFLANKCONFIG, ],
  enum_types=[
    _MICROBIGGEDATASETREQUEST_FILETYPE,
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
    _descriptor.OneofDescriptor(
      name='query', full_name='ncbi.datasets.v1alpha1.MicroBiggeDatasetRequest.query',
      index=0, containing_type=None, fields=[]),
  ],
  serialized_start=15643,
  serialized_end=16090,
)


_VERSIONREPLY = _descriptor.Descriptor(
  name='VersionReply',
  full_name='ncbi.datasets.v1alpha1.VersionReply',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='version', full_name='ncbi.datasets.v1alpha1.VersionReply.version', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='version', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='major_ver', full_name='ncbi.datasets.v1alpha1.VersionReply.major_ver', index=1,
      number=2, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='majorVer', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='minor_ver', full_name='ncbi.datasets.v1alpha1.VersionReply.minor_ver', index=2,
      number=3, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='minorVer', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='patch_ver', full_name='ncbi.datasets.v1alpha1.VersionReply.patch_ver', index=3,
      number=4, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='patchVer', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=16092,
  serialized_end=16219,
)


_PROKARYOTEGENEREQUEST_GENEFLANKCONFIG = _descriptor.Descriptor(
  name='GeneFlankConfig',
  full_name='ncbi.datasets.v1alpha1.ProkaryoteGeneRequest.GeneFlankConfig',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='length', full_name='ncbi.datasets.v1alpha1.ProkaryoteGeneRequest.GeneFlankConfig.length', index=0,
      number=1, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='length', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=16495,
  serialized_end=16536,
)

_PROKARYOTEGENEREQUEST = _descriptor.Descriptor(
  name='ProkaryoteGeneRequest',
  full_name='ncbi.datasets.v1alpha1.ProkaryoteGeneRequest',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='accessions', full_name='ncbi.datasets.v1alpha1.ProkaryoteGeneRequest.accessions', index=0,
      number=1, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='accessions', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='include_annotation_type', full_name='ncbi.datasets.v1alpha1.ProkaryoteGeneRequest.include_annotation_type', index=1,
      number=2, type=14, cpp_type=8, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='includeAnnotationType', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='gene_flank_config', full_name='ncbi.datasets.v1alpha1.ProkaryoteGeneRequest.gene_flank_config', index=2,
      number=3, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='geneFlankConfig', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='taxon', full_name='ncbi.datasets.v1alpha1.ProkaryoteGeneRequest.taxon', index=3,
      number=4, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='taxon', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[_PROKARYOTEGENEREQUEST_GENEFLANKCONFIG, ],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=16222,
  serialized_end=16536,
)

_ASSEMBLYMETADATAREQUEST_BIOPROJECTS.containing_type = _ASSEMBLYMETADATAREQUEST
_ASSEMBLYMETADATAREQUEST.fields_by_name['accessions'].message_type = _ACCESSIONS
_ASSEMBLYMETADATAREQUEST.fields_by_name['bioprojects'].message_type = _ASSEMBLYMETADATAREQUEST_BIOPROJECTS
_ASSEMBLYMETADATAREQUEST.fields_by_name['filters'].message_type = _ASSEMBLYDATASETDESCRIPTORSFILTER
_ASSEMBLYMETADATAREQUEST.fields_by_name['returned_content'].enum_type = _ASSEMBLYMETADATAREQUEST_CONTENTTYPE
_ASSEMBLYMETADATAREQUEST_CONTENTTYPE.containing_type = _ASSEMBLYMETADATAREQUEST
_ASSEMBLYMETADATAREQUEST.oneofs_by_name['id'].fields.append(
  _ASSEMBLYMETADATAREQUEST.fields_by_name['taxon'])
_ASSEMBLYMETADATAREQUEST.fields_by_name['taxon'].containing_oneof = _ASSEMBLYMETADATAREQUEST.oneofs_by_name['id']
_ASSEMBLYMETADATAREQUEST.oneofs_by_name['id'].fields.append(
  _ASSEMBLYMETADATAREQUEST.fields_by_name['accessions'])
_ASSEMBLYMETADATAREQUEST.fields_by_name['accessions'].containing_oneof = _ASSEMBLYMETADATAREQUEST.oneofs_by_name['id']
_ASSEMBLYMETADATAREQUEST.oneofs_by_name['id'].fields.append(
  _ASSEMBLYMETADATAREQUEST.fields_by_name['bioprojects'])
_ASSEMBLYMETADATAREQUEST.fields_by_name['bioprojects'].containing_oneof = _ASSEMBLYMETADATAREQUEST.oneofs_by_name['id']
_ASSEMBLYMETADATAREQUEST.oneofs_by_name['records'].fields.append(
  _ASSEMBLYMETADATAREQUEST.fields_by_name['limit'])
_ASSEMBLYMETADATAREQUEST.fields_by_name['limit'].containing_oneof = _ASSEMBLYMETADATAREQUEST.oneofs_by_name['records']
_ASSEMBLYDATASETDESCRIPTORSREQUEST.fields_by_name['filters'].message_type = _ASSEMBLYDATASETDESCRIPTORSFILTER
_ASSEMBLYDATASETDESCRIPTORSREQUEST.fields_by_name['returned_content'].enum_type = _ASSEMBLYDATASETDESCRIPTORSREQUEST_CONTENTTYPE
_ASSEMBLYDATASETDESCRIPTORSREQUEST_CONTENTTYPE.containing_type = _ASSEMBLYDATASETDESCRIPTORSREQUEST
_ASSEMBLYDATASETDESCRIPTORSREQUEST.oneofs_by_name['id'].fields.append(
  _ASSEMBLYDATASETDESCRIPTORSREQUEST.fields_by_name['tax_id'])
_ASSEMBLYDATASETDESCRIPTORSREQUEST.fields_by_name['tax_id'].containing_oneof = _ASSEMBLYDATASETDESCRIPTORSREQUEST.oneofs_by_name['id']
_ASSEMBLYDATASETDESCRIPTORSREQUEST.oneofs_by_name['id'].fields.append(
  _ASSEMBLYDATASETDESCRIPTORSREQUEST.fields_by_name['org_name'])
_ASSEMBLYDATASETDESCRIPTORSREQUEST.fields_by_name['org_name'].containing_oneof = _ASSEMBLYDATASETDESCRIPTORSREQUEST.oneofs_by_name['id']
_ASSEMBLYDATASETDESCRIPTORSREQUEST.oneofs_by_name['id'].fields.append(
  _ASSEMBLYDATASETDESCRIPTORSREQUEST.fields_by_name['tax_name'])
_ASSEMBLYDATASETDESCRIPTORSREQUEST.fields_by_name['tax_name'].containing_oneof = _ASSEMBLYDATASETDESCRIPTORSREQUEST.oneofs_by_name['id']
_ASSEMBLYDATASETDESCRIPTORSREQUEST.oneofs_by_name['id'].fields.append(
  _ASSEMBLYDATASETDESCRIPTORSREQUEST.fields_by_name['assembly_accession'])
_ASSEMBLYDATASETDESCRIPTORSREQUEST.fields_by_name['assembly_accession'].containing_oneof = _ASSEMBLYDATASETDESCRIPTORSREQUEST.oneofs_by_name['id']
_ASSEMBLYDATASETDESCRIPTORSREQUEST.oneofs_by_name['records'].fields.append(
  _ASSEMBLYDATASETDESCRIPTORSREQUEST.fields_by_name['cutoff'])
_ASSEMBLYDATASETDESCRIPTORSREQUEST.fields_by_name['cutoff'].containing_oneof = _ASSEMBLYDATASETDESCRIPTORSREQUEST.oneofs_by_name['records']
_ASSEMBLYDATASETDESCRIPTORSREQUEST.oneofs_by_name['records'].fields.append(
  _ASSEMBLYDATASETDESCRIPTORSREQUEST.fields_by_name['limit'])
_ASSEMBLYDATASETDESCRIPTORSREQUEST.fields_by_name['limit'].containing_oneof = _ASSEMBLYDATASETDESCRIPTORSREQUEST.oneofs_by_name['records']
_ASSEMBLYDATASETDESCRIPTORS.fields_by_name['datasets'].message_type = _ASSEMBLYDATASETDESCRIPTOR
_BIOPROJECTLINEAGE.fields_by_name['bioprojects'].message_type = _BIOPROJECT
_ASSEMBLYDATASETDESCRIPTOR.fields_by_name['bioproject_lineages'].message_type = _BIOPROJECTLINEAGE
_ASSEMBLYDATASETDESCRIPTOR.fields_by_name['org'].message_type = _ORGANISM
_ASSEMBLYDATASETDESCRIPTOR.fields_by_name['annotation_metadata'].message_type = _ANNOTATIONFORASSEMBLY
_ANNOTATIONFORASSEMBLY_FILE.fields_by_name['type'].enum_type = _ANNOTATIONFORASSEMBLY_TYPE
_ANNOTATIONFORASSEMBLY_FILE.containing_type = _ANNOTATIONFORASSEMBLY
_ANNOTATIONFORASSEMBLY.fields_by_name['file'].message_type = _ANNOTATIONFORASSEMBLY_FILE
_ANNOTATIONFORASSEMBLY_TYPE.containing_type = _ANNOTATIONFORASSEMBLY
_ASSEMBLYDATASETREQUEST.fields_by_name['include_annotation_type'].enum_type = _ANNOTATIONFORASSEMBLY_TYPE
_ASSEMBLYDATASETREQUEST.fields_by_name['hydrated'].enum_type = _ASSEMBLYDATASETREQUEST_RESOLUTION
_ASSEMBLYDATASETREQUEST_RESOLUTION.containing_type = _ASSEMBLYDATASETREQUEST
_ORGANISM_COUNTS.containing_type = _ORGANISM
_ORGANISM_COUNTBYTYPE.fields_by_name['type'].enum_type = _ORGANISM_COUNTTYPE
_ORGANISM_COUNTBYTYPE.fields_by_name['counts'].message_type = _ORGANISM_COUNTS
_ORGANISM_COUNTBYTYPE.containing_type = _ORGANISM
_ORGANISM.fields_by_name['rank'].enum_type = _ORGANISM_RANKTYPE
_ORGANISM.fields_by_name['assembly_counts'].message_type = _ORGANISM_COUNTS
_ORGANISM.fields_by_name['counts'].message_type = _ORGANISM_COUNTBYTYPE
_ORGANISM.fields_by_name['children'].message_type = _ORGANISM
_ORGANISM.fields_by_name['merged'].message_type = _ORGANISM
_ORGANISM_RANKTYPE.containing_type = _ORGANISM
_ORGANISM_COUNTTYPE.containing_type = _ORGANISM
_TAXTREEREQUEST.oneofs_by_name['taxonomy'].fields.append(
  _TAXTREEREQUEST.fields_by_name['tax_id'])
_TAXTREEREQUEST.fields_by_name['tax_id'].containing_oneof = _TAXTREEREQUEST.oneofs_by_name['taxonomy']
_TAXTREEREQUEST.oneofs_by_name['taxonomy'].fields.append(
  _TAXTREEREQUEST.fields_by_name['tax_token'])
_TAXTREEREQUEST.fields_by_name['tax_token'].containing_oneof = _TAXTREEREQUEST.oneofs_by_name['taxonomy']
_TAXTREEREQUEST.oneofs_by_name['taxonomy'].fields.append(
  _TAXTREEREQUEST.fields_by_name['taxon'])
_TAXTREEREQUEST.fields_by_name['taxon'].containing_oneof = _TAXTREEREQUEST.oneofs_by_name['taxonomy']
_ORGANISMQUERYREQUEST.oneofs_by_name['id'].fields.append(
  _ORGANISMQUERYREQUEST.fields_by_name['organism_query'])
_ORGANISMQUERYREQUEST.fields_by_name['organism_query'].containing_oneof = _ORGANISMQUERYREQUEST.oneofs_by_name['id']
_ORGANISMQUERYREQUEST.oneofs_by_name['id'].fields.append(
  _ORGANISMQUERYREQUEST.fields_by_name['taxon_query'])
_ORGANISMQUERYREQUEST.fields_by_name['taxon_query'].containing_oneof = _ORGANISMQUERYREQUEST.oneofs_by_name['id']
_SCINAMEANDIDS_SCINAMEANDID.containing_type = _SCINAMEANDIDS
_SCINAMEANDIDS.fields_by_name['sci_name_and_ids'].message_type = _SCINAMEANDIDS_SCINAMEANDID
_ORTHOLOGREQUEST.fields_by_name['returned_content'].enum_type = _ORTHOLOGREQUEST_CONTENTTYPE
_ORTHOLOGREQUEST_CONTENTTYPE.containing_type = _ORTHOLOGREQUEST
_ORTHOLOGSET.fields_by_name['genes'].message_type = _GENEMETADATA
_GENEDATASETREQUEST_SYMBOLSFORTAXON.containing_type = _GENEDATASETREQUEST
_GENEDATASETREQUEST_GENEIDS.containing_type = _GENEDATASETREQUEST
_GENEDATASETREQUEST_ACCESSIONS.containing_type = _GENEDATASETREQUEST
_GENEDATASETREQUEST_SORT.fields_by_name['field'].enum_type = _GENEDATASETREQUEST_SORTFIELD
_GENEDATASETREQUEST_SORT.fields_by_name['direction'].enum_type = _SORTDIRECTION
_GENEDATASETREQUEST_SORT.containing_type = _GENEDATASETREQUEST
_GENEDATASETREQUEST.fields_by_name['symbols_for_taxon'].message_type = _GENEDATASETREQUEST_SYMBOLSFORTAXON
_GENEDATASETREQUEST.fields_by_name['include_annotation_type'].enum_type = _FASTA
_GENEDATASETREQUEST.fields_by_name['returned_content'].enum_type = _GENEDATASETREQUEST_CONTENTTYPE
_GENEDATASETREQUEST.fields_by_name['sort_schema'].message_type = _GENEDATASETREQUEST_SORT
_GENEDATASETREQUEST_CONTENTTYPE.containing_type = _GENEDATASETREQUEST
_GENEDATASETREQUEST_SORTFIELD.containing_type = _GENEDATASETREQUEST
_RANGE.fields_by_name['orientation'].enum_type = _ORIENTATION
_SEQRANGESET.fields_by_name['range'].message_type = _RANGE
_GENOMICLOCATION.fields_by_name['genomic_range'].message_type = _RANGE
_GENOMICLOCATION.fields_by_name['exons'].message_type = _RANGE
_PROTEIN.fields_by_name['mature_peptides'].message_type = _MATUREPEPTIDE
_TRANSCRIPT.fields_by_name['genomic_range'].message_type = _SEQRANGESET
_TRANSCRIPT.fields_by_name['exons'].message_type = _SEQRANGESET
_TRANSCRIPT.fields_by_name['cds'].message_type = _SEQRANGESET
_TRANSCRIPT.fields_by_name['genomic_locations'].message_type = _GENOMICLOCATION
_TRANSCRIPT.fields_by_name['protein'].message_type = _PROTEIN
_TRANSCRIPT.fields_by_name['type'].enum_type = _TRANSCRIPT_TRANSCRIPTTYPE
_TRANSCRIPT_TRANSCRIPTTYPE.containing_type = _TRANSCRIPT
_GENOMICREGION.fields_by_name['gene_range'].message_type = _SEQRANGESET
_GENOMICREGION.fields_by_name['type'].enum_type = _GENOMICREGION_GENOMICREGIONTYPE
_GENOMICREGION_GENOMICREGIONTYPE.containing_type = _GENOMICREGION
_ANNOTATION.fields_by_name['assemblies_in_scope'].message_type = _ANNOTATEDASSEMBLIES
_GENEDESCRIPTOR.fields_by_name['type'].enum_type = _GENEDESCRIPTOR_GENETYPE
_GENEDESCRIPTOR.fields_by_name['rna_type'].enum_type = _GENEDESCRIPTOR_RNATYPE
_GENEDESCRIPTOR.fields_by_name['orientation'].enum_type = _ORIENTATION
_GENEDESCRIPTOR.fields_by_name['genomic_ranges'].message_type = _SEQRANGESET
_GENEDESCRIPTOR.fields_by_name['reference_standards'].message_type = _GENOMICREGION
_GENEDESCRIPTOR.fields_by_name['genomic_regions'].message_type = _GENOMICREGION
_GENEDESCRIPTOR.fields_by_name['transcripts'].message_type = _TRANSCRIPT
_GENEDESCRIPTOR.fields_by_name['proteins'].message_type = _PROTEIN
_GENEDESCRIPTOR.fields_by_name['nomenclature_authority'].message_type = _NOMENCLATUREAUTHORITY
_GENEDESCRIPTOR.fields_by_name['annotations'].message_type = _ANNOTATION
_GENEDESCRIPTOR_GENETYPE.containing_type = _GENEDESCRIPTOR
_GENEDESCRIPTOR_RNATYPE.containing_type = _GENEDESCRIPTOR
_ERROR.fields_by_name['assembly_error_code'].enum_type = _ERROR_ASSEMBLYERRORCODE
_ERROR.fields_by_name['gene_error_code'].enum_type = _ERROR_GENEERRORCODE
_ERROR.fields_by_name['virus_error_code'].enum_type = _ERROR_VIRUSERRORCODE
_ERROR_ASSEMBLYERRORCODE.containing_type = _ERROR
_ERROR_GENEERRORCODE.containing_type = _ERROR
_ERROR_VIRUSERRORCODE.containing_type = _ERROR
_ERROR.oneofs_by_name['code'].fields.append(
  _ERROR.fields_by_name['assembly_error_code'])
_ERROR.fields_by_name['assembly_error_code'].containing_oneof = _ERROR.oneofs_by_name['code']
_ERROR.oneofs_by_name['code'].fields.append(
  _ERROR.fields_by_name['gene_error_code'])
_ERROR.fields_by_name['gene_error_code'].containing_oneof = _ERROR.oneofs_by_name['code']
_ERROR.oneofs_by_name['code'].fields.append(
  _ERROR.fields_by_name['virus_error_code'])
_ERROR.fields_by_name['virus_error_code'].containing_oneof = _ERROR.oneofs_by_name['code']
_WARNING_REPLACEDID.containing_type = _WARNING
_WARNING.fields_by_name['gene_warning_code'].enum_type = _WARNING_GENEWARNINGCODE
_WARNING.fields_by_name['replaced_id'].message_type = _WARNING_REPLACEDID
_WARNING_GENEWARNINGCODE.containing_type = _WARNING
_WARNING.oneofs_by_name['code'].fields.append(
  _WARNING.fields_by_name['gene_warning_code'])
_WARNING.fields_by_name['gene_warning_code'].containing_oneof = _WARNING.oneofs_by_name['code']
_MESSAGE.fields_by_name['error'].message_type = _ERROR
_MESSAGE.fields_by_name['warning'].message_type = _WARNING
_MESSAGE.oneofs_by_name['message'].fields.append(
  _MESSAGE.fields_by_name['error'])
_MESSAGE.fields_by_name['error'].containing_oneof = _MESSAGE.oneofs_by_name['message']
_MESSAGE.oneofs_by_name['message'].fields.append(
  _MESSAGE.fields_by_name['warning'])
_MESSAGE.fields_by_name['warning'].containing_oneof = _MESSAGE.oneofs_by_name['message']
_GENEMATCH.fields_by_name['messages'].message_type = _MESSAGE
_GENEMATCH.fields_by_name['warnings'].message_type = _WARNING
_GENEMATCH.fields_by_name['gene'].message_type = _GENEDESCRIPTOR
_GENEMATCH.fields_by_name['errors'].message_type = _ERROR
_GENEMETADATA.fields_by_name['messages'].message_type = _MESSAGE
_GENEMETADATA.fields_by_name['genes'].message_type = _GENEMATCH
_ASSEMBLYMATCH.fields_by_name['messages'].message_type = _MESSAGE
_ASSEMBLYMATCH.fields_by_name['assembly'].message_type = _ASSEMBLYDATASETDESCRIPTOR
_ASSEMBLYMETADATA.fields_by_name['messages'].message_type = _MESSAGE
_ASSEMBLYMETADATA.fields_by_name['assemblies'].message_type = _ASSEMBLYMATCH
_GENEDESCRIPTORS.fields_by_name['genes'].message_type = _GENEDESCRIPTOR
_GENEDESCRIPTORS.fields_by_name['errors'].message_type = _ERROR
_ANNOTATIONFORVIRUS_TYPE.containing_type = _ANNOTATIONFORVIRUS
_VIRUSDATASETREQUEST.fields_by_name['released_since'].message_type = google_dot_protobuf_dot_timestamp__pb2._TIMESTAMP
_VIRUSDATASETREQUEST.fields_by_name['table_fields'].enum_type = _VIRUSTABLEFIELD
_VIRUSDATASETREQUEST.fields_by_name['include_annotation_type'].enum_type = _ANNOTATIONFORVIRUS_TYPE
_VIRUSDATASETREQUEST.fields_by_name['format'].enum_type = _TABLEFORMAT
_VIRUSDATASETREQUEST.oneofs_by_name['id'].fields.append(
  _VIRUSDATASETREQUEST.fields_by_name['tax_id'])
_VIRUSDATASETREQUEST.fields_by_name['tax_id'].containing_oneof = _VIRUSDATASETREQUEST.oneofs_by_name['id']
_VIRUSDATASETREQUEST.oneofs_by_name['id'].fields.append(
  _VIRUSDATASETREQUEST.fields_by_name['tax_name'])
_VIRUSDATASETREQUEST.fields_by_name['tax_name'].containing_oneof = _VIRUSDATASETREQUEST.oneofs_by_name['id']
_VIRUSDATASETREQUEST.oneofs_by_name['id'].fields.append(
  _VIRUSDATASETREQUEST.fields_by_name['accession'])
_VIRUSDATASETREQUEST.fields_by_name['accession'].containing_oneof = _VIRUSDATASETREQUEST.oneofs_by_name['id']
_VIRUSDATASETREQUEST.oneofs_by_name['id'].fields.append(
  _VIRUSDATASETREQUEST.fields_by_name['taxon'])
_VIRUSDATASETREQUEST.fields_by_name['taxon'].containing_oneof = _VIRUSDATASETREQUEST.oneofs_by_name['id']
_SARS2PROTEINDATASETREQUEST.fields_by_name['released_since'].message_type = google_dot_protobuf_dot_timestamp__pb2._TIMESTAMP
_SARS2PROTEINDATASETREQUEST.fields_by_name['table_fields'].enum_type = _VIRUSTABLEFIELD
_SARS2PROTEINDATASETREQUEST.fields_by_name['include_annotation_type'].enum_type = _ANNOTATIONFORVIRUS_TYPE
_SARS2PROTEINDATASETREQUEST.fields_by_name['format'].enum_type = _TABLEFORMAT
_DOWNLOADSUMMARY_HYDRATED.containing_type = _DOWNLOADSUMMARY
_DOWNLOADSUMMARY_DEHYDRATED.containing_type = _DOWNLOADSUMMARY
_DOWNLOADSUMMARY.fields_by_name['resource_updated_on'].message_type = google_dot_protobuf_dot_timestamp__pb2._TIMESTAMP
_DOWNLOADSUMMARY.fields_by_name['hydrated'].message_type = _DOWNLOADSUMMARY_HYDRATED
_DOWNLOADSUMMARY.fields_by_name['dehydrated'].message_type = _DOWNLOADSUMMARY_DEHYDRATED
_DOWNLOADSUMMARY.fields_by_name['errors'].message_type = _ERROR
_DOWNLOADSUMMARY.fields_by_name['messages'].message_type = _MESSAGE
_MICROBIGGEDATASETREQUEST_ELEMENTFLANKCONFIG.containing_type = _MICROBIGGEDATASETREQUEST
_MICROBIGGEDATASETREQUEST.fields_by_name['files'].enum_type = _MICROBIGGEDATASETREQUEST_FILETYPE
_MICROBIGGEDATASETREQUEST.fields_by_name['element_flank_config'].message_type = _MICROBIGGEDATASETREQUEST_ELEMENTFLANKCONFIG
_MICROBIGGEDATASETREQUEST_FILETYPE.containing_type = _MICROBIGGEDATASETREQUEST
_MICROBIGGEDATASETREQUEST.oneofs_by_name['query'].fields.append(
  _MICROBIGGEDATASETREQUEST.fields_by_name['opaque_solr_query'])
_MICROBIGGEDATASETREQUEST.fields_by_name['opaque_solr_query'].containing_oneof = _MICROBIGGEDATASETREQUEST.oneofs_by_name['query']
_PROKARYOTEGENEREQUEST_GENEFLANKCONFIG.containing_type = _PROKARYOTEGENEREQUEST
_PROKARYOTEGENEREQUEST.fields_by_name['include_annotation_type'].enum_type = _FASTA
_PROKARYOTEGENEREQUEST.fields_by_name['gene_flank_config'].message_type = _PROKARYOTEGENEREQUEST_GENEFLANKCONFIG
DESCRIPTOR.message_types_by_name['MethodPayloadRequest'] = _METHODPAYLOADREQUEST
DESCRIPTOR.message_types_by_name['AssemblyDatasetDescriptorsFilter'] = _ASSEMBLYDATASETDESCRIPTORSFILTER
DESCRIPTOR.message_types_by_name['Accessions'] = _ACCESSIONS
DESCRIPTOR.message_types_by_name['AssemblyMetadataRequest'] = _ASSEMBLYMETADATAREQUEST
DESCRIPTOR.message_types_by_name['AssemblyDatasetDescriptorsRequest'] = _ASSEMBLYDATASETDESCRIPTORSREQUEST
DESCRIPTOR.message_types_by_name['AssemblyDatasetDescriptors'] = _ASSEMBLYDATASETDESCRIPTORS
DESCRIPTOR.message_types_by_name['BioProject'] = _BIOPROJECT
DESCRIPTOR.message_types_by_name['BioProjectLineage'] = _BIOPROJECTLINEAGE
DESCRIPTOR.message_types_by_name['AssemblyDatasetDescriptor'] = _ASSEMBLYDATASETDESCRIPTOR
DESCRIPTOR.message_types_by_name['AnnotationForAssembly'] = _ANNOTATIONFORASSEMBLY
DESCRIPTOR.message_types_by_name['AssemblyDatasetRequest'] = _ASSEMBLYDATASETREQUEST
DESCRIPTOR.message_types_by_name['AssemblyDatasetAvailability'] = _ASSEMBLYDATASETAVAILABILITY
DESCRIPTOR.message_types_by_name['TaxidRequest'] = _TAXIDREQUEST
DESCRIPTOR.message_types_by_name['TaxidReply'] = _TAXIDREPLY
DESCRIPTOR.message_types_by_name['Organism'] = _ORGANISM
DESCRIPTOR.message_types_by_name['TaxTreeRequest'] = _TAXTREEREQUEST
DESCRIPTOR.message_types_by_name['OrganismQueryRequest'] = _ORGANISMQUERYREQUEST
DESCRIPTOR.message_types_by_name['SciNameAndIds'] = _SCINAMEANDIDS
DESCRIPTOR.message_types_by_name['OrthologRequest'] = _ORTHOLOGREQUEST
DESCRIPTOR.message_types_by_name['OrthologSet'] = _ORTHOLOGSET
DESCRIPTOR.message_types_by_name['GeneDatasetRequest'] = _GENEDATASETREQUEST
DESCRIPTOR.message_types_by_name['Range'] = _RANGE
DESCRIPTOR.message_types_by_name['SeqRangeSet'] = _SEQRANGESET
DESCRIPTOR.message_types_by_name['GenomicLocation'] = _GENOMICLOCATION
DESCRIPTOR.message_types_by_name['MaturePeptide'] = _MATUREPEPTIDE
DESCRIPTOR.message_types_by_name['Protein'] = _PROTEIN
DESCRIPTOR.message_types_by_name['Transcript'] = _TRANSCRIPT
DESCRIPTOR.message_types_by_name['GenomicRegion'] = _GENOMICREGION
DESCRIPTOR.message_types_by_name['AnnotatedAssemblies'] = _ANNOTATEDASSEMBLIES
DESCRIPTOR.message_types_by_name['Annotation'] = _ANNOTATION
DESCRIPTOR.message_types_by_name['GeneDescriptor'] = _GENEDESCRIPTOR
DESCRIPTOR.message_types_by_name['NomenclatureAuthority'] = _NOMENCLATUREAUTHORITY
DESCRIPTOR.message_types_by_name['Error'] = _ERROR
DESCRIPTOR.message_types_by_name['Warning'] = _WARNING
DESCRIPTOR.message_types_by_name['Message'] = _MESSAGE
DESCRIPTOR.message_types_by_name['GeneMatch'] = _GENEMATCH
DESCRIPTOR.message_types_by_name['GeneMetadata'] = _GENEMETADATA
DESCRIPTOR.message_types_by_name['AssemblyMatch'] = _ASSEMBLYMATCH
DESCRIPTOR.message_types_by_name['AssemblyMetadata'] = _ASSEMBLYMETADATA
DESCRIPTOR.message_types_by_name['GeneDescriptors'] = _GENEDESCRIPTORS
DESCRIPTOR.message_types_by_name['AnnotationForVirus'] = _ANNOTATIONFORVIRUS
DESCRIPTOR.message_types_by_name['VirusDatasetRequest'] = _VIRUSDATASETREQUEST
DESCRIPTOR.message_types_by_name['Sars2ProteinDatasetRequest'] = _SARS2PROTEINDATASETREQUEST
DESCRIPTOR.message_types_by_name['DownloadSummary'] = _DOWNLOADSUMMARY
DESCRIPTOR.message_types_by_name['TabularOutput'] = _TABULAROUTPUT
DESCRIPTOR.message_types_by_name['MicroBiggeDatasetRequest'] = _MICROBIGGEDATASETREQUEST
DESCRIPTOR.message_types_by_name['VersionReply'] = _VERSIONREPLY
DESCRIPTOR.message_types_by_name['ProkaryoteGeneRequest'] = _PROKARYOTEGENEREQUEST
DESCRIPTOR.enum_types_by_name['SortDirection'] = _SORTDIRECTION
DESCRIPTOR.enum_types_by_name['Fasta'] = _FASTA
DESCRIPTOR.enum_types_by_name['Orientation'] = _ORIENTATION
DESCRIPTOR.enum_types_by_name['VirusTableField'] = _VIRUSTABLEFIELD
DESCRIPTOR.enum_types_by_name['TableFormat'] = _TABLEFORMAT
_sym_db.RegisterFileDescriptor(DESCRIPTOR)

MethodPayloadRequest = _reflection.GeneratedProtocolMessageType('MethodPayloadRequest', (_message.Message,), {
  'DESCRIPTOR' : _METHODPAYLOADREQUEST,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.MethodPayloadRequest)
  })
_sym_db.RegisterMessage(MethodPayloadRequest)

AssemblyDatasetDescriptorsFilter = _reflection.GeneratedProtocolMessageType('AssemblyDatasetDescriptorsFilter', (_message.Message,), {
  'DESCRIPTOR' : _ASSEMBLYDATASETDESCRIPTORSFILTER,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsFilter)
  })
_sym_db.RegisterMessage(AssemblyDatasetDescriptorsFilter)

Accessions = _reflection.GeneratedProtocolMessageType('Accessions', (_message.Message,), {
  'DESCRIPTOR' : _ACCESSIONS,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.Accessions)
  })
_sym_db.RegisterMessage(Accessions)

AssemblyMetadataRequest = _reflection.GeneratedProtocolMessageType('AssemblyMetadataRequest', (_message.Message,), {

  'Bioprojects' : _reflection.GeneratedProtocolMessageType('Bioprojects', (_message.Message,), {
    'DESCRIPTOR' : _ASSEMBLYMETADATAREQUEST_BIOPROJECTS,
    '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
    # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.AssemblyMetadataRequest.Bioprojects)
    })
  ,
  'DESCRIPTOR' : _ASSEMBLYMETADATAREQUEST,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.AssemblyMetadataRequest)
  })
_sym_db.RegisterMessage(AssemblyMetadataRequest)
_sym_db.RegisterMessage(AssemblyMetadataRequest.Bioprojects)

AssemblyDatasetDescriptorsRequest = _reflection.GeneratedProtocolMessageType('AssemblyDatasetDescriptorsRequest', (_message.Message,), {
  'DESCRIPTOR' : _ASSEMBLYDATASETDESCRIPTORSREQUEST,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.AssemblyDatasetDescriptorsRequest)
  })
_sym_db.RegisterMessage(AssemblyDatasetDescriptorsRequest)

AssemblyDatasetDescriptors = _reflection.GeneratedProtocolMessageType('AssemblyDatasetDescriptors', (_message.Message,), {
  'DESCRIPTOR' : _ASSEMBLYDATASETDESCRIPTORS,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.AssemblyDatasetDescriptors)
  })
_sym_db.RegisterMessage(AssemblyDatasetDescriptors)

BioProject = _reflection.GeneratedProtocolMessageType('BioProject', (_message.Message,), {
  'DESCRIPTOR' : _BIOPROJECT,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.BioProject)
  })
_sym_db.RegisterMessage(BioProject)

BioProjectLineage = _reflection.GeneratedProtocolMessageType('BioProjectLineage', (_message.Message,), {
  'DESCRIPTOR' : _BIOPROJECTLINEAGE,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.BioProjectLineage)
  })
_sym_db.RegisterMessage(BioProjectLineage)

AssemblyDatasetDescriptor = _reflection.GeneratedProtocolMessageType('AssemblyDatasetDescriptor', (_message.Message,), {
  'DESCRIPTOR' : _ASSEMBLYDATASETDESCRIPTOR,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.AssemblyDatasetDescriptor)
  })
_sym_db.RegisterMessage(AssemblyDatasetDescriptor)

AnnotationForAssembly = _reflection.GeneratedProtocolMessageType('AnnotationForAssembly', (_message.Message,), {

  'File' : _reflection.GeneratedProtocolMessageType('File', (_message.Message,), {
    'DESCRIPTOR' : _ANNOTATIONFORASSEMBLY_FILE,
    '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
    # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.AnnotationForAssembly.File)
    })
  ,
  'DESCRIPTOR' : _ANNOTATIONFORASSEMBLY,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.AnnotationForAssembly)
  })
_sym_db.RegisterMessage(AnnotationForAssembly)
_sym_db.RegisterMessage(AnnotationForAssembly.File)

AssemblyDatasetRequest = _reflection.GeneratedProtocolMessageType('AssemblyDatasetRequest', (_message.Message,), {
  'DESCRIPTOR' : _ASSEMBLYDATASETREQUEST,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.AssemblyDatasetRequest)
  })
_sym_db.RegisterMessage(AssemblyDatasetRequest)

AssemblyDatasetAvailability = _reflection.GeneratedProtocolMessageType('AssemblyDatasetAvailability', (_message.Message,), {
  'DESCRIPTOR' : _ASSEMBLYDATASETAVAILABILITY,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.AssemblyDatasetAvailability)
  })
_sym_db.RegisterMessage(AssemblyDatasetAvailability)

TaxidRequest = _reflection.GeneratedProtocolMessageType('TaxidRequest', (_message.Message,), {
  'DESCRIPTOR' : _TAXIDREQUEST,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.TaxidRequest)
  })
_sym_db.RegisterMessage(TaxidRequest)

TaxidReply = _reflection.GeneratedProtocolMessageType('TaxidReply', (_message.Message,), {
  'DESCRIPTOR' : _TAXIDREPLY,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.TaxidReply)
  })
_sym_db.RegisterMessage(TaxidReply)

Organism = _reflection.GeneratedProtocolMessageType('Organism', (_message.Message,), {

  'Counts' : _reflection.GeneratedProtocolMessageType('Counts', (_message.Message,), {
    'DESCRIPTOR' : _ORGANISM_COUNTS,
    '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
    # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.Organism.Counts)
    })
  ,

  'CountByType' : _reflection.GeneratedProtocolMessageType('CountByType', (_message.Message,), {
    'DESCRIPTOR' : _ORGANISM_COUNTBYTYPE,
    '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
    # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.Organism.CountByType)
    })
  ,
  'DESCRIPTOR' : _ORGANISM,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.Organism)
  })
_sym_db.RegisterMessage(Organism)
_sym_db.RegisterMessage(Organism.Counts)
_sym_db.RegisterMessage(Organism.CountByType)

TaxTreeRequest = _reflection.GeneratedProtocolMessageType('TaxTreeRequest', (_message.Message,), {
  'DESCRIPTOR' : _TAXTREEREQUEST,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.TaxTreeRequest)
  })
_sym_db.RegisterMessage(TaxTreeRequest)

OrganismQueryRequest = _reflection.GeneratedProtocolMessageType('OrganismQueryRequest', (_message.Message,), {
  'DESCRIPTOR' : _ORGANISMQUERYREQUEST,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.OrganismQueryRequest)
  })
_sym_db.RegisterMessage(OrganismQueryRequest)

SciNameAndIds = _reflection.GeneratedProtocolMessageType('SciNameAndIds', (_message.Message,), {

  'SciNameAndId' : _reflection.GeneratedProtocolMessageType('SciNameAndId', (_message.Message,), {
    'DESCRIPTOR' : _SCINAMEANDIDS_SCINAMEANDID,
    '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
    # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.SciNameAndIds.SciNameAndId)
    })
  ,
  'DESCRIPTOR' : _SCINAMEANDIDS,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.SciNameAndIds)
  })
_sym_db.RegisterMessage(SciNameAndIds)
_sym_db.RegisterMessage(SciNameAndIds.SciNameAndId)

OrthologRequest = _reflection.GeneratedProtocolMessageType('OrthologRequest', (_message.Message,), {
  'DESCRIPTOR' : _ORTHOLOGREQUEST,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.OrthologRequest)
  })
_sym_db.RegisterMessage(OrthologRequest)

OrthologSet = _reflection.GeneratedProtocolMessageType('OrthologSet', (_message.Message,), {
  'DESCRIPTOR' : _ORTHOLOGSET,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.OrthologSet)
  })
_sym_db.RegisterMessage(OrthologSet)

GeneDatasetRequest = _reflection.GeneratedProtocolMessageType('GeneDatasetRequest', (_message.Message,), {

  'SymbolsForTaxon' : _reflection.GeneratedProtocolMessageType('SymbolsForTaxon', (_message.Message,), {
    'DESCRIPTOR' : _GENEDATASETREQUEST_SYMBOLSFORTAXON,
    '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
    # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.GeneDatasetRequest.SymbolsForTaxon)
    })
  ,

  'GeneIds' : _reflection.GeneratedProtocolMessageType('GeneIds', (_message.Message,), {
    'DESCRIPTOR' : _GENEDATASETREQUEST_GENEIDS,
    '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
    # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.GeneDatasetRequest.GeneIds)
    })
  ,

  'Accessions' : _reflection.GeneratedProtocolMessageType('Accessions', (_message.Message,), {
    'DESCRIPTOR' : _GENEDATASETREQUEST_ACCESSIONS,
    '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
    # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.GeneDatasetRequest.Accessions)
    })
  ,

  'Sort' : _reflection.GeneratedProtocolMessageType('Sort', (_message.Message,), {
    'DESCRIPTOR' : _GENEDATASETREQUEST_SORT,
    '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
    # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.GeneDatasetRequest.Sort)
    })
  ,
  'DESCRIPTOR' : _GENEDATASETREQUEST,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.GeneDatasetRequest)
  })
_sym_db.RegisterMessage(GeneDatasetRequest)
_sym_db.RegisterMessage(GeneDatasetRequest.SymbolsForTaxon)
_sym_db.RegisterMessage(GeneDatasetRequest.GeneIds)
_sym_db.RegisterMessage(GeneDatasetRequest.Accessions)
_sym_db.RegisterMessage(GeneDatasetRequest.Sort)

Range = _reflection.GeneratedProtocolMessageType('Range', (_message.Message,), {
  'DESCRIPTOR' : _RANGE,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.Range)
  })
_sym_db.RegisterMessage(Range)

SeqRangeSet = _reflection.GeneratedProtocolMessageType('SeqRangeSet', (_message.Message,), {
  'DESCRIPTOR' : _SEQRANGESET,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.SeqRangeSet)
  })
_sym_db.RegisterMessage(SeqRangeSet)

GenomicLocation = _reflection.GeneratedProtocolMessageType('GenomicLocation', (_message.Message,), {
  'DESCRIPTOR' : _GENOMICLOCATION,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.GenomicLocation)
  })
_sym_db.RegisterMessage(GenomicLocation)

MaturePeptide = _reflection.GeneratedProtocolMessageType('MaturePeptide', (_message.Message,), {
  'DESCRIPTOR' : _MATUREPEPTIDE,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.MaturePeptide)
  })
_sym_db.RegisterMessage(MaturePeptide)

Protein = _reflection.GeneratedProtocolMessageType('Protein', (_message.Message,), {
  'DESCRIPTOR' : _PROTEIN,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.Protein)
  })
_sym_db.RegisterMessage(Protein)

Transcript = _reflection.GeneratedProtocolMessageType('Transcript', (_message.Message,), {
  'DESCRIPTOR' : _TRANSCRIPT,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.Transcript)
  })
_sym_db.RegisterMessage(Transcript)

GenomicRegion = _reflection.GeneratedProtocolMessageType('GenomicRegion', (_message.Message,), {
  'DESCRIPTOR' : _GENOMICREGION,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.GenomicRegion)
  })
_sym_db.RegisterMessage(GenomicRegion)

AnnotatedAssemblies = _reflection.GeneratedProtocolMessageType('AnnotatedAssemblies', (_message.Message,), {
  'DESCRIPTOR' : _ANNOTATEDASSEMBLIES,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.AnnotatedAssemblies)
  })
_sym_db.RegisterMessage(AnnotatedAssemblies)

Annotation = _reflection.GeneratedProtocolMessageType('Annotation', (_message.Message,), {
  'DESCRIPTOR' : _ANNOTATION,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.Annotation)
  })
_sym_db.RegisterMessage(Annotation)

GeneDescriptor = _reflection.GeneratedProtocolMessageType('GeneDescriptor', (_message.Message,), {
  'DESCRIPTOR' : _GENEDESCRIPTOR,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.GeneDescriptor)
  })
_sym_db.RegisterMessage(GeneDescriptor)

NomenclatureAuthority = _reflection.GeneratedProtocolMessageType('NomenclatureAuthority', (_message.Message,), {
  'DESCRIPTOR' : _NOMENCLATUREAUTHORITY,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.NomenclatureAuthority)
  })
_sym_db.RegisterMessage(NomenclatureAuthority)

Error = _reflection.GeneratedProtocolMessageType('Error', (_message.Message,), {
  'DESCRIPTOR' : _ERROR,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.Error)
  })
_sym_db.RegisterMessage(Error)

Warning = _reflection.GeneratedProtocolMessageType('Warning', (_message.Message,), {

  'ReplacedId' : _reflection.GeneratedProtocolMessageType('ReplacedId', (_message.Message,), {
    'DESCRIPTOR' : _WARNING_REPLACEDID,
    '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
    # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.Warning.ReplacedId)
    })
  ,
  'DESCRIPTOR' : _WARNING,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.Warning)
  })
_sym_db.RegisterMessage(Warning)
_sym_db.RegisterMessage(Warning.ReplacedId)

Message = _reflection.GeneratedProtocolMessageType('Message', (_message.Message,), {
  'DESCRIPTOR' : _MESSAGE,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.Message)
  })
_sym_db.RegisterMessage(Message)

GeneMatch = _reflection.GeneratedProtocolMessageType('GeneMatch', (_message.Message,), {
  'DESCRIPTOR' : _GENEMATCH,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.GeneMatch)
  })
_sym_db.RegisterMessage(GeneMatch)

GeneMetadata = _reflection.GeneratedProtocolMessageType('GeneMetadata', (_message.Message,), {
  'DESCRIPTOR' : _GENEMETADATA,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.GeneMetadata)
  })
_sym_db.RegisterMessage(GeneMetadata)

AssemblyMatch = _reflection.GeneratedProtocolMessageType('AssemblyMatch', (_message.Message,), {
  'DESCRIPTOR' : _ASSEMBLYMATCH,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.AssemblyMatch)
  })
_sym_db.RegisterMessage(AssemblyMatch)

AssemblyMetadata = _reflection.GeneratedProtocolMessageType('AssemblyMetadata', (_message.Message,), {
  'DESCRIPTOR' : _ASSEMBLYMETADATA,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.AssemblyMetadata)
  })
_sym_db.RegisterMessage(AssemblyMetadata)

GeneDescriptors = _reflection.GeneratedProtocolMessageType('GeneDescriptors', (_message.Message,), {
  'DESCRIPTOR' : _GENEDESCRIPTORS,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.GeneDescriptors)
  })
_sym_db.RegisterMessage(GeneDescriptors)

AnnotationForVirus = _reflection.GeneratedProtocolMessageType('AnnotationForVirus', (_message.Message,), {
  'DESCRIPTOR' : _ANNOTATIONFORVIRUS,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.AnnotationForVirus)
  })
_sym_db.RegisterMessage(AnnotationForVirus)

VirusDatasetRequest = _reflection.GeneratedProtocolMessageType('VirusDatasetRequest', (_message.Message,), {
  'DESCRIPTOR' : _VIRUSDATASETREQUEST,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.VirusDatasetRequest)
  })
_sym_db.RegisterMessage(VirusDatasetRequest)

Sars2ProteinDatasetRequest = _reflection.GeneratedProtocolMessageType('Sars2ProteinDatasetRequest', (_message.Message,), {
  'DESCRIPTOR' : _SARS2PROTEINDATASETREQUEST,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.Sars2ProteinDatasetRequest)
  })
_sym_db.RegisterMessage(Sars2ProteinDatasetRequest)

DownloadSummary = _reflection.GeneratedProtocolMessageType('DownloadSummary', (_message.Message,), {

  'Hydrated' : _reflection.GeneratedProtocolMessageType('Hydrated', (_message.Message,), {
    'DESCRIPTOR' : _DOWNLOADSUMMARY_HYDRATED,
    '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
    # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.DownloadSummary.Hydrated)
    })
  ,

  'Dehydrated' : _reflection.GeneratedProtocolMessageType('Dehydrated', (_message.Message,), {
    'DESCRIPTOR' : _DOWNLOADSUMMARY_DEHYDRATED,
    '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
    # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.DownloadSummary.Dehydrated)
    })
  ,
  'DESCRIPTOR' : _DOWNLOADSUMMARY,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.DownloadSummary)
  })
_sym_db.RegisterMessage(DownloadSummary)
_sym_db.RegisterMessage(DownloadSummary.Hydrated)
_sym_db.RegisterMessage(DownloadSummary.Dehydrated)

TabularOutput = _reflection.GeneratedProtocolMessageType('TabularOutput', (_message.Message,), {
  'DESCRIPTOR' : _TABULAROUTPUT,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.TabularOutput)
  })
_sym_db.RegisterMessage(TabularOutput)

MicroBiggeDatasetRequest = _reflection.GeneratedProtocolMessageType('MicroBiggeDatasetRequest', (_message.Message,), {

  'ElementFlankConfig' : _reflection.GeneratedProtocolMessageType('ElementFlankConfig', (_message.Message,), {
    'DESCRIPTOR' : _MICROBIGGEDATASETREQUEST_ELEMENTFLANKCONFIG,
    '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
    # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.MicroBiggeDatasetRequest.ElementFlankConfig)
    })
  ,
  'DESCRIPTOR' : _MICROBIGGEDATASETREQUEST,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.MicroBiggeDatasetRequest)
  })
_sym_db.RegisterMessage(MicroBiggeDatasetRequest)
_sym_db.RegisterMessage(MicroBiggeDatasetRequest.ElementFlankConfig)

VersionReply = _reflection.GeneratedProtocolMessageType('VersionReply', (_message.Message,), {
  'DESCRIPTOR' : _VERSIONREPLY,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.VersionReply)
  })
_sym_db.RegisterMessage(VersionReply)

ProkaryoteGeneRequest = _reflection.GeneratedProtocolMessageType('ProkaryoteGeneRequest', (_message.Message,), {

  'GeneFlankConfig' : _reflection.GeneratedProtocolMessageType('GeneFlankConfig', (_message.Message,), {
    'DESCRIPTOR' : _PROKARYOTEGENEREQUEST_GENEFLANKCONFIG,
    '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
    # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.ProkaryoteGeneRequest.GeneFlankConfig)
    })
  ,
  'DESCRIPTOR' : _PROKARYOTEGENEREQUEST,
  '__module__' : 'ncbi.datasets.v1alpha1.datasets_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.ProkaryoteGeneRequest)
  })
_sym_db.RegisterMessage(ProkaryoteGeneRequest)
_sym_db.RegisterMessage(ProkaryoteGeneRequest.GeneFlankConfig)


DESCRIPTOR._options = None
_ASSEMBLYDATASETDESCRIPTORSREQUEST.fields_by_name['org_name']._options = None
_ASSEMBLYDATASETDESCRIPTORSREQUEST.fields_by_name['cutoff']._options = None
_ANNOTATIONFORASSEMBLY_TYPE._options = None
_ANNOTATIONFORASSEMBLY_TYPE.values_by_name["GENOME_GB"]._options = None
_ASSEMBLYDATASETREQUEST.fields_by_name['assembly_accessions']._options = None
_ASSEMBLYDATASETREQUEST.fields_by_name['include_annotation']._options = None
_ORGANISM.fields_by_name['assembly_count']._options = None
_ORGANISM.fields_by_name['assembly_counts']._options = None
_ORGANISM.fields_by_name['counts']._options = None
_ORGANISM.fields_by_name['children']._options = None
_ORGANISM.fields_by_name['merged']._options = None
_ORGANISM.fields_by_name['merged_tax_ids']._options = None
_ORGANISM.fields_by_name['min_ord']._options = None
_ORGANISM.fields_by_name['max_ord']._options = None
_ORGANISM.fields_by_name['key']._options = None
_ORGANISM.fields_by_name['title']._options = None
_ORGANISM.fields_by_name['icon']._options = None
_TAXTREEREQUEST.fields_by_name['tax_id']._options = None
_TAXTREEREQUEST.fields_by_name['tax_token']._options = None
_ORGANISMQUERYREQUEST.fields_by_name['organism_query']._options = None
_TRANSCRIPT.fields_by_name['genomic_range']._options = None
_TRANSCRIPT.fields_by_name['exons']._options = None
_GENEDESCRIPTOR.fields_by_name['chromosome']._options = None
_ERROR.fields_by_name['valid_identifiers']._options = None
_GENEMATCH.fields_by_name['messages']._options = None
_VIRUSDATASETREQUEST.fields_by_name['tax_id']._options = None
_VIRUSDATASETREQUEST.fields_by_name['tax_name']._options = None
_DOWNLOADSUMMARY.fields_by_name['assembly_count']._options = None
_DOWNLOADSUMMARY.fields_by_name['errors']._options = None
# @@protoc_insertion_point(module_scope)
