
Generation of test_meales.sam/bam/cram
======================================

looks like the sam and bam were generated with a measles.fa edited to simplify
the name of the chromosome into 'chr1'. Earlier version of samtools could read
the cram without the reference, but not newest one (v1.9). So I copied the
samtools locally, rename the chromosome into chr1 and then typed::

     samtools faidx measles.fa
     samtools view -@ 8 -C test_measles.bam -T measles.fa -o test_measles.cram

test_bamtools.py worked afterwards



test_pacbio_barcode_report.csv: file from smrtlink pacbio report after
demultiplexing to test BarCode class

test_laa_amplicon_analysis_summary_bc2*.csv: files from laa tool output to test
laa.LAA class





This file is the output of the barcode report from smrtlink v6.0:
./resources/testing/test_pacbio_barcode_report.csv


The following files are used by test_multiqc to test the sub packages
./sequana/multiqc. Those files are generated by the pacbio_qc and
quality_control pipelines (summary.json). 

- summary_pacbio_qc1.json
- summary_pacbio_qc2.json
- summary_pacbio_qc3.json
- summary_qc.json


test_CS_tiny.bam is a BAM file created with minimap2 using --cs option


./test_isoseq_lq_sirv.bam is a BAM file related to isoseq analysis. It is the
mapping of a set of LQ isoforms on the SIRV reference. This can be used to test
the isoseq.PacbioIsoSeqMultipleIsoforms class.

adapters_with_duplicates.fa
test_adapter_removal_output.txt
test_atropos_paired.txt

test.bam
test_blast_ebi.xml
test_blast_local.xml
test_coverage.csv
test_cutadapt_paired.txt
test_dag.ann.dot
test_dag.dot
test_expdesign_generic.csv
test_expdesign_hiseq.csv
test_expdesign_hiseq_doubleindex.csv
test_expdesign_hiseq_duplicated_index.csv
test_expdesign_miseq_illumina_1.csv
test_expdesign_miseq_illumina2.csv
test_expdesign_miseq_illumina.csv
test_expdesign_rubicon.csv
test_expdesign_wrong.csv
test.fasta
test_fasta.fasta
test.fastq
test.fastq.gz
test_gui_generic_config.yaml
test_gui_generic_snakefile_configfile.rules
test_gui_generic_snakefile_config.rules
test_gui_generic_snakefile_noconfig.rules
test_index_mapper.csv
test_JB409847.bed
test_kraken.out
test_kraken.out.summary
test_krona_k1.tsv
test_krona_k2.tsv
test_pacbio_subreads.bam
test_R1_.mapped.fastq
test_R1.mapped.fastq
test_R1_.unmapped.fastq
test_R1.unmapped.fastq
test_R2_.mapped.fastq
test_R2.mapped.fastq
test_R2_.unmapped.fastq
test_R2.unmapped.fastq
test.sam
test_snakemake_stats.txt
test_snpeff_ref.gb
test_summary_fastq_stats.json
test_summary_module.svg
test_taxon_rtd.csv
test.vcf

