# -*- coding: utf-8 -*-
# Generated by the protocol buffer compiler.  DO NOT EDIT!
# source: ncbi/datasets/v1alpha1/reports/assembly.proto

from google.protobuf import descriptor as _descriptor
from google.protobuf import message as _message
from google.protobuf import reflection as _reflection
from google.protobuf import symbol_database as _symbol_database
# @@protoc_insertion_point(imports)

_sym_db = _symbol_database.Default()


from ncbi.datasets.v1alpha1.reports import options_pb2 as ncbi_dot_datasets_dot_v1alpha1_dot_reports_dot_options__pb2


DESCRIPTOR = _descriptor.FileDescriptor(
  name='ncbi/datasets/v1alpha1/reports/assembly.proto',
  package='ncbi.datasets.v1alpha1.reports',
  syntax='proto3',
  serialized_options=b'Z\036ncbi/datasets/v1alpha1/reports\370\001\001\302\363\030\370\010\n\033Genome Assembly Data Report\022\022AssemblyDataReport\032G<p>The downloaded genome package contains a genome assembly data report\032^in <a href=\"https://jsonlines.readthedocs.io/\">JSON lines</a> format at the following location\032tin your downloaded data package: <em>ncbi_dataset/data/data_report.jsonl</em>. <br> Each line of the genome assembly\032pdata report file is a hierarchichal object serialized as JSON and each line represents a single assembly record.\032\215\001The tables below represent the schema. The outermost object is an <em>AssemblyDataReport</em> object. Its fields are documented in the tables\032\nbelow.</p>\032O<p>The second column in each table documents the mnemonic string used by the <a\032Phref=\"https://api.ncbi.nlm.nih.gov/datasets/docs/cli/dataformat/\">dataformat</a>\032ccommand-line tool to transform genome assembly data reports from JSON lines to tabular formats.</p>\032k<p>Check the <a href=\"https://api.ncbi.nlm.nih.gov/datasets/docs/data-report-schemas-overview/\">Data Report\032DOverview</a> documentation for examples of using this table with the\032]<a href=\"https://api.ncbi.nlm.nih.gov/datasets/docs/cli/dataformat/\">dataformat</a> tool.</p>',
  serialized_pb=b'\n-ncbi/datasets/v1alpha1/reports/assembly.proto\x12\x1encbi.datasets.v1alpha1.reports\x1a,ncbi/datasets/v1alpha1/reports/options.proto\"\xc6\x08\n\x12\x41ssemblyDataReport\x12g\n\x0b\x63ommon_name\x18\x01 \x01(\tBF\xc2\xf3\x18\x42\n\x0b\x63ommon-name\x12\x0b\x43ommon name2\x05Human2\tzebrafish2\x14pacific white shrimpR\ncommonName\x12\x83\x01\n\rorganism_name\x18\x02 \x01(\tB^\xc2\xf3\x18Z\n\rorganism-name\x12\rOrganism name2\x0cHomo sapiens2\x14\x41rabidopsis thaliana2\x16\x43\x61nis lupus familiarisR\x0corganismName\x12\x39\n\x05\x62reed\x18\x03 \x01(\tB#\xc2\xf3\x18\x1f\n\x05\x62reed\x12\x05\x42reed2\x08Hereford2\x05\x62oxerR\x05\x62reed\x12\x39\n\x08\x63ultivar\x18\x04 \x01(\tB\x1d\xc2\xf3\x18\x19\n\x08\x63ultivar\x12\x08\x43ultivar2\x03\x42\x37\x33R\x08\x63ultivar\x12\x38\n\x07\x65\x63otype\x18\x05 \x01(\tB\x1e\xc2\xf3\x18\x1a\n\x07\x65\x63otype\x12\x07\x45\x63otype2\x06\x41lpineR\x07\x65\x63otype\x12j\n\x07isolate\x18\x06 \x01(\tBP\xc2\xf3\x18L\n\x07isolate\x12\x07Isolate2/L1 Dominette 01449 registration number 421906802\x07Pmale09R\x07isolate\x12(\n\x03sex\x18\x07 \x01(\tB\x16\xc2\xf3\x18\x12\n\x03sex\x12\x03Sex2\x06\x66\x65maleR\x03sex\x12,\n\x06strain\x18\x08 \x01(\tB\x14\xc2\xf3\x18\x10\n\x06strain\x12\x06StrainR\x06strain\x12\x31\n\x06tax_id\x18\t \x01(\rB\x1a\xc2\xf3\x18\x16\n\x06tax-id\x12\x0cTaxonomic IDR\x05taxId\x12Q\n\rassembly_info\x18\n \x01(\x0b\x32,.ncbi.datasets.v1alpha1.reports.AssemblyInfoR\x0c\x61ssemblyInfo\x12T\n\x0e\x61ssembly_stats\x18\x0b \x01(\x0b\x32-.ncbi.datasets.v1alpha1.reports.AssemblyStatsR\rassemblyStats\x12T\n\x0eorganelle_info\x18\x0c \x03(\x0b\x32-.ncbi.datasets.v1alpha1.reports.OrganelleInfoR\rorganelleInfo\x12W\n\x0f\x61nnotation_info\x18\r \x01(\x0b\x32..ncbi.datasets.v1alpha1.reports.AnnotationInfoR\x0e\x61nnotationInfo\x12\x42\n\x08wgs_info\x18\x0e \x01(\x0b\x32\'.ncbi.datasets.v1alpha1.reports.WGSInfoR\x07wgsInfo\"\xf0\x03\n\nBioProject\x12Z\n\taccession\x18\x01 \x01(\tB<\xc2\xf3\x18\x38\n\x14\x62ioproject-accession\x12\x14\x42ioProject accession2\nPRJEB35387R\taccession\x12~\n\x05title\x18\x02 \x01(\tBh\xc2\xf3\x18\x64\n\x10\x62ioproject-title\x12\x10\x42ioProject title2>Sciurus carolinensis (grey squirrel) genome assembly, mSciCar1R\x05title\x12k\n\x10parent_accession\x18\x03 \x01(\tB@\x18\x01\xc2\xf3\x18:\n\x1b\x62ioproject-parent-accession\x12\x1b\x42ioProject parent accessionR\x0fparentAccession\x12\x98\x01\n\x11parent_accessions\x18\x04 \x03(\tBk\xc2\xf3\x18g\n\x1c\x62ioproject-parent-accessions\x12\x1c\x42ioProject parent accessions2)[\"PRJNA489243\",\"PRJEB33226\",\"PRJEB40665\"]R\x10parentAccessions\"a\n\x11\x42ioProjectLineage\x12L\n\x0b\x62ioprojects\x18\x01 \x03(\x0b\x32*.ncbi.datasets.v1alpha1.reports.BioProjectR\x0b\x62ioprojects\"\xe1\n\n\x0c\x41ssemblyInfo\x12\x65\n\x0e\x61ssembly_level\x18\x01 \x01(\tB>\xc2\xf3\x18:\n\nassm-level\x12\x0e\x41ssembly level2\nchromosome2\x08scaffold2\x06\x63ontigR\rassemblyLevel\x12\\\n\rassembly_name\x18\x02 \x01(\tB7\xc2\xf3\x18\x33\n\tassm-name\x12\rAssembly name2\nGRCh38.p132\x0b\x41SM985889v3R\x0c\x61ssemblyName\x12\x63\n\rassembly_type\x18\x03 \x01(\tB>\xc2\xf3\x18:\n\tassm-type\x12\rAssembly type2\x15haploid-with-alt-loci2\x07haploidR\x0c\x61ssemblyType\x12`\n\x12\x62ioproject_lineage\x18\r \x03(\x0b\x32\x31.ncbi.datasets.v1alpha1.reports.BioProjectLineageR\x11\x62ioprojectLineage\x12O\n\x0fsubmission_date\x18\x05 \x01(\tB&\xc2\xf3\x18\"\n\x0fsubmission-date\x12\x0fSubmission dateR\x0esubmissionDate\x12N\n\x0b\x64\x65scription\x18\x06 \x01(\tB,\xc2\xf3\x18(\n\x10\x61ssm-description\x12\x14\x41ssembly descriptionR\x0b\x64\x65scription\x12\x80\x01\n\x16genbank_assm_accession\x18\x07 \x01(\tBJ\xc2\xf3\x18\x46\n\x16genbank-assm-accession\x12\x1aGenBank assembly accession2\x10GCA_000001405.28R\x14genbankAssmAccession\x12\x46\n\tsubmitter\x18\x08 \x01(\tB(\xc2\xf3\x18$\n\x0e\x61ssm-submitter\x12\x12\x41ssembly submitterR\tsubmitter\x12x\n\x0frefseq_category\x18\t \x01(\tBO\xc2\xf3\x18K\n\x0frefseq-category\x12\x0fRefseq category2\x10reference genome2\x15representative genomeR\x0erefseqCategory\x12|\n\x15refseq_assm_accession\x18\n \x01(\tBH\xc2\xf3\x18\x44\n\x15refseq-assm-accession\x12\x19RefSeq assembly accession2\x10GCF_000001405.39R\x13refseqAssmAccession\x12T\n\x0eucsc_assm_name\x18\x0b \x01(\tB.\xc2\xf3\x18*\n\x0eucsc-assm-name\x12\x12UCSC assembly name2\x04hg38R\x0cucscAssmName\x12K\n\x0flinked_assembly\x18\x0c \x01(\tB\"\xc2\xf3\x18\x1e\n\x0blinked-assm\x12\x0fLinked assemblyR\x0elinkedAssembly\x12O\n\x0fsequencing_tech\x18\x0e \x01(\tB&\xc2\xf3\x18\"\n\x0fsequencing-tech\x12\x0fSequencing techR\x0esequencingTech\x12m\n\x13\x62iosample_accession\x18\x0f \x01(\tB<\xc2\xf3\x18\x38\n\x13\x62iosample-accession\x12\x13\x42ioSample accession2\x0cSAMN03145444R\x12\x62iosampleAccession\"\xa0\x08\n\rAssemblyStats\x12}\n\x1btotal_number_of_chromosomes\x18\x01 \x01(\rB>\xc2\xf3\x18:\n\x1btotal-number-of-chromosomes\x12\x1bTotal number of chromosomesR\x18totalNumberOfChromosomes\x12\x63\n\x15total_sequence_length\x18\x02 \x01(\x04\x42/\xc2\xf3\x18+\n\x12total-sequence-len\x12\x15Total sequence lengthR\x13totalSequenceLength\x12\x63\n\x15total_ungapped_length\x18\x03 \x01(\x04\x42/\xc2\xf3\x18+\n\x12total-ungapped-len\x12\x15Total ungapped lengthR\x13totalUngappedLength\x12V\n\x11number_of_contigs\x18\x04 \x01(\rB*\xc2\xf3\x18&\n\x11number-of-contigs\x12\x11Number of contigsR\x0fnumberOfContigs\x12;\n\ncontig_n50\x18\x05 \x01(\rB\x1c\xc2\xf3\x18\x18\n\ncontig-n50\x12\nContig N50R\tcontigN50\x12;\n\ncontig_l50\x18\x06 \x01(\rB\x1c\xc2\xf3\x18\x18\n\ncontig-l50\x12\nContig L50R\tcontigL50\x12^\n\x13number_of_scaffolds\x18\x07 \x01(\rB.\xc2\xf3\x18*\n\x13number-of-scaffolds\x12\x13Number of scaffoldsR\x11numberOfScaffolds\x12\x43\n\x0cscaffold_n50\x18\x08 \x01(\rB \xc2\xf3\x18\x1c\n\x0cscaffold-n50\x12\x0cScaffold N50R\x0bscaffoldN50\x12\x43\n\x0cscaffold_l50\x18\t \x01(\rB \xc2\xf3\x18\x1c\n\x0cscaffold-l50\x12\x0cScaffold L50R\x0bscaffoldL50\x12\x81\x01\n\x1cgaps_between_scaffolds_count\x18\n \x01(\rB@\xc2\xf3\x18<\n\x1cgaps-between-scaffolds-count\x12\x1cGaps between scaffolds countR\x19gapsBetweenScaffoldsCount\x12\x85\x01\n\x1dnumber_of_component_sequences\x18\x0b \x01(\rBB\xc2\xf3\x18>\n\x1dnumber-of-component-sequences\x12\x1dNumber of component sequencesR\x1anumberOfComponentSequences\"\xed\x01\n\rOrganelleInfo\x12#\n\rassembly_name\x18\x01 \x01(\tR\x0c\x61ssemblyName\x12-\n\x12infraspecific_name\x18\x02 \x01(\tR\x11infraspecificName\x12\x1e\n\nbioproject\x18\x03 \x03(\tR\nbioproject\x12 \n\x0b\x64\x65scription\x18\x04 \x01(\tR\x0b\x64\x65scription\x12(\n\x10total_seq_length\x18\x05 \x01(\x04R\x0etotalSeqLength\x12\x1c\n\tsubmitter\x18\x06 \x01(\tR\tsubmitter\"\x89\x07\n\x0e\x41nnotationInfo\x12\x35\n\x04name\x18\x01 \x01(\tB!\xc2\xf3\x18\x1d\n\nannot-name\x12\x0f\x41nnotation nameR\x04name\x12=\n\x06source\x18\x02 \x01(\tB%\xc2\xf3\x18!\n\x0c\x61nnot-source\x12\x11\x41nnotation sourceR\x06source\x12T\n\x0crelease_date\x18\x03 \x01(\tB1\xc2\xf3\x18-\n\x12\x61nnot-release-date\x12\x17\x41nnotation release dateR\x0breleaseDate\x12L\n\nreport_url\x18\x04 \x01(\tB-\xc2\xf3\x18)\n\x10\x61nnot-report-url\x12\x15\x41nnotation report URLR\treportUrl\x12R\n\x05stats\x18\x05 \x01(\x0b\x32<.ncbi.datasets.v1alpha1.reports.AnnotationInfo.FeatureCountsR\x05stats\x1ak\n\rFeatureCounts\x12Z\n\x0bgene_counts\x18\x01 \x01(\x0b\x32\x39.ncbi.datasets.v1alpha1.reports.AnnotationInfo.GeneCountsR\ngeneCounts\x1a\x9b\x03\n\nGeneCounts\x12?\n\x05total\x18\x01 \x01(\rB)\xc2\xf3\x18%\n\x11gene-counts-total\x12\x10Gene count totalR\x05total\x12\x62\n\x0eprotein_coding\x18\x02 \x01(\rB;\xc2\xf3\x18\x37\n\x1agene-counts-protein-coding\x12\x19Gene count protein-codingR\rproteinCoding\x12R\n\nnon_coding\x18\x03 \x01(\rB3\xc2\xf3\x18/\n\x16gene-counts-non-coding\x12\x15Gene count non-codingR\tnonCoding\x12S\n\npseudogene\x18\x04 \x01(\rB3\xc2\xf3\x18/\n\x16gene-counts-pseudogene\x12\x15Gene count pseudogeneR\npseudogene\x12?\n\x05other\x18\x05 \x01(\rB)\xc2\xf3\x18%\n\x11gene-counts-other\x12\x10Gene count otherR\x05other\"\xf6\x02\n\x07WGSInfo\x12x\n\x15wgs_project_accession\x18\x01 \x01(\tBD\xc2\xf3\x18@\n\x15wgs-project-accession\x12\x15WGS project accession2\x06\x41\x41\x45X032\x08\x43\x41\x42HLF01R\x13wgsProjectAccession\x12q\n\x0emaster_wgs_url\x18\x02 \x01(\tBK\xc2\xf3\x18G\n\x07wgs-url\x12\x07WGS URL23https://www.ncbi.nlm.nih.gov/nuccore/AAEX00000000.3R\x0cmasterWgsUrl\x12~\n\x0fwgs_contigs_url\x18\x03 \x01(\tBV\xc2\xf3\x18R\n\x0fwgs-contigs-url\x12\x0fWGS contigs URL2.https://www.ncbi.nlm.nih.gov/Traces/wgs/AAEX03R\rwgsContigsUrlB\xa0\tZ\x1encbi/datasets/v1alpha1/reports\xf8\x01\x01\xc2\xf3\x18\xf8\x08\n\x1bGenome Assembly Data Report\x12\x12\x41ssemblyDataReport\x1aG<p>The downloaded genome package contains a genome assembly data report\x1a^in <a href=\"https://jsonlines.readthedocs.io/\">JSON lines</a> format at the following location\x1atin your downloaded data package: <em>ncbi_dataset/data/data_report.jsonl</em>. <br> Each line of the genome assembly\x1apdata report file is a hierarchichal object serialized as JSON and each line represents a single assembly record.\x1a\x8d\x01The tables below represent the schema. The outermost object is an <em>AssemblyDataReport</em> object. Its fields are documented in the tables\x1a\nbelow.</p>\x1aO<p>The second column in each table documents the mnemonic string used by the <a\x1aPhref=\"https://api.ncbi.nlm.nih.gov/datasets/docs/cli/dataformat/\">dataformat</a>\x1a\x63\x63ommand-line tool to transform genome assembly data reports from JSON lines to tabular formats.</p>\x1ak<p>Check the <a href=\"https://api.ncbi.nlm.nih.gov/datasets/docs/data-report-schemas-overview/\">Data Report\x1a\x44Overview</a> documentation for examples of using this table with the\x1a]<a href=\"https://api.ncbi.nlm.nih.gov/datasets/docs/cli/dataformat/\">dataformat</a> tool.</p>b\x06proto3'
  ,
  dependencies=[ncbi_dot_datasets_dot_v1alpha1_dot_reports_dot_options__pb2.DESCRIPTOR,])




_ASSEMBLYDATAREPORT = _descriptor.Descriptor(
  name='AssemblyDataReport',
  full_name='ncbi.datasets.v1alpha1.reports.AssemblyDataReport',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='common_name', full_name='ncbi.datasets.v1alpha1.reports.AssemblyDataReport.common_name', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030B\n\013common-name\022\013Common name2\005Human2\tzebrafish2\024pacific white shrimp', json_name='commonName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='organism_name', full_name='ncbi.datasets.v1alpha1.reports.AssemblyDataReport.organism_name', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030Z\n\rorganism-name\022\rOrganism name2\014Homo sapiens2\024Arabidopsis thaliana2\026Canis lupus familiaris', json_name='organismName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='breed', full_name='ncbi.datasets.v1alpha1.reports.AssemblyDataReport.breed', index=2,
      number=3, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\037\n\005breed\022\005Breed2\010Hereford2\005boxer', json_name='breed', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='cultivar', full_name='ncbi.datasets.v1alpha1.reports.AssemblyDataReport.cultivar', index=3,
      number=4, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\031\n\010cultivar\022\010Cultivar2\003B73', json_name='cultivar', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='ecotype', full_name='ncbi.datasets.v1alpha1.reports.AssemblyDataReport.ecotype', index=4,
      number=5, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\032\n\007ecotype\022\007Ecotype2\006Alpine', json_name='ecotype', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='isolate', full_name='ncbi.datasets.v1alpha1.reports.AssemblyDataReport.isolate', index=5,
      number=6, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030L\n\007isolate\022\007Isolate2/L1 Dominette 01449 registration number 421906802\007Pmale09', json_name='isolate', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='sex', full_name='ncbi.datasets.v1alpha1.reports.AssemblyDataReport.sex', index=6,
      number=7, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\022\n\003sex\022\003Sex2\006female', json_name='sex', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='strain', full_name='ncbi.datasets.v1alpha1.reports.AssemblyDataReport.strain', index=7,
      number=8, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\020\n\006strain\022\006Strain', json_name='strain', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='tax_id', full_name='ncbi.datasets.v1alpha1.reports.AssemblyDataReport.tax_id', index=8,
      number=9, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\026\n\006tax-id\022\014Taxonomic ID', json_name='taxId', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='assembly_info', full_name='ncbi.datasets.v1alpha1.reports.AssemblyDataReport.assembly_info', index=9,
      number=10, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='assemblyInfo', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='assembly_stats', full_name='ncbi.datasets.v1alpha1.reports.AssemblyDataReport.assembly_stats', index=10,
      number=11, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='assemblyStats', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='organelle_info', full_name='ncbi.datasets.v1alpha1.reports.AssemblyDataReport.organelle_info', index=11,
      number=12, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='organelleInfo', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='annotation_info', full_name='ncbi.datasets.v1alpha1.reports.AssemblyDataReport.annotation_info', index=12,
      number=13, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='annotationInfo', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='wgs_info', full_name='ncbi.datasets.v1alpha1.reports.AssemblyDataReport.wgs_info', index=13,
      number=14, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='wgsInfo', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=128,
  serialized_end=1222,
)


_BIOPROJECT = _descriptor.Descriptor(
  name='BioProject',
  full_name='ncbi.datasets.v1alpha1.reports.BioProject',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='accession', full_name='ncbi.datasets.v1alpha1.reports.BioProject.accession', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\0308\n\024bioproject-accession\022\024BioProject accession2\nPRJEB35387', json_name='accession', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='title', full_name='ncbi.datasets.v1alpha1.reports.BioProject.title', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030d\n\020bioproject-title\022\020BioProject title2>Sciurus carolinensis (grey squirrel) genome assembly, mSciCar1', json_name='title', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='parent_accession', full_name='ncbi.datasets.v1alpha1.reports.BioProject.parent_accession', index=2,
      number=3, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001\302\363\030:\n\033bioproject-parent-accession\022\033BioProject parent accession', json_name='parentAccession', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='parent_accessions', full_name='ncbi.datasets.v1alpha1.reports.BioProject.parent_accessions', index=3,
      number=4, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030g\n\034bioproject-parent-accessions\022\034BioProject parent accessions2)[\"PRJNA489243\",\"PRJEB33226\",\"PRJEB40665\"]', json_name='parentAccessions', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=1225,
  serialized_end=1721,
)


_BIOPROJECTLINEAGE = _descriptor.Descriptor(
  name='BioProjectLineage',
  full_name='ncbi.datasets.v1alpha1.reports.BioProjectLineage',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='bioprojects', full_name='ncbi.datasets.v1alpha1.reports.BioProjectLineage.bioprojects', index=0,
      number=1, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='bioprojects', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=1723,
  serialized_end=1820,
)


_ASSEMBLYINFO = _descriptor.Descriptor(
  name='AssemblyInfo',
  full_name='ncbi.datasets.v1alpha1.reports.AssemblyInfo',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='assembly_level', full_name='ncbi.datasets.v1alpha1.reports.AssemblyInfo.assembly_level', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030:\n\nassm-level\022\016Assembly level2\nchromosome2\010scaffold2\006contig', json_name='assemblyLevel', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='assembly_name', full_name='ncbi.datasets.v1alpha1.reports.AssemblyInfo.assembly_name', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\0303\n\tassm-name\022\rAssembly name2\nGRCh38.p132\013ASM985889v3', json_name='assemblyName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='assembly_type', full_name='ncbi.datasets.v1alpha1.reports.AssemblyInfo.assembly_type', index=2,
      number=3, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030:\n\tassm-type\022\rAssembly type2\025haploid-with-alt-loci2\007haploid', json_name='assemblyType', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='bioproject_lineage', full_name='ncbi.datasets.v1alpha1.reports.AssemblyInfo.bioproject_lineage', index=3,
      number=13, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='bioprojectLineage', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='submission_date', full_name='ncbi.datasets.v1alpha1.reports.AssemblyInfo.submission_date', index=4,
      number=5, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\"\n\017submission-date\022\017Submission date', json_name='submissionDate', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='description', full_name='ncbi.datasets.v1alpha1.reports.AssemblyInfo.description', index=5,
      number=6, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030(\n\020assm-description\022\024Assembly description', json_name='description', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='genbank_assm_accession', full_name='ncbi.datasets.v1alpha1.reports.AssemblyInfo.genbank_assm_accession', index=6,
      number=7, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030F\n\026genbank-assm-accession\022\032GenBank assembly accession2\020GCA_000001405.28', json_name='genbankAssmAccession', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='submitter', full_name='ncbi.datasets.v1alpha1.reports.AssemblyInfo.submitter', index=7,
      number=8, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030$\n\016assm-submitter\022\022Assembly submitter', json_name='submitter', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='refseq_category', full_name='ncbi.datasets.v1alpha1.reports.AssemblyInfo.refseq_category', index=8,
      number=9, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030K\n\017refseq-category\022\017Refseq category2\020reference genome2\025representative genome', json_name='refseqCategory', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='refseq_assm_accession', full_name='ncbi.datasets.v1alpha1.reports.AssemblyInfo.refseq_assm_accession', index=9,
      number=10, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030D\n\025refseq-assm-accession\022\031RefSeq assembly accession2\020GCF_000001405.39', json_name='refseqAssmAccession', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='ucsc_assm_name', full_name='ncbi.datasets.v1alpha1.reports.AssemblyInfo.ucsc_assm_name', index=10,
      number=11, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030*\n\016ucsc-assm-name\022\022UCSC assembly name2\004hg38', json_name='ucscAssmName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='linked_assembly', full_name='ncbi.datasets.v1alpha1.reports.AssemblyInfo.linked_assembly', index=11,
      number=12, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\036\n\013linked-assm\022\017Linked assembly', json_name='linkedAssembly', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='sequencing_tech', full_name='ncbi.datasets.v1alpha1.reports.AssemblyInfo.sequencing_tech', index=12,
      number=14, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\"\n\017sequencing-tech\022\017Sequencing tech', json_name='sequencingTech', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='biosample_accession', full_name='ncbi.datasets.v1alpha1.reports.AssemblyInfo.biosample_accession', index=13,
      number=15, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\0308\n\023biosample-accession\022\023BioSample accession2\014SAMN03145444', json_name='biosampleAccession', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=1823,
  serialized_end=3200,
)


_ASSEMBLYSTATS = _descriptor.Descriptor(
  name='AssemblyStats',
  full_name='ncbi.datasets.v1alpha1.reports.AssemblyStats',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='total_number_of_chromosomes', full_name='ncbi.datasets.v1alpha1.reports.AssemblyStats.total_number_of_chromosomes', index=0,
      number=1, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030:\n\033total-number-of-chromosomes\022\033Total number of chromosomes', json_name='totalNumberOfChromosomes', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='total_sequence_length', full_name='ncbi.datasets.v1alpha1.reports.AssemblyStats.total_sequence_length', index=1,
      number=2, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030+\n\022total-sequence-len\022\025Total sequence length', json_name='totalSequenceLength', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='total_ungapped_length', full_name='ncbi.datasets.v1alpha1.reports.AssemblyStats.total_ungapped_length', index=2,
      number=3, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030+\n\022total-ungapped-len\022\025Total ungapped length', json_name='totalUngappedLength', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='number_of_contigs', full_name='ncbi.datasets.v1alpha1.reports.AssemblyStats.number_of_contigs', index=3,
      number=4, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030&\n\021number-of-contigs\022\021Number of contigs', json_name='numberOfContigs', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='contig_n50', full_name='ncbi.datasets.v1alpha1.reports.AssemblyStats.contig_n50', index=4,
      number=5, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\030\n\ncontig-n50\022\nContig N50', json_name='contigN50', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='contig_l50', full_name='ncbi.datasets.v1alpha1.reports.AssemblyStats.contig_l50', index=5,
      number=6, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\030\n\ncontig-l50\022\nContig L50', json_name='contigL50', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='number_of_scaffolds', full_name='ncbi.datasets.v1alpha1.reports.AssemblyStats.number_of_scaffolds', index=6,
      number=7, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030*\n\023number-of-scaffolds\022\023Number of scaffolds', json_name='numberOfScaffolds', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='scaffold_n50', full_name='ncbi.datasets.v1alpha1.reports.AssemblyStats.scaffold_n50', index=7,
      number=8, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\034\n\014scaffold-n50\022\014Scaffold N50', json_name='scaffoldN50', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='scaffold_l50', full_name='ncbi.datasets.v1alpha1.reports.AssemblyStats.scaffold_l50', index=8,
      number=9, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\034\n\014scaffold-l50\022\014Scaffold L50', json_name='scaffoldL50', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='gaps_between_scaffolds_count', full_name='ncbi.datasets.v1alpha1.reports.AssemblyStats.gaps_between_scaffolds_count', index=9,
      number=10, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030<\n\034gaps-between-scaffolds-count\022\034Gaps between scaffolds count', json_name='gapsBetweenScaffoldsCount', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='number_of_component_sequences', full_name='ncbi.datasets.v1alpha1.reports.AssemblyStats.number_of_component_sequences', index=10,
      number=11, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030>\n\035number-of-component-sequences\022\035Number of component sequences', json_name='numberOfComponentSequences', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=3203,
  serialized_end=4259,
)


_ORGANELLEINFO = _descriptor.Descriptor(
  name='OrganelleInfo',
  full_name='ncbi.datasets.v1alpha1.reports.OrganelleInfo',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='assembly_name', full_name='ncbi.datasets.v1alpha1.reports.OrganelleInfo.assembly_name', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='assemblyName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='infraspecific_name', full_name='ncbi.datasets.v1alpha1.reports.OrganelleInfo.infraspecific_name', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='infraspecificName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='bioproject', full_name='ncbi.datasets.v1alpha1.reports.OrganelleInfo.bioproject', index=2,
      number=3, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='bioproject', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='description', full_name='ncbi.datasets.v1alpha1.reports.OrganelleInfo.description', index=3,
      number=4, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='description', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='total_seq_length', full_name='ncbi.datasets.v1alpha1.reports.OrganelleInfo.total_seq_length', index=4,
      number=5, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='totalSeqLength', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='submitter', full_name='ncbi.datasets.v1alpha1.reports.OrganelleInfo.submitter', index=5,
      number=6, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='submitter', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=4262,
  serialized_end=4499,
)


_ANNOTATIONINFO_FEATURECOUNTS = _descriptor.Descriptor(
  name='FeatureCounts',
  full_name='ncbi.datasets.v1alpha1.reports.AnnotationInfo.FeatureCounts',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='gene_counts', full_name='ncbi.datasets.v1alpha1.reports.AnnotationInfo.FeatureCounts.gene_counts', index=0,
      number=1, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='geneCounts', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=4886,
  serialized_end=4993,
)

_ANNOTATIONINFO_GENECOUNTS = _descriptor.Descriptor(
  name='GeneCounts',
  full_name='ncbi.datasets.v1alpha1.reports.AnnotationInfo.GeneCounts',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='total', full_name='ncbi.datasets.v1alpha1.reports.AnnotationInfo.GeneCounts.total', index=0,
      number=1, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030%\n\021gene-counts-total\022\020Gene count total', json_name='total', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='protein_coding', full_name='ncbi.datasets.v1alpha1.reports.AnnotationInfo.GeneCounts.protein_coding', index=1,
      number=2, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\0307\n\032gene-counts-protein-coding\022\031Gene count protein-coding', json_name='proteinCoding', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='non_coding', full_name='ncbi.datasets.v1alpha1.reports.AnnotationInfo.GeneCounts.non_coding', index=2,
      number=3, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030/\n\026gene-counts-non-coding\022\025Gene count non-coding', json_name='nonCoding', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='pseudogene', full_name='ncbi.datasets.v1alpha1.reports.AnnotationInfo.GeneCounts.pseudogene', index=3,
      number=4, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030/\n\026gene-counts-pseudogene\022\025Gene count pseudogene', json_name='pseudogene', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='other', full_name='ncbi.datasets.v1alpha1.reports.AnnotationInfo.GeneCounts.other', index=4,
      number=5, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030%\n\021gene-counts-other\022\020Gene count other', json_name='other', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=4996,
  serialized_end=5407,
)

_ANNOTATIONINFO = _descriptor.Descriptor(
  name='AnnotationInfo',
  full_name='ncbi.datasets.v1alpha1.reports.AnnotationInfo',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='name', full_name='ncbi.datasets.v1alpha1.reports.AnnotationInfo.name', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\035\n\nannot-name\022\017Annotation name', json_name='name', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='source', full_name='ncbi.datasets.v1alpha1.reports.AnnotationInfo.source', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030!\n\014annot-source\022\021Annotation source', json_name='source', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='release_date', full_name='ncbi.datasets.v1alpha1.reports.AnnotationInfo.release_date', index=2,
      number=3, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030-\n\022annot-release-date\022\027Annotation release date', json_name='releaseDate', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='report_url', full_name='ncbi.datasets.v1alpha1.reports.AnnotationInfo.report_url', index=3,
      number=4, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030)\n\020annot-report-url\022\025Annotation report URL', json_name='reportUrl', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='stats', full_name='ncbi.datasets.v1alpha1.reports.AnnotationInfo.stats', index=4,
      number=5, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='stats', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[_ANNOTATIONINFO_FEATURECOUNTS, _ANNOTATIONINFO_GENECOUNTS, ],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=4502,
  serialized_end=5407,
)


_WGSINFO = _descriptor.Descriptor(
  name='WGSInfo',
  full_name='ncbi.datasets.v1alpha1.reports.WGSInfo',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='wgs_project_accession', full_name='ncbi.datasets.v1alpha1.reports.WGSInfo.wgs_project_accession', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030@\n\025wgs-project-accession\022\025WGS project accession2\006AAEX032\010CABHLF01', json_name='wgsProjectAccession', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='master_wgs_url', full_name='ncbi.datasets.v1alpha1.reports.WGSInfo.master_wgs_url', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030G\n\007wgs-url\022\007WGS URL23https://www.ncbi.nlm.nih.gov/nuccore/AAEX00000000.3', json_name='masterWgsUrl', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='wgs_contigs_url', full_name='ncbi.datasets.v1alpha1.reports.WGSInfo.wgs_contigs_url', index=2,
      number=3, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030R\n\017wgs-contigs-url\022\017WGS contigs URL2.https://www.ncbi.nlm.nih.gov/Traces/wgs/AAEX03', json_name='wgsContigsUrl', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=5410,
  serialized_end=5784,
)

_ASSEMBLYDATAREPORT.fields_by_name['assembly_info'].message_type = _ASSEMBLYINFO
_ASSEMBLYDATAREPORT.fields_by_name['assembly_stats'].message_type = _ASSEMBLYSTATS
_ASSEMBLYDATAREPORT.fields_by_name['organelle_info'].message_type = _ORGANELLEINFO
_ASSEMBLYDATAREPORT.fields_by_name['annotation_info'].message_type = _ANNOTATIONINFO
_ASSEMBLYDATAREPORT.fields_by_name['wgs_info'].message_type = _WGSINFO
_BIOPROJECTLINEAGE.fields_by_name['bioprojects'].message_type = _BIOPROJECT
_ASSEMBLYINFO.fields_by_name['bioproject_lineage'].message_type = _BIOPROJECTLINEAGE
_ANNOTATIONINFO_FEATURECOUNTS.fields_by_name['gene_counts'].message_type = _ANNOTATIONINFO_GENECOUNTS
_ANNOTATIONINFO_FEATURECOUNTS.containing_type = _ANNOTATIONINFO
_ANNOTATIONINFO_GENECOUNTS.containing_type = _ANNOTATIONINFO
_ANNOTATIONINFO.fields_by_name['stats'].message_type = _ANNOTATIONINFO_FEATURECOUNTS
DESCRIPTOR.message_types_by_name['AssemblyDataReport'] = _ASSEMBLYDATAREPORT
DESCRIPTOR.message_types_by_name['BioProject'] = _BIOPROJECT
DESCRIPTOR.message_types_by_name['BioProjectLineage'] = _BIOPROJECTLINEAGE
DESCRIPTOR.message_types_by_name['AssemblyInfo'] = _ASSEMBLYINFO
DESCRIPTOR.message_types_by_name['AssemblyStats'] = _ASSEMBLYSTATS
DESCRIPTOR.message_types_by_name['OrganelleInfo'] = _ORGANELLEINFO
DESCRIPTOR.message_types_by_name['AnnotationInfo'] = _ANNOTATIONINFO
DESCRIPTOR.message_types_by_name['WGSInfo'] = _WGSINFO
_sym_db.RegisterFileDescriptor(DESCRIPTOR)

AssemblyDataReport = _reflection.GeneratedProtocolMessageType('AssemblyDataReport', (_message.Message,), {
  'DESCRIPTOR' : _ASSEMBLYDATAREPORT,
  '__module__' : 'ncbi.datasets.v1alpha1.reports.assembly_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.AssemblyDataReport)
  })
_sym_db.RegisterMessage(AssemblyDataReport)

BioProject = _reflection.GeneratedProtocolMessageType('BioProject', (_message.Message,), {
  'DESCRIPTOR' : _BIOPROJECT,
  '__module__' : 'ncbi.datasets.v1alpha1.reports.assembly_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.BioProject)
  })
_sym_db.RegisterMessage(BioProject)

BioProjectLineage = _reflection.GeneratedProtocolMessageType('BioProjectLineage', (_message.Message,), {
  'DESCRIPTOR' : _BIOPROJECTLINEAGE,
  '__module__' : 'ncbi.datasets.v1alpha1.reports.assembly_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.BioProjectLineage)
  })
_sym_db.RegisterMessage(BioProjectLineage)

AssemblyInfo = _reflection.GeneratedProtocolMessageType('AssemblyInfo', (_message.Message,), {
  'DESCRIPTOR' : _ASSEMBLYINFO,
  '__module__' : 'ncbi.datasets.v1alpha1.reports.assembly_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.AssemblyInfo)
  })
_sym_db.RegisterMessage(AssemblyInfo)

AssemblyStats = _reflection.GeneratedProtocolMessageType('AssemblyStats', (_message.Message,), {
  'DESCRIPTOR' : _ASSEMBLYSTATS,
  '__module__' : 'ncbi.datasets.v1alpha1.reports.assembly_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.AssemblyStats)
  })
_sym_db.RegisterMessage(AssemblyStats)

OrganelleInfo = _reflection.GeneratedProtocolMessageType('OrganelleInfo', (_message.Message,), {
  'DESCRIPTOR' : _ORGANELLEINFO,
  '__module__' : 'ncbi.datasets.v1alpha1.reports.assembly_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.OrganelleInfo)
  })
_sym_db.RegisterMessage(OrganelleInfo)

AnnotationInfo = _reflection.GeneratedProtocolMessageType('AnnotationInfo', (_message.Message,), {

  'FeatureCounts' : _reflection.GeneratedProtocolMessageType('FeatureCounts', (_message.Message,), {
    'DESCRIPTOR' : _ANNOTATIONINFO_FEATURECOUNTS,
    '__module__' : 'ncbi.datasets.v1alpha1.reports.assembly_pb2'
    # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.AnnotationInfo.FeatureCounts)
    })
  ,

  'GeneCounts' : _reflection.GeneratedProtocolMessageType('GeneCounts', (_message.Message,), {
    'DESCRIPTOR' : _ANNOTATIONINFO_GENECOUNTS,
    '__module__' : 'ncbi.datasets.v1alpha1.reports.assembly_pb2'
    # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.AnnotationInfo.GeneCounts)
    })
  ,
  'DESCRIPTOR' : _ANNOTATIONINFO,
  '__module__' : 'ncbi.datasets.v1alpha1.reports.assembly_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.AnnotationInfo)
  })
_sym_db.RegisterMessage(AnnotationInfo)
_sym_db.RegisterMessage(AnnotationInfo.FeatureCounts)
_sym_db.RegisterMessage(AnnotationInfo.GeneCounts)

WGSInfo = _reflection.GeneratedProtocolMessageType('WGSInfo', (_message.Message,), {
  'DESCRIPTOR' : _WGSINFO,
  '__module__' : 'ncbi.datasets.v1alpha1.reports.assembly_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.WGSInfo)
  })
_sym_db.RegisterMessage(WGSInfo)


DESCRIPTOR._options = None
_ASSEMBLYDATAREPORT.fields_by_name['common_name']._options = None
_ASSEMBLYDATAREPORT.fields_by_name['organism_name']._options = None
_ASSEMBLYDATAREPORT.fields_by_name['breed']._options = None
_ASSEMBLYDATAREPORT.fields_by_name['cultivar']._options = None
_ASSEMBLYDATAREPORT.fields_by_name['ecotype']._options = None
_ASSEMBLYDATAREPORT.fields_by_name['isolate']._options = None
_ASSEMBLYDATAREPORT.fields_by_name['sex']._options = None
_ASSEMBLYDATAREPORT.fields_by_name['strain']._options = None
_ASSEMBLYDATAREPORT.fields_by_name['tax_id']._options = None
_BIOPROJECT.fields_by_name['accession']._options = None
_BIOPROJECT.fields_by_name['title']._options = None
_BIOPROJECT.fields_by_name['parent_accession']._options = None
_BIOPROJECT.fields_by_name['parent_accessions']._options = None
_ASSEMBLYINFO.fields_by_name['assembly_level']._options = None
_ASSEMBLYINFO.fields_by_name['assembly_name']._options = None
_ASSEMBLYINFO.fields_by_name['assembly_type']._options = None
_ASSEMBLYINFO.fields_by_name['submission_date']._options = None
_ASSEMBLYINFO.fields_by_name['description']._options = None
_ASSEMBLYINFO.fields_by_name['genbank_assm_accession']._options = None
_ASSEMBLYINFO.fields_by_name['submitter']._options = None
_ASSEMBLYINFO.fields_by_name['refseq_category']._options = None
_ASSEMBLYINFO.fields_by_name['refseq_assm_accession']._options = None
_ASSEMBLYINFO.fields_by_name['ucsc_assm_name']._options = None
_ASSEMBLYINFO.fields_by_name['linked_assembly']._options = None
_ASSEMBLYINFO.fields_by_name['sequencing_tech']._options = None
_ASSEMBLYINFO.fields_by_name['biosample_accession']._options = None
_ASSEMBLYSTATS.fields_by_name['total_number_of_chromosomes']._options = None
_ASSEMBLYSTATS.fields_by_name['total_sequence_length']._options = None
_ASSEMBLYSTATS.fields_by_name['total_ungapped_length']._options = None
_ASSEMBLYSTATS.fields_by_name['number_of_contigs']._options = None
_ASSEMBLYSTATS.fields_by_name['contig_n50']._options = None
_ASSEMBLYSTATS.fields_by_name['contig_l50']._options = None
_ASSEMBLYSTATS.fields_by_name['number_of_scaffolds']._options = None
_ASSEMBLYSTATS.fields_by_name['scaffold_n50']._options = None
_ASSEMBLYSTATS.fields_by_name['scaffold_l50']._options = None
_ASSEMBLYSTATS.fields_by_name['gaps_between_scaffolds_count']._options = None
_ASSEMBLYSTATS.fields_by_name['number_of_component_sequences']._options = None
_ANNOTATIONINFO_GENECOUNTS.fields_by_name['total']._options = None
_ANNOTATIONINFO_GENECOUNTS.fields_by_name['protein_coding']._options = None
_ANNOTATIONINFO_GENECOUNTS.fields_by_name['non_coding']._options = None
_ANNOTATIONINFO_GENECOUNTS.fields_by_name['pseudogene']._options = None
_ANNOTATIONINFO_GENECOUNTS.fields_by_name['other']._options = None
_ANNOTATIONINFO.fields_by_name['name']._options = None
_ANNOTATIONINFO.fields_by_name['source']._options = None
_ANNOTATIONINFO.fields_by_name['release_date']._options = None
_ANNOTATIONINFO.fields_by_name['report_url']._options = None
_WGSINFO.fields_by_name['wgs_project_accession']._options = None
_WGSINFO.fields_by_name['master_wgs_url']._options = None
_WGSINFO.fields_by_name['wgs_contigs_url']._options = None
# @@protoc_insertion_point(module_scope)
