# -*- coding: utf-8 -*-
# Generated by the protocol buffer compiler.  DO NOT EDIT!
# source: ncbi/datasets/v1alpha1/reports/gene.proto

from google.protobuf import descriptor as _descriptor
from google.protobuf import message as _message
from google.protobuf import reflection as _reflection
from google.protobuf import symbol_database as _symbol_database
# @@protoc_insertion_point(imports)

_sym_db = _symbol_database.Default()


from ncbi.datasets.v1alpha1.reports import common_pb2 as ncbi_dot_datasets_dot_v1alpha1_dot_reports_dot_common__pb2
from ncbi.datasets.v1alpha1.reports import options_pb2 as ncbi_dot_datasets_dot_v1alpha1_dot_reports_dot_options__pb2


DESCRIPTOR = _descriptor.FileDescriptor(
  name='ncbi/datasets/v1alpha1/reports/gene.proto',
  package='ncbi.datasets.v1alpha1.reports',
  syntax='proto3',
  serialized_options=b'Z\036ncbi/datasets/v1alpha1/reports\370\001\001\302\363\030\270\010\n\013Gene Report\022\016GeneDescriptor\032:<p>The downloaded gene package contains a gene data report\032^in <a href=\"https://jsonlines.readthedocs.io/\">JSON lines</a> format at the following location\032iin your downloaded data package: <em>ncbi_dataset/data/data_report.jsonl</em>. <br> Each line of the gene\032ldata report file is a hierarchichal object serialized as JSON and each line represents a single gene record.\032\210\001The tables below represent the schema. The outermost object is a <em>GeneDescriptor</em> object. Its fields are documented in the tables\032\nbelow.</p>\032O<p>The second column in each table documents the mnemonic string used by the <a\032Phref=\"https://api.ncbi.nlm.nih.gov/datasets/docs/cli/dataformat/\">dataformat</a>\032Xcommand-line tool to transform gene data reports from JSON lines to tabular formats.</p>\032k<p>Check the <a href=\"https://api.ncbi.nlm.nih.gov/datasets/docs/data-report-schemas-overview/\">Data Report\032DOverview</a> documentation for examples of using this table with the\032]<a href=\"https://api.ncbi.nlm.nih.gov/datasets/docs/cli/dataformat/\">dataformat</a> tool.</p>',
  serialized_pb=b'\n)ncbi/datasets/v1alpha1/reports/gene.proto\x12\x1encbi.datasets.v1alpha1.reports\x1a+ncbi/datasets/v1alpha1/reports/common.proto\x1a,ncbi/datasets/v1alpha1/reports/options.proto\"h\n\rMaturePeptide\x12+\n\x11\x61\x63\x63\x65ssion_version\x18\x01 \x01(\tR\x10\x61\x63\x63\x65ssionVersion\x12\x12\n\x04name\x18\x02 \x01(\tR\x04name\x12\x16\n\x06length\x18\x03 \x01(\rR\x06length\"\x8e\x04\n\x07Protein\x12g\n\x11\x61\x63\x63\x65ssion_version\x18\x01 \x01(\tB:\xc2\xf3\x18\x36\n\x11protein-accession\x12\x11Protein accession2\x0eNP_001296812.1R\x10\x61\x63\x63\x65ssionVersion\x12\x42\n\x04name\x18\x02 \x01(\tB.\xc2\xf3\x18*\n\x0cprotein-name\x12\x0cProtein name2\x0cprotein ALEXR\x04name\x12\x41\n\x06length\x18\x03 \x01(\rB)\xc2\xf3\x18%\n\x0eprotein-length\x12\x0eProtein length2\x03\x36\x32\x36R\x06length\x12W\n\x0cisoform_name\x18\x05 \x01(\tB4\xc2\xf3\x18\x30\n\x0fprotein-isoform\x12\x0fProtein isoform2\x0cisoform AlexR\x0bisoformName\x12\x62\n\x0f\x65nsembl_protein\x18\x06 \x01(\tB9\xc2\xf3\x18\x35\n\x0f\x65nsembl-protein\x12\x0f\x45nsembl protein2\x11\x45NSP00000302237.3R\x0e\x65nsemblProtein\x12V\n\x0fmature_peptides\x18\x04 \x03(\x0b\x32-.ncbi.datasets.v1alpha1.reports.MaturePeptideR\x0ematurePeptides\"\xfb\x01\n\x0fGenomicLocation\x12:\n\x19genomic_accession_version\x18\x01 \x01(\tR\x17genomicAccessionVersion\x12#\n\rsequence_name\x18\x02 \x01(\tR\x0csequenceName\x12J\n\rgenomic_range\x18\x03 \x01(\x0b\x32%.ncbi.datasets.v1alpha1.reports.RangeR\x0cgenomicRange\x12;\n\x05\x65xons\x18\x04 \x03(\x0b\x32%.ncbi.datasets.v1alpha1.reports.RangeR\x05\x65xons\"\xd9\x07\n\nTranscript\x12+\n\x11\x61\x63\x63\x65ssion_version\x18\x01 \x01(\tR\x10\x61\x63\x63\x65ssionVersion\x12Q\n\x04name\x18\x02 \x01(\tB=\xc2\xf3\x18\x39\n\x0ftranscript-name\x12\x0fTranscript Name2\x15transcript variant 12R\x04name\x12H\n\x06length\x18\x03 \x01(\rB0\xc2\xf3\x18,\n\x11transcript-length\x12\x11Transcript Length2\x04\x33\x31\x38\x30R\x06length\x12T\n\rgenomic_range\x18\x04 \x01(\x0b\x32+.ncbi.datasets.v1alpha1.reports.SeqRangeSetB\x02\x18\x01R\x0cgenomicRange\x12\x45\n\x05\x65xons\x18\x05 \x01(\x0b\x32+.ncbi.datasets.v1alpha1.reports.SeqRangeSetB\x02\x18\x01R\x05\x65xons\x12\x41\n\x03\x63\x64s\x18\x06 \x01(\x0b\x32+.ncbi.datasets.v1alpha1.reports.SeqRangeSetB\x02\x18\x01R\x03\x63\x64s\x12\x83\x01\n\x11genomic_locations\x18\n \x03(\x0b\x32/.ncbi.datasets.v1alpha1.reports.GenomicLocationB%\xc2\xf3\x18!\n\x0fgenomic-ranges-\x12\x0eGenomic Range R\x10genomicLocations\x12n\n\x12\x65nsembl_transcript\x18\x08 \x01(\tB?\xc2\xf3\x18;\n\x12\x65nsembl-transcript\x12\x12\x45nsembl Transcript2\x11\x45NST00000306120.3R\x11\x65nsemblTranscript\x12\x41\n\x07protein\x18\x07 \x01(\x0b\x32\'.ncbi.datasets.v1alpha1.reports.ProteinR\x07protein\x12u\n\x04type\x18\t \x01(\x0e\x32\x39.ncbi.datasets.v1alpha1.reports.Transcript.TranscriptTypeB&\xc2\xf3\x18\"\n\x0ftranscript-type\x12\x0fTranscript TypeR\x04type\"q\n\x0eTranscriptType\x12\x0b\n\x07UNKNOWN\x10\x00\x12\x12\n\x0ePROTEIN_CODING\x10\x01\x12\x0e\n\nNON_CODING\x10\x02\x12\x18\n\x14PROTEIN_CODING_MODEL\x10\x03\x12\x14\n\x10NON_CODING_MODEL\x10\x04\"\xc6\x02\n\rGenomicRegion\x12J\n\ngene_range\x18\x01 \x01(\x0b\x32+.ncbi.datasets.v1alpha1.reports.SeqRangeSetR\tgeneRange\x12\x83\x01\n\x04type\x18\x02 \x01(\x0e\x32?.ncbi.datasets.v1alpha1.reports.GenomicRegion.GenomicRegionTypeB.\xc2\xf3\x18*\n\x13genomic-region-type\x12\x13Genomic Region TypeR\x04type\"c\n\x11GenomicRegionType\x12\x0b\n\x07UNKNOWN\x10\x00\x12\x0f\n\x0bREFSEQ_GENE\x10\x01\x12\x0e\n\nPSEUDOGENE\x10\x02\x12\x15\n\x11\x42IOLOGICAL_REGION\x10\x03\x12\t\n\x05OTHER\x10\x04\"\xb7\x01\n\nAnnotation\x12!\n\x0crelease_name\x18\x01 \x01(\tR\x0breleaseName\x12!\n\x0crelease_date\x18\x02 \x01(\tR\x0breleaseDate\x12\x63\n\x13\x61ssemblies_in_scope\x18\x03 \x03(\x0b\x32\x33.ncbi.datasets.v1alpha1.reports.AnnotatedAssembliesR\x11\x61ssembliesInScope\"G\n\x13\x41nnotatedAssemblies\x12\x1c\n\taccession\x18\x01 \x01(\tR\taccession\x12\x12\n\x04name\x18\x02 \x01(\tR\x04name\"\xd1\x0f\n\x0eGeneDescriptor\x12\x35\n\x07gene_id\x18\x01 \x01(\x04\x42\x1c\xc2\xf3\x18\x18\n\x07gene-id\x12\x07Gene ID2\x04\x32\x37\x37\x38R\x06geneId\x12\x32\n\x06symbol\x18\x02 \x01(\tB\x1a\xc2\xf3\x18\x16\n\x06symbol\x12\x06Symbol2\x04GNASR\x06symbol\x12T\n\x0b\x64\x65scription\x18\x03 \x01(\tB2\xc2\xf3\x18.\n\x0b\x64\x65scription\x12\x0b\x44\x65scription2\x12GNAS complex locusR\x0b\x64\x65scription\x12\x37\n\x06tax_id\x18\x04 \x01(\x04\x42 \xc2\xf3\x18\x1c\n\x06tax-id\x12\x0cTaxonomic ID2\x04\x39\x36\x30\x36R\x05taxId\x12\x46\n\x07taxname\x18\x05 \x01(\tB,\xc2\xf3\x18(\n\x08tax-name\x12\x0eTaxonomic Name2\x0cHomo sapiensR\x07taxname\x12\x46\n\x0b\x63ommon_name\x18\x0c \x01(\tB%\xc2\xf3\x18!\n\x0b\x63ommon-name\x12\x0b\x43ommon Name2\x05humanR\ncommonName\x12g\n\x04type\x18\x06 \x01(\x0e\x32\x37.ncbi.datasets.v1alpha1.reports.GeneDescriptor.GeneTypeB\x1a\xc2\xf3\x18\x16\n\tgene-type\x12\tGene typeR\x04type\x12k\n\x08rna_type\x18\x0b \x01(\x0e\x32\x36.ncbi.datasets.v1alpha1.reports.GeneDescriptor.RnaTypeB\x18\xc2\xf3\x18\x14\n\x08rna-type\x12\x08RNA typeR\x07rnaType\x12w\n\x0borientation\x18\x07 \x01(\x0e\x32+.ncbi.datasets.v1alpha1.reports.OrientationB(\xc2\xf3\x18$\n\x10gene-orientation\x12\x10Gene OrientationR\x0borientation\x12y\n\x0egenomic_ranges\x18\x08 \x03(\x0b\x32+.ncbi.datasets.v1alpha1.reports.SeqRangeSetB%\xc2\xf3\x18!\n\x0fgenomic-ranges-\x12\x0eGenomic Range R\rgenomicRanges\x12^\n\x13reference_standards\x18\x15 \x03(\x0b\x32-.ncbi.datasets.v1alpha1.reports.GenomicRegionR\x12referenceStandards\x12V\n\x0fgenomic_regions\x18\x16 \x03(\x0b\x32-.ncbi.datasets.v1alpha1.reports.GenomicRegionR\x0egenomicRegions\x12m\n\x0btranscripts\x18\t \x03(\x0b\x32*.ncbi.datasets.v1alpha1.reports.TranscriptB\x1f\xc2\xf3\x18\x1b\n\x0ctranscripts-\x12\x0bTranscript R\x0btranscripts\x12\x43\n\x08proteins\x18\n \x03(\x0b\x32\'.ncbi.datasets.v1alpha1.reports.ProteinR\x08proteins\x12\"\n\nchromosome\x18\r \x01(\tB\x02\x18\x01R\nchromosome\x12H\n\x0b\x63hromosomes\x18\x0e \x03(\tB&\xc2\xf3\x18\"\n\x0b\x63hromosomes\x12\x0b\x43hromosomes2\x01\x31\x32\x03X,YR\x0b\x63hromosomes\x12l\n\x16nomenclature_authority\x18\x0f \x01(\x0b\x32\x35.ncbi.datasets.v1alpha1.reports.NomenclatureAuthorityR\x15nomenclatureAuthority\x12\x32\n\x15swiss_prot_accessions\x18\x10 \x03(\tR\x13swissProtAccessions\x12(\n\x10\x65nsembl_gene_ids\x18\x11 \x03(\tR\x0e\x65nsemblGeneIds\x12\x19\n\x08omim_ids\x18\x12 \x03(\tR\x07omimIds\x12\x1a\n\x08synonyms\x18\x13 \x03(\tR\x08synonyms\x12R\n\x10replaced_gene_id\x18\x14 \x01(\x04\x42(\xc2\xf3\x18$\n\x10replaced-gene-id\x12\x10Replaced Gene IDR\x0ereplacedGeneId\x12L\n\x0b\x61nnotations\x18\x17 \x03(\x0b\x32*.ncbi.datasets.v1alpha1.reports.AnnotationR\x0b\x61nnotations\"\xb8\x01\n\x08GeneType\x12\x0b\n\x07UNKNOWN\x10\x00\x12\x08\n\x04tRNA\x10\x01\x12\x08\n\x04rRNA\x10\x02\x12\t\n\x05snRNA\x10\x03\x12\t\n\x05scRNA\x10\x04\x12\n\n\x06snoRNA\x10\x05\x12\x12\n\x0ePROTEIN_CODING\x10\x06\x12\n\n\x06PSEUDO\x10\x07\x12\x0e\n\nTRANSPOSON\x10\x08\x12\x0b\n\x07miscRNA\x10\t\x12\t\n\x05ncRNA\x10\n\x12\x15\n\x11\x42IOLOGICAL_REGION\x10\x0b\x12\n\n\x05OTHER\x10\xff\x01\"1\n\x07RnaType\x12\x0f\n\x0brna_UNKNOWN\x10\x00\x12\n\n\x06premsg\x10\x01\x12\t\n\x05tmRna\x10\x02\"\xb4\x01\n\x15NomenclatureAuthority\x12P\n\tauthority\x18\x01 \x01(\tB2\xc2\xf3\x18.\n\x0ename-authority\x12\x16Nomenclature Authority2\x04HGNCR\tauthority\x12I\n\nidentifier\x18\x02 \x01(\tB)\xc2\xf3\x18%\n\x07name-id\x12\x0fNomenclature ID2\tHGNC:4392R\nidentifier\"W\n\x0fGeneDescriptors\x12\x44\n\x05genes\x18\x01 \x03(\x0b\x32..ncbi.datasets.v1alpha1.reports.GeneDescriptorR\x05genesB\xe0\x08Z\x1encbi/datasets/v1alpha1/reports\xf8\x01\x01\xc2\xf3\x18\xb8\x08\n\x0bGene Report\x12\x0eGeneDescriptor\x1a:<p>The downloaded gene package contains a gene data report\x1a^in <a href=\"https://jsonlines.readthedocs.io/\">JSON lines</a> format at the following location\x1aiin your downloaded data package: <em>ncbi_dataset/data/data_report.jsonl</em>. <br> Each line of the gene\x1aldata report file is a hierarchichal object serialized as JSON and each line represents a single gene record.\x1a\x88\x01The tables below represent the schema. The outermost object is a <em>GeneDescriptor</em> object. Its fields are documented in the tables\x1a\nbelow.</p>\x1aO<p>The second column in each table documents the mnemonic string used by the <a\x1aPhref=\"https://api.ncbi.nlm.nih.gov/datasets/docs/cli/dataformat/\">dataformat</a>\x1aXcommand-line tool to transform gene data reports from JSON lines to tabular formats.</p>\x1ak<p>Check the <a href=\"https://api.ncbi.nlm.nih.gov/datasets/docs/data-report-schemas-overview/\">Data Report\x1a\x44Overview</a> documentation for examples of using this table with the\x1a]<a href=\"https://api.ncbi.nlm.nih.gov/datasets/docs/cli/dataformat/\">dataformat</a> tool.</p>b\x06proto3'
  ,
  dependencies=[ncbi_dot_datasets_dot_v1alpha1_dot_reports_dot_common__pb2.DESCRIPTOR,ncbi_dot_datasets_dot_v1alpha1_dot_reports_dot_options__pb2.DESCRIPTOR,])



_TRANSCRIPT_TRANSCRIPTTYPE = _descriptor.EnumDescriptor(
  name='TranscriptType',
  full_name='ncbi.datasets.v1alpha1.reports.Transcript.TranscriptType',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='UNKNOWN', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='PROTEIN_CODING', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='NON_CODING', index=2, number=2,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='PROTEIN_CODING_MODEL', index=3, number=3,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='NON_CODING_MODEL', index=4, number=4,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=1930,
  serialized_end=2043,
)
_sym_db.RegisterEnumDescriptor(_TRANSCRIPT_TRANSCRIPTTYPE)

_GENOMICREGION_GENOMICREGIONTYPE = _descriptor.EnumDescriptor(
  name='GenomicRegionType',
  full_name='ncbi.datasets.v1alpha1.reports.GenomicRegion.GenomicRegionType',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='UNKNOWN', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='REFSEQ_GENE', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='PSEUDOGENE', index=2, number=2,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='BIOLOGICAL_REGION', index=3, number=3,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='OTHER', index=4, number=4,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=2273,
  serialized_end=2372,
)
_sym_db.RegisterEnumDescriptor(_GENOMICREGION_GENOMICREGIONTYPE)

_GENEDESCRIPTOR_GENETYPE = _descriptor.EnumDescriptor(
  name='GeneType',
  full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.GeneType',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='UNKNOWN', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='tRNA', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='rRNA', index=2, number=2,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='snRNA', index=3, number=3,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='scRNA', index=4, number=4,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='snoRNA', index=5, number=5,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='PROTEIN_CODING', index=6, number=6,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='PSEUDO', index=7, number=7,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='TRANSPOSON', index=8, number=8,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='miscRNA', index=9, number=9,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='ncRNA', index=10, number=10,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='BIOLOGICAL_REGION', index=11, number=11,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='OTHER', index=12, number=255,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=4400,
  serialized_end=4584,
)
_sym_db.RegisterEnumDescriptor(_GENEDESCRIPTOR_GENETYPE)

_GENEDESCRIPTOR_RNATYPE = _descriptor.EnumDescriptor(
  name='RnaType',
  full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.RnaType',
  filename=None,
  file=DESCRIPTOR,
  values=[
    _descriptor.EnumValueDescriptor(
      name='rna_UNKNOWN', index=0, number=0,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='premsg', index=1, number=1,
      serialized_options=None,
      type=None),
    _descriptor.EnumValueDescriptor(
      name='tmRna', index=2, number=2,
      serialized_options=None,
      type=None),
  ],
  containing_type=None,
  serialized_options=None,
  serialized_start=4586,
  serialized_end=4635,
)
_sym_db.RegisterEnumDescriptor(_GENEDESCRIPTOR_RNATYPE)


_MATUREPEPTIDE = _descriptor.Descriptor(
  name='MaturePeptide',
  full_name='ncbi.datasets.v1alpha1.reports.MaturePeptide',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='accession_version', full_name='ncbi.datasets.v1alpha1.reports.MaturePeptide.accession_version', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='accessionVersion', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='name', full_name='ncbi.datasets.v1alpha1.reports.MaturePeptide.name', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='name', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='length', full_name='ncbi.datasets.v1alpha1.reports.MaturePeptide.length', index=2,
      number=3, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='length', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=168,
  serialized_end=272,
)


_PROTEIN = _descriptor.Descriptor(
  name='Protein',
  full_name='ncbi.datasets.v1alpha1.reports.Protein',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='accession_version', full_name='ncbi.datasets.v1alpha1.reports.Protein.accession_version', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\0306\n\021protein-accession\022\021Protein accession2\016NP_001296812.1', json_name='accessionVersion', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='name', full_name='ncbi.datasets.v1alpha1.reports.Protein.name', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030*\n\014protein-name\022\014Protein name2\014protein ALEX', json_name='name', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='length', full_name='ncbi.datasets.v1alpha1.reports.Protein.length', index=2,
      number=3, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030%\n\016protein-length\022\016Protein length2\003626', json_name='length', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='isoform_name', full_name='ncbi.datasets.v1alpha1.reports.Protein.isoform_name', index=3,
      number=5, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\0300\n\017protein-isoform\022\017Protein isoform2\014isoform Alex', json_name='isoformName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='ensembl_protein', full_name='ncbi.datasets.v1alpha1.reports.Protein.ensembl_protein', index=4,
      number=6, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\0305\n\017ensembl-protein\022\017Ensembl protein2\021ENSP00000302237.3', json_name='ensemblProtein', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='mature_peptides', full_name='ncbi.datasets.v1alpha1.reports.Protein.mature_peptides', index=5,
      number=4, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='maturePeptides', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=275,
  serialized_end=801,
)


_GENOMICLOCATION = _descriptor.Descriptor(
  name='GenomicLocation',
  full_name='ncbi.datasets.v1alpha1.reports.GenomicLocation',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='genomic_accession_version', full_name='ncbi.datasets.v1alpha1.reports.GenomicLocation.genomic_accession_version', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='genomicAccessionVersion', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='sequence_name', full_name='ncbi.datasets.v1alpha1.reports.GenomicLocation.sequence_name', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='sequenceName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='genomic_range', full_name='ncbi.datasets.v1alpha1.reports.GenomicLocation.genomic_range', index=2,
      number=3, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='genomicRange', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='exons', full_name='ncbi.datasets.v1alpha1.reports.GenomicLocation.exons', index=3,
      number=4, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='exons', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=804,
  serialized_end=1055,
)


_TRANSCRIPT = _descriptor.Descriptor(
  name='Transcript',
  full_name='ncbi.datasets.v1alpha1.reports.Transcript',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='accession_version', full_name='ncbi.datasets.v1alpha1.reports.Transcript.accession_version', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='accessionVersion', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='name', full_name='ncbi.datasets.v1alpha1.reports.Transcript.name', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\0309\n\017transcript-name\022\017Transcript Name2\025transcript variant 12', json_name='name', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='length', full_name='ncbi.datasets.v1alpha1.reports.Transcript.length', index=2,
      number=3, type=13, cpp_type=3, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030,\n\021transcript-length\022\021Transcript Length2\0043180', json_name='length', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='genomic_range', full_name='ncbi.datasets.v1alpha1.reports.Transcript.genomic_range', index=3,
      number=4, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='genomicRange', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='exons', full_name='ncbi.datasets.v1alpha1.reports.Transcript.exons', index=4,
      number=5, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='exons', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='cds', full_name='ncbi.datasets.v1alpha1.reports.Transcript.cds', index=5,
      number=6, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='cds', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='genomic_locations', full_name='ncbi.datasets.v1alpha1.reports.Transcript.genomic_locations', index=6,
      number=10, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030!\n\017genomic-ranges-\022\016Genomic Range ', json_name='genomicLocations', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='ensembl_transcript', full_name='ncbi.datasets.v1alpha1.reports.Transcript.ensembl_transcript', index=7,
      number=8, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030;\n\022ensembl-transcript\022\022Ensembl Transcript2\021ENST00000306120.3', json_name='ensemblTranscript', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='protein', full_name='ncbi.datasets.v1alpha1.reports.Transcript.protein', index=8,
      number=7, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='protein', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='type', full_name='ncbi.datasets.v1alpha1.reports.Transcript.type', index=9,
      number=9, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\"\n\017transcript-type\022\017Transcript Type', json_name='type', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
    _TRANSCRIPT_TRANSCRIPTTYPE,
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=1058,
  serialized_end=2043,
)


_GENOMICREGION = _descriptor.Descriptor(
  name='GenomicRegion',
  full_name='ncbi.datasets.v1alpha1.reports.GenomicRegion',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='gene_range', full_name='ncbi.datasets.v1alpha1.reports.GenomicRegion.gene_range', index=0,
      number=1, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='geneRange', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='type', full_name='ncbi.datasets.v1alpha1.reports.GenomicRegion.type', index=1,
      number=2, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030*\n\023genomic-region-type\022\023Genomic Region Type', json_name='type', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
    _GENOMICREGION_GENOMICREGIONTYPE,
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=2046,
  serialized_end=2372,
)


_ANNOTATION = _descriptor.Descriptor(
  name='Annotation',
  full_name='ncbi.datasets.v1alpha1.reports.Annotation',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='release_name', full_name='ncbi.datasets.v1alpha1.reports.Annotation.release_name', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='releaseName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='release_date', full_name='ncbi.datasets.v1alpha1.reports.Annotation.release_date', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='releaseDate', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='assemblies_in_scope', full_name='ncbi.datasets.v1alpha1.reports.Annotation.assemblies_in_scope', index=2,
      number=3, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='assembliesInScope', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=2375,
  serialized_end=2558,
)


_ANNOTATEDASSEMBLIES = _descriptor.Descriptor(
  name='AnnotatedAssemblies',
  full_name='ncbi.datasets.v1alpha1.reports.AnnotatedAssemblies',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='accession', full_name='ncbi.datasets.v1alpha1.reports.AnnotatedAssemblies.accession', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='accession', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='name', full_name='ncbi.datasets.v1alpha1.reports.AnnotatedAssemblies.name', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='name', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=2560,
  serialized_end=2631,
)


_GENEDESCRIPTOR = _descriptor.Descriptor(
  name='GeneDescriptor',
  full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='gene_id', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.gene_id', index=0,
      number=1, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\030\n\007gene-id\022\007Gene ID2\0042778', json_name='geneId', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='symbol', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.symbol', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\026\n\006symbol\022\006Symbol2\004GNAS', json_name='symbol', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='description', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.description', index=2,
      number=3, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030.\n\013description\022\013Description2\022GNAS complex locus', json_name='description', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='tax_id', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.tax_id', index=3,
      number=4, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\034\n\006tax-id\022\014Taxonomic ID2\0049606', json_name='taxId', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='taxname', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.taxname', index=4,
      number=5, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030(\n\010tax-name\022\016Taxonomic Name2\014Homo sapiens', json_name='taxname', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='common_name', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.common_name', index=5,
      number=12, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030!\n\013common-name\022\013Common Name2\005human', json_name='commonName', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='type', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.type', index=6,
      number=6, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\026\n\tgene-type\022\tGene type', json_name='type', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='rna_type', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.rna_type', index=7,
      number=11, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\024\n\010rna-type\022\010RNA type', json_name='rnaType', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='orientation', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.orientation', index=8,
      number=7, type=14, cpp_type=8, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030$\n\020gene-orientation\022\020Gene Orientation', json_name='orientation', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='genomic_ranges', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.genomic_ranges', index=9,
      number=8, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030!\n\017genomic-ranges-\022\016Genomic Range ', json_name='genomicRanges', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='reference_standards', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.reference_standards', index=10,
      number=21, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='referenceStandards', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='genomic_regions', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.genomic_regions', index=11,
      number=22, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='genomicRegions', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='transcripts', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.transcripts', index=12,
      number=9, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\033\n\014transcripts-\022\013Transcript ', json_name='transcripts', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='proteins', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.proteins', index=13,
      number=10, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='proteins', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='chromosome', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.chromosome', index=14,
      number=13, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\030\001', json_name='chromosome', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='chromosomes', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.chromosomes', index=15,
      number=14, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030\"\n\013chromosomes\022\013Chromosomes2\00112\003X,Y', json_name='chromosomes', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='nomenclature_authority', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.nomenclature_authority', index=16,
      number=15, type=11, cpp_type=10, label=1,
      has_default_value=False, default_value=None,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='nomenclatureAuthority', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='swiss_prot_accessions', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.swiss_prot_accessions', index=17,
      number=16, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='swissProtAccessions', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='ensembl_gene_ids', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.ensembl_gene_ids', index=18,
      number=17, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='ensemblGeneIds', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='omim_ids', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.omim_ids', index=19,
      number=18, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='omimIds', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='synonyms', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.synonyms', index=20,
      number=19, type=9, cpp_type=9, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='synonyms', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='replaced_gene_id', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.replaced_gene_id', index=21,
      number=20, type=4, cpp_type=4, label=1,
      has_default_value=False, default_value=0,
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030$\n\020replaced-gene-id\022\020Replaced Gene ID', json_name='replacedGeneId', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='annotations', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptor.annotations', index=22,
      number=23, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='annotations', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
    _GENEDESCRIPTOR_GENETYPE,
    _GENEDESCRIPTOR_RNATYPE,
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=2634,
  serialized_end=4635,
)


_NOMENCLATUREAUTHORITY = _descriptor.Descriptor(
  name='NomenclatureAuthority',
  full_name='ncbi.datasets.v1alpha1.reports.NomenclatureAuthority',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='authority', full_name='ncbi.datasets.v1alpha1.reports.NomenclatureAuthority.authority', index=0,
      number=1, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030.\n\016name-authority\022\026Nomenclature Authority2\004HGNC', json_name='authority', file=DESCRIPTOR),
    _descriptor.FieldDescriptor(
      name='identifier', full_name='ncbi.datasets.v1alpha1.reports.NomenclatureAuthority.identifier', index=1,
      number=2, type=9, cpp_type=9, label=1,
      has_default_value=False, default_value=b"".decode('utf-8'),
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=b'\302\363\030%\n\007name-id\022\017Nomenclature ID2\tHGNC:4392', json_name='identifier', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=4638,
  serialized_end=4818,
)


_GENEDESCRIPTORS = _descriptor.Descriptor(
  name='GeneDescriptors',
  full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptors',
  filename=None,
  file=DESCRIPTOR,
  containing_type=None,
  fields=[
    _descriptor.FieldDescriptor(
      name='genes', full_name='ncbi.datasets.v1alpha1.reports.GeneDescriptors.genes', index=0,
      number=1, type=11, cpp_type=10, label=3,
      has_default_value=False, default_value=[],
      message_type=None, enum_type=None, containing_type=None,
      is_extension=False, extension_scope=None,
      serialized_options=None, json_name='genes', file=DESCRIPTOR),
  ],
  extensions=[
  ],
  nested_types=[],
  enum_types=[
  ],
  serialized_options=None,
  is_extendable=False,
  syntax='proto3',
  extension_ranges=[],
  oneofs=[
  ],
  serialized_start=4820,
  serialized_end=4907,
)

_PROTEIN.fields_by_name['mature_peptides'].message_type = _MATUREPEPTIDE
_GENOMICLOCATION.fields_by_name['genomic_range'].message_type = ncbi_dot_datasets_dot_v1alpha1_dot_reports_dot_common__pb2._RANGE
_GENOMICLOCATION.fields_by_name['exons'].message_type = ncbi_dot_datasets_dot_v1alpha1_dot_reports_dot_common__pb2._RANGE
_TRANSCRIPT.fields_by_name['genomic_range'].message_type = ncbi_dot_datasets_dot_v1alpha1_dot_reports_dot_common__pb2._SEQRANGESET
_TRANSCRIPT.fields_by_name['exons'].message_type = ncbi_dot_datasets_dot_v1alpha1_dot_reports_dot_common__pb2._SEQRANGESET
_TRANSCRIPT.fields_by_name['cds'].message_type = ncbi_dot_datasets_dot_v1alpha1_dot_reports_dot_common__pb2._SEQRANGESET
_TRANSCRIPT.fields_by_name['genomic_locations'].message_type = _GENOMICLOCATION
_TRANSCRIPT.fields_by_name['protein'].message_type = _PROTEIN
_TRANSCRIPT.fields_by_name['type'].enum_type = _TRANSCRIPT_TRANSCRIPTTYPE
_TRANSCRIPT_TRANSCRIPTTYPE.containing_type = _TRANSCRIPT
_GENOMICREGION.fields_by_name['gene_range'].message_type = ncbi_dot_datasets_dot_v1alpha1_dot_reports_dot_common__pb2._SEQRANGESET
_GENOMICREGION.fields_by_name['type'].enum_type = _GENOMICREGION_GENOMICREGIONTYPE
_GENOMICREGION_GENOMICREGIONTYPE.containing_type = _GENOMICREGION
_ANNOTATION.fields_by_name['assemblies_in_scope'].message_type = _ANNOTATEDASSEMBLIES
_GENEDESCRIPTOR.fields_by_name['type'].enum_type = _GENEDESCRIPTOR_GENETYPE
_GENEDESCRIPTOR.fields_by_name['rna_type'].enum_type = _GENEDESCRIPTOR_RNATYPE
_GENEDESCRIPTOR.fields_by_name['orientation'].enum_type = ncbi_dot_datasets_dot_v1alpha1_dot_reports_dot_common__pb2._ORIENTATION
_GENEDESCRIPTOR.fields_by_name['genomic_ranges'].message_type = ncbi_dot_datasets_dot_v1alpha1_dot_reports_dot_common__pb2._SEQRANGESET
_GENEDESCRIPTOR.fields_by_name['reference_standards'].message_type = _GENOMICREGION
_GENEDESCRIPTOR.fields_by_name['genomic_regions'].message_type = _GENOMICREGION
_GENEDESCRIPTOR.fields_by_name['transcripts'].message_type = _TRANSCRIPT
_GENEDESCRIPTOR.fields_by_name['proteins'].message_type = _PROTEIN
_GENEDESCRIPTOR.fields_by_name['nomenclature_authority'].message_type = _NOMENCLATUREAUTHORITY
_GENEDESCRIPTOR.fields_by_name['annotations'].message_type = _ANNOTATION
_GENEDESCRIPTOR_GENETYPE.containing_type = _GENEDESCRIPTOR
_GENEDESCRIPTOR_RNATYPE.containing_type = _GENEDESCRIPTOR
_GENEDESCRIPTORS.fields_by_name['genes'].message_type = _GENEDESCRIPTOR
DESCRIPTOR.message_types_by_name['MaturePeptide'] = _MATUREPEPTIDE
DESCRIPTOR.message_types_by_name['Protein'] = _PROTEIN
DESCRIPTOR.message_types_by_name['GenomicLocation'] = _GENOMICLOCATION
DESCRIPTOR.message_types_by_name['Transcript'] = _TRANSCRIPT
DESCRIPTOR.message_types_by_name['GenomicRegion'] = _GENOMICREGION
DESCRIPTOR.message_types_by_name['Annotation'] = _ANNOTATION
DESCRIPTOR.message_types_by_name['AnnotatedAssemblies'] = _ANNOTATEDASSEMBLIES
DESCRIPTOR.message_types_by_name['GeneDescriptor'] = _GENEDESCRIPTOR
DESCRIPTOR.message_types_by_name['NomenclatureAuthority'] = _NOMENCLATUREAUTHORITY
DESCRIPTOR.message_types_by_name['GeneDescriptors'] = _GENEDESCRIPTORS
_sym_db.RegisterFileDescriptor(DESCRIPTOR)

MaturePeptide = _reflection.GeneratedProtocolMessageType('MaturePeptide', (_message.Message,), {
  'DESCRIPTOR' : _MATUREPEPTIDE,
  '__module__' : 'ncbi.datasets.v1alpha1.reports.gene_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.MaturePeptide)
  })
_sym_db.RegisterMessage(MaturePeptide)

Protein = _reflection.GeneratedProtocolMessageType('Protein', (_message.Message,), {
  'DESCRIPTOR' : _PROTEIN,
  '__module__' : 'ncbi.datasets.v1alpha1.reports.gene_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.Protein)
  })
_sym_db.RegisterMessage(Protein)

GenomicLocation = _reflection.GeneratedProtocolMessageType('GenomicLocation', (_message.Message,), {
  'DESCRIPTOR' : _GENOMICLOCATION,
  '__module__' : 'ncbi.datasets.v1alpha1.reports.gene_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.GenomicLocation)
  })
_sym_db.RegisterMessage(GenomicLocation)

Transcript = _reflection.GeneratedProtocolMessageType('Transcript', (_message.Message,), {
  'DESCRIPTOR' : _TRANSCRIPT,
  '__module__' : 'ncbi.datasets.v1alpha1.reports.gene_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.Transcript)
  })
_sym_db.RegisterMessage(Transcript)

GenomicRegion = _reflection.GeneratedProtocolMessageType('GenomicRegion', (_message.Message,), {
  'DESCRIPTOR' : _GENOMICREGION,
  '__module__' : 'ncbi.datasets.v1alpha1.reports.gene_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.GenomicRegion)
  })
_sym_db.RegisterMessage(GenomicRegion)

Annotation = _reflection.GeneratedProtocolMessageType('Annotation', (_message.Message,), {
  'DESCRIPTOR' : _ANNOTATION,
  '__module__' : 'ncbi.datasets.v1alpha1.reports.gene_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.Annotation)
  })
_sym_db.RegisterMessage(Annotation)

AnnotatedAssemblies = _reflection.GeneratedProtocolMessageType('AnnotatedAssemblies', (_message.Message,), {
  'DESCRIPTOR' : _ANNOTATEDASSEMBLIES,
  '__module__' : 'ncbi.datasets.v1alpha1.reports.gene_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.AnnotatedAssemblies)
  })
_sym_db.RegisterMessage(AnnotatedAssemblies)

GeneDescriptor = _reflection.GeneratedProtocolMessageType('GeneDescriptor', (_message.Message,), {
  'DESCRIPTOR' : _GENEDESCRIPTOR,
  '__module__' : 'ncbi.datasets.v1alpha1.reports.gene_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.GeneDescriptor)
  })
_sym_db.RegisterMessage(GeneDescriptor)

NomenclatureAuthority = _reflection.GeneratedProtocolMessageType('NomenclatureAuthority', (_message.Message,), {
  'DESCRIPTOR' : _NOMENCLATUREAUTHORITY,
  '__module__' : 'ncbi.datasets.v1alpha1.reports.gene_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.NomenclatureAuthority)
  })
_sym_db.RegisterMessage(NomenclatureAuthority)

GeneDescriptors = _reflection.GeneratedProtocolMessageType('GeneDescriptors', (_message.Message,), {
  'DESCRIPTOR' : _GENEDESCRIPTORS,
  '__module__' : 'ncbi.datasets.v1alpha1.reports.gene_pb2'
  # @@protoc_insertion_point(class_scope:ncbi.datasets.v1alpha1.reports.GeneDescriptors)
  })
_sym_db.RegisterMessage(GeneDescriptors)


DESCRIPTOR._options = None
_PROTEIN.fields_by_name['accession_version']._options = None
_PROTEIN.fields_by_name['name']._options = None
_PROTEIN.fields_by_name['length']._options = None
_PROTEIN.fields_by_name['isoform_name']._options = None
_PROTEIN.fields_by_name['ensembl_protein']._options = None
_TRANSCRIPT.fields_by_name['name']._options = None
_TRANSCRIPT.fields_by_name['length']._options = None
_TRANSCRIPT.fields_by_name['genomic_range']._options = None
_TRANSCRIPT.fields_by_name['exons']._options = None
_TRANSCRIPT.fields_by_name['cds']._options = None
_TRANSCRIPT.fields_by_name['genomic_locations']._options = None
_TRANSCRIPT.fields_by_name['ensembl_transcript']._options = None
_TRANSCRIPT.fields_by_name['type']._options = None
_GENOMICREGION.fields_by_name['type']._options = None
_GENEDESCRIPTOR.fields_by_name['gene_id']._options = None
_GENEDESCRIPTOR.fields_by_name['symbol']._options = None
_GENEDESCRIPTOR.fields_by_name['description']._options = None
_GENEDESCRIPTOR.fields_by_name['tax_id']._options = None
_GENEDESCRIPTOR.fields_by_name['taxname']._options = None
_GENEDESCRIPTOR.fields_by_name['common_name']._options = None
_GENEDESCRIPTOR.fields_by_name['type']._options = None
_GENEDESCRIPTOR.fields_by_name['rna_type']._options = None
_GENEDESCRIPTOR.fields_by_name['orientation']._options = None
_GENEDESCRIPTOR.fields_by_name['genomic_ranges']._options = None
_GENEDESCRIPTOR.fields_by_name['transcripts']._options = None
_GENEDESCRIPTOR.fields_by_name['chromosome']._options = None
_GENEDESCRIPTOR.fields_by_name['chromosomes']._options = None
_GENEDESCRIPTOR.fields_by_name['replaced_gene_id']._options = None
_NOMENCLATUREAUTHORITY.fields_by_name['authority']._options = None
_NOMENCLATUREAUTHORITY.fields_by_name['identifier']._options = None
# @@protoc_insertion_point(module_scope)
