{
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Input of Mapping Pipeline\n",
        "\n",
        "## Input Files\n",
        "\n",
        "In order to run the pipeline for a single cell library, you need to have 2 things:\n",
        "\n",
        "1. FASTQ files generated by bcl2fastq\n",
        "    - If the samplesheet used in bcl2fastq is made by yap, you don't need to make fastq dataframe about the fastq files, just provide the path is fine.\n",
        "    - If the samplesheet is not made by yap, you need to make a FASTQ dataframe, see next step.\n",
        "2. mapping_config.ini for mapping parameters\n",
        "\n\n",
        "## FASTQ File Name Requirements\n",
        "- FASTQ file should be generated by SampleSheet made from previous step. Because the pipeline heavily relies on pre-defined FASTQ file name patterns to automatically parse uid, lane, read_type etc.\n",
        "- If the SampleSheet is not generated by previous step, FASTQ names may not support automatic parse. You need to provide the FASTQ dataframe by yourself. See the documentation of making FASTQ dataframe. The mapping summary part should also be done manually.\n"
      ],
      "metadata": {}
    },
    {
      "cell_type": "markdown",
      "source": [
        "## mapping_config.ini\n",
        "\n",
        "### What is mapping_config.ini file?\n",
        "\n",
        "- It's a place gather all adjustable parameters of mapping pipeline into a single file in [INI format](https://en.wikipedia.org/wiki/INI_file), so you don't need to put in 100 parameters in a shell command...\n",
        "\n",
        "- INI format is super simple:\n",
        "    ```\n",
        "    ; comment start with semicolon\n",
        "    [section1]\n",
        "    key1=value1\n",
        "    key2=value2\n",
        "    \n",
        "    [section2]\n",
        "    key1=value1\n",
        "    key2=value2\n",
        "    ```\n",
        "- Currently, the pipeline don't allow to change the sections and keys, so just change values according to your needs."
      ],
      "metadata": {}
    },
    {
      "cell_type": "markdown",
      "source": [
        "### How to prepare mapping_config.ini file?\n",
        "You can print out the default config file, save it to your own place and modify the value.\n",
        "\n",
        "```shell\n",
        "# MODE should be in snmC, NOMe, snmCT, snmCT-NOMe, depending on the library type\n",
        "yap default-mapping-config --mode MODE\n",
        "```\n",
        "\nHere is an example of getting snmC-seq default mapping_config.ini file. You need to change the place holders to correct value, such as providing correct barcode version (V1, V2) or path to the reference genome"
      ],
      "metadata": {}
    },
    {
      "cell_type": "code",
      "source": [
        "!yap default-mapping-config --mode mc"
      ],
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "# Executing default-mapping-config...\n",
            "; Mapping configurations\n",
            ";\n",
            "; INI format\n",
            "; [Section1]\n",
            "; KEY1 = VALUE1\n",
            "; KEY2 = VALUE2\n",
            ";\n",
            "; [Section2]\n",
            "; KEY1 = VALUE1\n",
            "; KEY2 = VALUE2\n",
            ";\n",
            "; lines start with \";\" is comment.\n",
            ";\n",
            "; NOTE: Don't change any section or key names.\n",
            "; Custom keys won't work, only change value when adjust parameters.\n",
            ";\n",
            "\n",
            "[mode]\n",
            "mode = mc\n",
            "\n\n",
            "[multiplexIndex]\n",
            "; This section is for demultiplex step\n",
            "; V1: 8 random index version\n",
            "; V2: 384 random index version\n",
            "barcode_version = USE_CORRECT_BARCODE_VERSION_HERE\n",
            "\n\n",
            "[fastqTrim]\n",
            "r1_adapter = AGATCGGAAGAGCACACGTCTGAAC\n",
            "r2_adapter = AGATCGGAAGAGCGTCGTGTAGGGA\n",
            "; Universal illumina adapter\n",
            "\n",
            "overlap = 6\n",
            "; least overlap of base and illumina adapter\n",
            "\n",
            "r1_left_cut = 10\n",
            "; constant length to trim at 5 prime end, apply before quality trim.\n",
            "; Aim to cut random primer part, determined by random primer length.\n",
            "; Random primer can impact results, see bellow:\n",
            "; https://sequencing.qcfail.com/articles/mispriming-in-pbat-libraries-causes-methylation-bias-and-poor-mapping-efficiencies/\n",
            "\n",
            "r1_right_cut = 10\n",
            "; constant length to trim at 3 prime end, apply before quality trim.\n",
            "\n",
            "r2_left_cut = 10\n",
            "; constant length to trim at 5 prime end, apply before quality trim.\n",
            "; Aim to cut Y-tailing by adaptase, exact length is uncertain.\n",
            "\n",
            "r2_right_cut = 10\n",
            "; constant length to trim at 3 prime end, apply before quality trim.\n",
            "\n",
            "quality_threshold = 20\n",
            "; reads quality score threshold for trimming.\n",
            "\n",
            "length_threshold = 30\n",
            "; reads length threshold after all trim steps.\n",
            "\n",
            "total_reads_threshold = 10000\n",
            "; total minimum reads number threshold for a cell to be analyzed in subsequent steps.\n",
            "; Change this to 1 for miseq library\n",
            "\n\n",
            "[bismark]\n",
            "bismark_reference= CHANGE_THIS_TO_YOUR_BISMARK_REFERENCE_DIR\n",
            "; reference directory of bismark\n",
            "\n",
            "read_min = 10000\n",
            "; total minimum reads number threshold for a cell to be analyzed in subsequent steps.\n",
            "; Change this to 1 for miseq library\n",
            "\n",
            "read_max = 12000000\n",
            "; total maximum reads number threshold for a cell to be analyzed in subsequent steps.\n",
            "\n",
            "remove_fastq_input = True\n",
            "; whether remove input fastq or not.\n",
            "\n\n",
            "[bamFilter]\n",
            "mapq_threshold = 10\n",
            "; reads MAPQ threshold\n",
            "\n\n",
            "[callMethylation]\n",
            "reference_fasta = CHANGE_THIS_TO_YOUR_REFERENCE_FASTA\n",
            "; reference fasta file, use the same one that bismark_mapping reference is prepared from\n",
            "\n",
            "num_upstr_bases = 0\n",
            "; number of base to include before mC\n",
            "\n",
            "num_downstr_bases = 2\n",
            "; number of base to include after mC\n",
            "\n",
            "compress_level = 5\n",
            "; ALLC file compress level\n",
            "# default-mapping-config finished.\n"
          ]
        }
      ],
      "execution_count": 1,
      "metadata": {
        "ExecuteTime": {
          "end_time": "2019-10-06T04:51:55.829577Z",
          "start_time": "2019-10-06T04:51:17.940371Z"
        }
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Mapping Modes of the config file\n",
        "yap support several different mapping mode for different experiments, which is controlled by mapping_config.ini as described bellow.\n",
        "\n",
        "#### snmC-seq2 (mc)\n",
        "- Normal snmC-seq2 library\n",
        "\n",
        "#### snmCT-seq (mct)\n",
        "- snmCT-seq library, each cell contain mixed reads from DNA and RNA\n",
        "- The major differences are:\n",
        "    - Need to do STAR mapping\n",
        "    - Filter bismark BAM file to get DNA reads\n",
        "    - Filter STAR BAM file to get RNA reads\n",
        "\n",
        "#### snmC-seq + NOMe treatment (nome)\n",
        "- snmC-seq with NOMe treatment, where the GCH contain open chromatin information, HCN contain normal methylation information.\n",
        "- The major differences are:\n",
        "    - Add additional one base in the context column of ALLC, so we can distinguish GpC sites with HpC sites\n",
        "\n",
        "####  snmCT-seq + NOMe treatment (mct-nome)\n",
        "- snmCT-seq with NOMe treatment, each cell contain mixed reads from DNA and RNA, and in DNA reads, the GCH contain open chromatin information, HCN contain normal methylation.\n",
        "- The major differences are:\n",
        "    - Need to do STAR mapping\n",
        "    - Filter bismark BAM file to get DNA reads\n",
        "    - Filter STAR BAM file to get RNA reads\n",
        "    - Add additional one base in the context column of ALLC, so we can distinguish GpC sites with HpC sites"
      ],
      "metadata": {
        "collapsed": false,
        "pycharm": {
          "name": "#%% md\n"
        }
      }
    }
  ],
  "metadata": {
    "kernelspec": {
      "name": "python3",
      "language": "python",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python",
      "version": "3.7.3",
      "mimetype": "text/x-python",
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "pygments_lexer": "ipython3",
      "nbconvert_exporter": "python",
      "file_extension": ".py"
    },
    "toc": {
      "toc_position": {},
      "skip_h1_title": false,
      "number_sections": true,
      "title_cell": "Table of Contents",
      "toc_window_display": true,
      "base_numbering": 1,
      "toc_section_display": true,
      "title_sidebar": "Contents",
      "toc_cell": false,
      "nav_menu": {},
      "sideBar": true
    },
    "pycharm": {
      "stem_cell": {
        "cell_type": "raw",
        "source": [],
        "metadata": {
          "collapsed": false
        }
      }
    },
    "kernel_info": {
      "name": "python3"
    },
    "nteract": {
      "version": "0.12.3"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}